// @builderx-core/builderx/src/components-x/demo/Editors/TextToTextInputEditor/index.tsx
import * as React from "react";
import {
  PropertyGroup,
  PropertyHeader,
  COLORS,
  FormRow,
  FormCol,
  Checkbox,
  CheckboxTick,
  CheckboxLabel,
  spaceMagnitudes
} from "components-x";

export default class TextToTextInputEditor extends React.Component<any, any> {
  render() {
    return (
      <PropertyGroup
        uiBackground={700}
        style={{ borderRadius: 2, marginTop: 4 }}
      >
        <PropertyHeader
          uiBackground={700}
          style={{
            borderRadius: 2,
            paddingTop: spaceMagnitudes[2],
            paddingBottom: spaceMagnitudes[2],
            backgroundColor: COLORS[700]
          }}
        >
          <FormRow>
            <FormCol first>
              <Checkbox style={{ color: "#FFF" }}>
                <CheckboxTick
                  noMargin
                  uiColor={50}
                  noBorder={true}
                  checked={true}
                  onChange={() => {
                    //
                  }}
                />
                <CheckboxLabel style={{ color: COLORS[300] }}>
                  Editable
                </CheckboxLabel>
              </Checkbox>
            </FormCol>
          </FormRow>
        </PropertyHeader>
      </PropertyGroup>
    );
  }
}
