// @builderx-core/builderx/jest.config.js
const tsconfig = require("./tsconfig.json");
const moduleNameMapper = require("tsconfig-paths-jest")(tsconfig);
const forEach = require("lodash").forEach;
const resolvedModules = {};
forEach(moduleNameMapper, (value, key) => {
  resolvedModules[`^${key}$`] = value;
});

module.exports = {
  collectCoverageFrom: ["src/**/*.{js,jsx,ts,tsx}"],
  setupFiles: ["<rootDir>/config/polyfills.js", "jest-canvas-mock"],
  testMatch: ["<rootDir>/src/**/?(*.)(spec|test).(j|t)s?(x)"],
  testPathIgnorePatterns: ["/smart-flex/"],
  testEnvironment: "node",
  testURL: "http://localhost",
  transform: {
    "^.+\\.(js|jsx|mjs)$": "<rootDir>/node_modules/babel-jest",
    "^.+\\.tsx?$": "ts-jest",
    "^.+\\.css$": "<rootDir>/config/jest/cssTransform.js",
    "^(?!.*\\.(js|jsx|mjs|css|json)$)": "<rootDir>/config/jest/fileTransform.js"
  },
  transformIgnorePatterns: [
    "[/\\\\]node_modules[/\\\\].+\\.(js|jsx|mjs|ts|tsx)$"
  ],
  moduleNameMapper: {
    "^react-native$": "react-native-web",
    "monaco-editor": "<rootDir>/node_modules/react-monaco-editor",
    "\\.worker.ts": "<rootDir>/src/session/__mocks__/worker.ts",
    ...resolvedModules
  },
  moduleFileExtensions: [
    "web.ts",
    "ts",
    "web.tsx",
    "tsx",
    "web.js",
    "js",
    "web.jsx",
    "jsx",
    "json",
    "node",
    "mjs"
  ],
  globals: {
    "ts-jest": {
      tsConfig: "./tsconfig.test.json"
    },
    window: true
  }
};
