// @builderx-core/builderx/src/form-builder/lib/FormBuilder/Editors/ResizingTypeEditor/ResizingTypeBox.tsx
import * as React from "react";
import {
  FormRow,
  Row,
  FormCol,
  ResizingType,
  ResizingTypeButton
} from "components-x";
import { Session } from "session";
import { ResizingTypeAttributes } from "common-interfaces";
import { Layer } from "domain-x";
import { get } from "lodash";
type propType = {
  layer: Layer;
  session: Session;
  toggleResizingType: (attribute: ResizingTypeAttributes, e: any) => void;
  additionalOptions?: any;
};
export default class ResizingTypeBox extends React.PureComponent<
  propType,
  any
> {
  onUpdate = () => {
    this.forceUpdate();
  };
  componentDidMount() {
    this.props.layer.props.style.on(
      ["top", "bottom", "left", "right", "width", "height"],
      this.onUpdate
    );
    this.props.layer.ui.on(["resizingType"], this.onUpdate);
  }
  componentDidUpdate(prevProps: propType) {
    prevProps.layer.props.style.off(
      ["top", "bottom", "left", "right"],
      this.onUpdate
    );
    prevProps.layer.ui.off(["resizingType"], this.onUpdate);
    this.props.layer.props.style.on(
      ["top", "bottom", "left", "right"],
      this.onUpdate
    );
    this.props.layer.ui.on(["resizingType"], this.onUpdate);
  }
  componentWillUnmount() {
    this.props.layer.props.style.off(
      ["top", "bottom", "left", "right"],
      this.onUpdate
    );
    this.props.layer.ui.off(["resizingType"], this.onUpdate);
  }

  render() {
    const resizingType = this.props.layer.getResizingType();
    const top = resizingType.top;
    const right = resizingType.right;
    const bottom = resizingType.bottom;
    const left = resizingType.left;
    return (
      <ResizingType>
        <Row>
          <ResizingTypeButton
            position="top"
            selected={top ? true : false}
            onClick={(e: any) => {
              this.props.toggleResizingType("top", e);
            }}
            data-tip="Fix top"
            data-for="tooltip-top"
          />
        </Row>
        <FormRow center>
          <FormCol first>
            <ResizingTypeButton
              onClick={(e: any) => {
                this.props.toggleResizingType("left", e);
              }}
              position="left"
              selected={left ? true : false}
              data-tip="Fix left"
              data-for="tooltip-right"
            />
          </FormCol>
          <FormCol first>
            <ResizingTypeButton
              onClick={(e: any) => {
                this.props.toggleResizingType("right", e);
              }}
              position="right"
              selected={right ? true : false}
              data-tip="Fix right"
              data-for="tooltip-right"
            />
          </FormCol>
        </FormRow>
        <Row>
          <ResizingTypeButton
            selected={bottom ? true : false}
            onClick={(e: any) => {
              this.props.toggleResizingType("bottom", e);
            }}
            data-tip="Fix bottom"
          />
        </Row>
      </ResizingType>
    );
  }
}
