// @builderx-core/builderx/src/components-x/demo/Editors/ScrollViewEditor/index.tsx
import * as React from "react";
import {
  FormGroup,
  PropertyGroup,
  PropertyHeader,
  COLORS,
  FormRow,
  FormCol,
  Checkbox,
  CheckboxTick,
  CheckboxLabel,
  Button,
  Icon,
  Text
} from "components-x";

export default class ScrollViewEditor extends React.Component<any, any> {
  render() {
    return (
      <React.Fragment>
        <FormGroup noPadding={true} paddingTop={true}>
          <PropertyGroup uiBackground={700} style={{ borderRadius: 2 }}>
            <PropertyHeader
              uiBackground={700}
              style={{
                paddingTop: 0,
                paddingBottom: 0,
                height: 25,
                borderTopLeftRadius: 2,
                borderTopRightRadius: 2,
                borderBottomLeftRadius: 0,
                borderBottomRightRadius: 0,
                backgroundColor: COLORS[700]
              }}
            >
              <FormRow>
                <FormCol>
                  <Checkbox style={{ color: "#FFF" }}>
                    <CheckboxTick
                      noMargin
                      noBorder={true}
                      checked={true}
                      uiColor={50}
                      onChange={() => {
                        //
                      }}
                    />

                    <CheckboxLabel
                      style={{
                        color: COLORS[300]
                      }}
                    >
                      Scrollable
                    </CheckboxLabel>
                  </Checkbox>
                </FormCol>
              </FormRow>
              <Button
                transparent={true}
                icon={<Icon name={"white-downarrow"} size={10} />}
                uiSize="xs"
              />
            </PropertyHeader>
            <FormGroup style={{ paddingTop: 0, paddingBottom: 0 }}>
              <FormRow>
                <FormCol first uiSize={4}>
                  <FormRow>
                    <Text>Direction</Text>
                  </FormRow>
                </FormCol>
                <FormCol uiSize={7} />
              </FormRow>
            </FormGroup>
          </PropertyGroup>
        </FormGroup>
      </React.Fragment>
    );
  }
}
