// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/List/list5.js
import React, { Component } from "react";
import { MaterialCommunityIcons as Icon } from "@expo/vector-icons";
import { View, FlatList, StyleSheet, Text, Image } from "react-native";
export default class cupertinoList5 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list1}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Image
                  source={require("../assets/cardImage.png")}
                  style={styles.imageStyle}
                />
                <View style={styles.content}>
                  <Text style={styles.rowText}>List Row</Text>
                  <Text style={styles.rowRightText}>16</Text>
                </View>
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  list1: {
    flex: 1,
    backgroundColor: "#FFF",
    height: 43
  },
  rowBgColor: {
    height: 43,
    flexDirection: "row",
    paddingLeft: 16,
    alignItems: "center"
  },
  imageStyle: {
    width: 29,
    height: 29,
    borderRadius: 5
  },
  content: {
    flex: 1,
    height: 43,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    marginLeft: 16,
    paddingRight: 16,
    borderBottomColor: "#CECED2",
    borderBottomWidth: 1
  },
  rowText: {
    fontSize: 17,
    color: "#212121"
  },
  rowRightText: {
    fontSize: 17,
    color: "#888"
  }
});
