// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/List/list3.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, FlatList, StyleSheet, Text } from "react-native";
export default class cupertinoList3 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list1}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <View style={styles.iconWrapper}>
                  <Icon
                    style={styles.iconStyle}
                    name="ios-wifi"
                    type="Ionicons"
                  />
                </View>
                <View style={styles.contentBgColor}>
                  <Text style={styles.rowText}>List Row</Text>
                  <Icon
                    style={styles.icon}
                    name="ios-arrow-forward"
                    type="Ionicons"
                  />
                </View>
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  list1: {
    flex: 1,
    backgroundColor: "#FFF",
    height: 43
  },
  rowBgColor: {
    height: 43,
    flexDirection: "row",
    paddingLeft: 16,
    alignItems: "center"
  },
  iconWrapper: {
    width: 29,
    height: 29,
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#007AFF",
    borderRadius: 5
  },
  iconStyle: {
    color: "#FFF",
    fontSize: 20
  },
  contentBgColor: {
    flex: 1,
    height: 43,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    marginLeft: 16,
    paddingRight: 16,
    borderBottomColor: "#CECED2",
    borderBottomWidth: 1
  },
  rowText: {
    fontSize: 17,
    color: "#212121"
  },
  icon: {
    opacity: 0.5,
    color: "#212121",
    fontSize: 24
  }
});
