// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/List/list4.js
import React, { Component } from "react";
import { MaterialCommunityIcons as Icon } from "@expo/vector-icons";
import { View, FlatList, StyleSheet, Text, Image } from "react-native";
export default class cupertinoList4 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list1}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Image
                  source={require("../assets/cardImage.png")}
                  style={styles.imageStyle}
                />
                <View style={styles.contentBgColor}>
                  <View>
                    <Text style={styles.rowText}>List Row</Text>
                    <Text style={styles.subText}>SubText</Text>
                  </View>
                  <Icon
                    style={styles.icon}
                    name="ios-arrow-forward"
                    type="Ionicons"
                  />
                </View>
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  list1: {
    flex: 1,
    backgroundColor: "#FFF",
    height: 43
  },
  rowBgColor: {
    height: 43,
    flexDirection: "row",
    paddingLeft: 16,
    alignItems: "center"
  },
  imageStyle: {
    width: 29,
    height: 29,
    borderRadius: 5
  },
  contentBgColor: {
    flex: 1,
    height: 43,
    flexDirection: "row",
    alignItems: "center",
    paddingRight: 16,
    justifycontent: "space-between",
    marginLeft: 16,
    borderBottomColor: "#CECED2",
    borderBottomWidth: 1
  },
  rowText: {
    fontSize: 15,
    lineHeight: 22,
    color: "#212121"
  },
  subText: {
    fontSize: 13,
    color: "#888"
  },
  icon: {
    opacity: 0.5,
    color: "#212121",
    fontSize: 24
  }
});
