// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/List/list2.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, StyleSheet, FlatList, Text } from "react-native";
export default class CupertinoList2 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list1}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Text style={styles.rowText}>List Row</Text>
                <Icon
                  name="ios-arrow-forward"
                  style={styles.icon}
                  type="Ionicons"
                />
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF",
    paddingTop: 8
  },
  list1: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  rowBgColor: {
    flexDirection: "row",
    padding: 16,
    paddingLeft: 0,
    height: 43,
    marginLeft: 16,
    borderBottomWidth: 1,
    borderBottomColor: "#CECED2",
    justifyContent: "space-between",
    alignItems: "center"
  },
  icon: {
    color: "#212121",
    opacity: 0.5,
    fontSize: 24
  },
  rowText: {
    color: "#212121",
    fontSize: 17
  }
});
