// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/List/list7.js
import React, { Component } from "react";
import { View, StyleSheet, FlatList, Text } from "react-native";
export default class CupertinoList7 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list1}
          ListHeaderComponent={({ item, separators }) => {
            return (
              <View style={styles.listHeaderBg}>
                <Text style={styles.headerText}>List Header</Text>
              </View>
            );
          }}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Text style={styles.rowText}>List Row</Text>
              </View>
            );
          }}
          ListFooterComponent={({ item, separators }) => {
            return (
              <View style={styles.listFooterBg}>
                <Text style={styles.footerText}>List Header</Text>
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  list1: {
    flex: 1,
    backgroundColor: "#FFF",
    paddingTop: 8
  },
  listHeaderBg: {
    flexDirection: "row",
    padding: 16,
    height: 43,
    backgroundColor: "#EFEFF4",
    borderBottomWidth: 1,
    borderBottomColor: "#CECED2"
  },
  headerText: {
    color: "#616161",
    fontSize: 13
  },
  listFooterBg: {
    flexDirection: "row",
    padding: 16,
    height: 43,
    backgroundColor: "#EFEFF4"
  },
  footerText: {
    color: "#616161",
    fontSize: 13
  },
  rowBgColor: {
    flexDirection: "row",
    padding: 16,
    paddingLeft: 0,
    height: 43,
    marginLeft: 16,
    borderBottomWidth: 1,
    borderBottomColor: "#CECED2"
  },
  rowText: {
    color: "#212121",
    fontSize: 17
  }
});
