// @builderx-core/builderx/src/smart-flex/demo/App.tsx
import * as React from "react";
import smartFlex from "../lib/smartFlex";

const overlappingLayers = {
  type: "root",
  children: [
    {
      type: "view",
      children: [],
      import: {
        layerName: "View",
        packageName: "native",
        isDefaultImport: false,
        asName: ""
      },
      styleName: "rect",
      locked: false,
      overrideKeys: [],
      template: { centerVertical: false, centerHorizontal: false },
      _top: 97.41,
      _left: 61.66,
      _height: 74.61,
      _width: 77.72,
      rcDims: {
        y: 97.41,
        x: 61.66,
        height: 74.61,
        width: 77.72
      },
      props: {
        style: {
          top: 97.41,
          left: 61.66,
          width: 77.72,
          height: 74.61,
          backgroundColor: "rgba(230, 230, 230,1)",
          position: "absolute"
        }
      }
    },

    {
      type: "view",
      children: [],
      import: {
        layerName: "View",
        packageName: "native",
        isDefaultImport: false,
        asName: ""
      },
      styleName: "rect2",
      locked: false,
      overrideKeys: [],
      template: { centerVertical: false, centerHorizontal: false },
      _top: 97.41,
      _left: 162.69,
      _height: 74.61,
      _width: 77.72,
      rcDims: {
        y: 97.41,
        x: 263.72,
        height: 74.61,
        width: 77.72
      },
      props: {
        style: {
          top: 97.41,
          left: 162.69,
          width: 77.72,
          height: 74.61,
          backgroundColor: "rgba(230, 230, 230,1)",
          position: "absolute"
        }
      }
    },
    {
      type: "view",
      children: [],
      import: {
        layerName: "View",
        packageName: "native",
        isDefaultImport: false,
        asName: ""
      },
      styleName: "rect3",
      locked: false,
      overrideKeys: [],
      template: { centerVertical: false, centerHorizontal: false },
      _top: 97.41,
      _left: 263.72,
      _height: 74.61,
      _width: 77.72,
      rcDims: {
        y: 97.41,
        x: 162.69,
        height: 74.61,
        width: 77.72
      },
      props: {
        style: {
          top: 97.41,
          left: 263.72,
          width: 77.72,
          height: 74.61,
          backgroundColor: "rgba(230, 230, 230,1)",
          position: "absolute"
        }
      }
    },
    {
      type: "view",
      children: [],
      import: {
        layerName: "View",
        packageName: "native",
        isDefaultImport: false,
        asName: ""
      },
      styleName: "rect4",
      locked: false,
      overrideKeys: [],
      template: { centerVertical: false, centerHorizontal: false },
      _top: 240.42,
      _left: 185.23,
      _height: 43.52,
      _width: 32.65,
      rcDims: {
        y: 240.42,
        x: 185.23,
        height: 43.52,
        width: 32.65
      },
      props: {
        style: {
          top: 240.42,
          left: 185.23,
          width: 32.65,
          height: 43.52,
          backgroundColor: "rgba(230, 230, 230,1)",
          position: "absolute"
        }
      }
    },
    {
      type: "view",
      children: [],
      import: {
        layerName: "View",
        packageName: "native",
        isDefaultImport: false,
        asName: ""
      },
      styleName: "rect5",
      locked: false,
      overrideKeys: [],
      template: { centerVertical: false, centerHorizontal: false },
      _top: 244.3,
      _left: 240.41,
      _height: 35.75,
      _width: 352.85,
      rcDims: {
        y: 244.3,
        x: 240.41,
        height: 35.75,
        width: 352.85
      },
      props: {
        style: {
          top: 244.3,
          left: 240.41,
          width: 352.85,
          height: 35.75,
          backgroundColor: "rgba(230, 230, 230,1)",
          position: "absolute"
        }
      }
    },
    {
      type: "view",
      children: [],
      import: {
        layerName: "View",
        packageName: "native",
        isDefaultImport: false,
        asName: ""
      },
      styleName: "rect6",
      locked: false,
      overrideKeys: [],
      template: { centerVertical: false, centerHorizontal: false },
      _top: 335.24,
      _left: 185.23,
      _height: 43.52,
      _width: 32.65,
      rcDims: {
        y: 335.24,
        x: 185.23,
        height: 43.52,
        width: 32.65
      },
      props: {
        style: {
          top: 335.24,
          left: 185.23,
          width: 32.65,
          height: 43.52,
          backgroundColor: "rgba(230, 230, 230,1)",
          position: "absolute"
        }
      }
    },
    {
      type: "view",
      children: [],
      import: {
        layerName: "View",
        packageName: "native",
        isDefaultImport: false,
        asName: ""
      },
      styleName: "rect7",
      locked: false,
      overrideKeys: [],
      template: { centerVertical: false, centerHorizontal: false },
      _top: 339.12,
      _left: 240.41,
      _height: 35.75,
      _width: 352.85,
      rcDims: {
        y: 339.12,
        x: 240.41,
        height: 35.75,
        width: 352.85
      },
      props: {
        style: {
          top: 339.12,
          left: 240.41,
          width: 352.85,
          height: 35.75,
          backgroundColor: "rgba(230, 230, 230,1)",
          position: "absolute"
        }
      }
    },
    {
      type: "view",
      children: [],
      import: {
        layerName: "View",
        packageName: "native",
        isDefaultImport: false,
        asName: ""
      },
      styleName: "rect8",
      locked: false,
      overrideKeys: [],
      template: { centerVertical: false, centerHorizontal: false },
      _top: 447.16,
      _left: 185.23,
      _height: 51.3,
      _width: 411.92,
      rcDims: {
        y: 447.16,
        x: 185.23,
        height: 51.3,
        width: 411.92
      },
      props: {
        style: {
          top: 447.16,
          left: 185.23,
          width: 411.92,
          height: 51.3,
          backgroundColor: "rgba(230, 230, 230,1)",
          position: "absolute"
        }
      }
    }
  ],
  import: {
    layerName: "View",
    packageName: "native",
    isDefaultImport: false,
    asName: ""
  },
  styleName: "root",
  rcDims: {
    y: 0,
    x: 0,
    height: 1000,
    width: 1000
  },
  locked: false,
  overrideKeys: [],
  template: { centerVertical: false, centerHorizontal: false },
  props: {
    style: {
      flex: 1
      // backgroundColor: "rgb(255,255,255)"
    }
  }
};

const file = { type: "file", layer: overlappingLayers };
console.log(overlappingLayers, "originalLayer");
console.log(smartFlex(file), "transfomedLayer");
console.log(overlappingLayers, "transfomedLayer");
const App: React.FC = () => {
  return (
    <div style={{ display: "flex", justifyContent: "center" }}>
      Welcome to BuilderX Smart Flex Algorithm
    </div>
  );
};
export default App;
