// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/ButtonDelete.ts
export default {
  path: "/src/components/CupertinoButtonDelete.js",
  isDirty: false,
  props: {
    style: {
      height: 44,
      width: 44,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "TouchableOpacity",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "transparent",
        justifyContent: "center",
        alignItems: "center",
        flexDirection: "row",
        borderRadius: 5
      }
    },
    children: [
      {
        type: "icon",
        styleName: "icon",
        import: {
          layerName: "Icon",
          packageName: "VectorIcons",
          isDefaultImport: true
        },
        props: {
          type: "Ionicons",
          name: "ios-trash",
          style: {
            color: "#000",
            fontSize: 24
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
