// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/Footer2.ts
export default {
  path: "/src/components/CupertinoFooter2.js",
  isDirty: false,
  props: {
    style: {
      height: 49,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        width: "100%",
        flexDirection: "row",
        backgroundColor: "rgba(255,255,255,1)",
        justifyContent: "space-between"
      }
    },
    children: [
      {
        type: "touchableOpacity",
        styleName: "btnWrapper1",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "column",
            alignItems: "center",
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                backgroundColor: "transparent",
                opacity: 0.8,
                fontSize: 24,
                color: {
                  type: "conditionalExpression",
                  test: {
                    type: "expression",
                    expression: "this.props.active"
                  },
                  consequent: {
                    type: "literal",
                    value: "#007AFF"
                  },
                  alternate: {
                    type: "literal",
                    value: "#616161"
                  }
                }
              },
              type: "MaterialCommunityIcons",
              name: "av-timer"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "btn1Caption",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                backgroundColor: "transparent",
                paddingTop: 4,
                fontSize: 12,
                fontFamily: "Roboto",
                color: {
                  type: "conditionalExpression",
                  test: {
                    type: "expression",
                    expression: "this.props.active"
                  },
                  consequent: {
                    type: "literal",
                    value: "#007AFF"
                  },
                  alternate: {
                    type: "literal",
                    value: "#9E9E9E"
                  }
                }
              }
            },
            children: [
              {
                type: "literal",
                value: "Dashboard"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "btnWrapper3",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "column",
            alignItems: "center",
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon2",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              type: "MaterialCommunityIcons",
              style: {
                backgroundColor: "transparent",
                opacity: 0.8,
                fontSize: 24,
                color: "#616161"
              },
              name: "library-books"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "btn3Caption",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                backgroundColor: "transparent",
                paddingTop: 4,
                fontSize: 12,
                fontFamily: "Roboto",
                color: "#9E9E9E"
              }
            },
            children: [
              {
                type: "literal",
                value: "Orderbook"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "btnWrapper4",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "column",
            alignItems: "center",
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon3",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              type: "MaterialCommunityIcons",
              style: {
                backgroundColor: "transparent",
                opacity: 0.8,
                fontSize: 24,
                color: "#616161"
              },
              name: "poll"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "btn4Caption",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                backgroundColor: "transparent",
                paddingTop: 4,
                fontSize: 12,
                fontFamily: "Roboto",
                color: "#9E9E9E"
              }
            },
            children: [
              {
                type: "literal",
                value: "Stats"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "btnWrapper2",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "column",
            alignItems: "center",
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon1",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              type: "MaterialCommunityIcons",
              style: {
                backgroundColor: "transparent",
                opacity: 0.8,
                fontSize: 24,
                color: "#616161"
              },
              name: "credit-card"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "btn2Caption",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                backgroundColor: "transparent",
                paddingTop: 4,
                fontSize: 12,
                fontFamily: "Roboto",
                color: "#9E9E9E"
              }
            },
            children: [
              {
                type: "literal",
                value: "Payment"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "btnWrapper5",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "column",
            alignItems: "center",
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon4",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                backgroundColor: "transparent",
                opacity: 0.8,
                fontSize: 24,
                color: "#616161"
              },
              name: "ios-heart-empty",
              type: "Ionicons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "btn5Caption",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                backgroundColor: "transparent",
                paddingTop: 4,
                fontSize: 12,
                fontFamily: "Roboto",
                color: "#9E9E9E"
              }
            },
            children: [
              {
                type: "literal",
                value: "Favorite"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
