// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/HeaderWithLargeTitle.ts
export default {
  path: "/src/components/CupertinoHeaderWithLargeTitle.js",
  isDirty: false,
  props: {
    style: {
      height: 96,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "#EFEFF4",
        paddingRight: 8,
        paddingLeft: 8
      }
    },
    children: [
      {
        type: "view",
        styleName: "header",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            width: "100%",
            height: 44,
            flexDirection: "row"
          }
        },
        children: [
          {
            type: "view",
            styleName: "leftWrapper",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flex: 1,
                alignItems: "flex-start",
                justifyContent: "center"
              }
            },
            children: [
              {
                type: "touchableOpacity",
                styleName: "leftIconButton",
                import: {
                  layerName: "TouchableOpacity",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    flexDirection: "row"
                  }
                },
                children: [
                  {
                    type: "icon",
                    styleName: "leftIcon",
                    import: {
                      layerName: "Icon",
                      packageName: "VectorIcons",
                      isDefaultImport: true
                    },
                    props: {
                      style: {
                        color: "#007AFF",
                        fontSize: 32
                      },
                      name: "ios-arrow-back",
                      type: "Ionicons"
                    },
                    children: [],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  },
                  {
                    type: "text",
                    styleName: "leftText",
                    import: {
                      layerName: "Text",
                      packageName: "native",
                      isDefaultImport: false
                    },
                    props: {
                      style: {
                        fontSize: 17,
                        color: "#007AFF",
                        paddingLeft: 5,
                        alignSelf: "center"
                      }
                    },
                    children: [
                      {
                        type: "literal",
                        value: "Back"
                      }
                    ],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "view",
            styleName: "rightWrapper",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flex: 1,
                alignItems: "flex-end",
                justifyContent: "center"
              }
            },
            children: [
              {
                type: "touchableOpacity",
                styleName: "button",
                import: {
                  layerName: "TouchableOpacity",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {},
                children: [
                  {
                    type: "text",
                    styleName: "rightText",
                    import: {
                      layerName: "Text",
                      packageName: "native",
                      isDefaultImport: false
                    },
                    props: {
                      style: {
                        color: "#007AFF",
                        fontSize: 17,
                        alignSelf: "center"
                      }
                    },
                    children: [
                      {
                        type: "literal",
                        value: "Action"
                      }
                    ],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "textWrapper",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            height: 52,
            paddingLeft: 5,
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "text",
            styleName: "title",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                fontSize: 34,
                fontWeight: "600",
                fontFamily: "Roboto",
                lineHeight: 40,
                color: "#000"
              },
              numberOfLines: 1
            },
            children: [
              {
                type: "literal",
                value: "Large Title"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
