// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/HeaderWithSubTitle.ts
export default {
  path: "/src/components/CupertinoHeaderWithSubTitle.js",
  isDirty: false,
  props: {
    style: {
      height: 44,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flexDirection: "row",
        backgroundColor: "#EFEFF4"
      }
    },
    children: [
      {
        type: "touchableOpacity",
        styleName: "leftIconButton",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "row",
            alignItems: "flex-start",
            alignSelf: "center",
            paddingLeft: 8,
            paddingRight: 8
          }
        },
        children: [
          {
            type: "text",
            styleName: "leftText",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                alignSelf: "center",
                fontSize: 17,
                color: "#007AFF"
              }
            },
            children: [
              {
                type: "literal",
                value: "All Files"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "textWrapper",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            alignItems: "center",
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "text",
            styleName: "title",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                fontSize: 15,
                fontWeight: "600",
                fontFamily: "Roboto",
                lineHeight: 15,
                color: "#000"
              },
              numberOfLines: 1
            },
            children: [
              {
                type: "literal",
                value: "Title"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "subTitle",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                fontSize: 12,
                color: "#8E8E93"
              },
              numberOfLines: 1
            },
            children: [
              {
                type: "literal",
                value: "SubTitle"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "rightIconButton",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            alignItems: "flex-end",
            justifyContent: "center",
            paddingLeft: 8,
            paddingRight: 8
          }
        },
        children: [
          {
            type: "text",
            styleName: "rightText",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                fontSize: 17,
                color: "#007AFF"
              }
            },
            children: [
              {
                type: "literal",
                value: "Done"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
