// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/HeaderWithBackground.ts
export default {
  path: "/src/components/CupertinoHeaderWithBackground.js",
  isDirty: false,
  props: {
    style: {
      height: 44,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flexDirection: "row",
        backgroundColor: "#1EC8FF",
        paddingRight: 8,
        paddingLeft: 8
      }
    },
    children: [
      {
        type: "view",
        styleName: "leftWrapper",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            alignItems: "flex-start",
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "touchableOpacity",
            styleName: "leftIconButton",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flexDirection: "row"
              }
            },
            children: [
              {
                type: "icon",
                styleName: "leftIcon",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  style: {
                    color: "#FFFFFF",
                    fontSize: 32
                  },
                  name: "ios-arrow-back",
                  type: "Ionicons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              },
              {
                type: "text",
                styleName: "leftText",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    alignSelf: "center",
                    fontSize: 17,
                    paddingLeft: 5,
                    color: "#FFFFFF"
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "Back"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "textWrapper",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            alignItems: "center",
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "text",
            styleName: "title",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                fontSize: 17,
                fontWeight: "600",
                fontFamily: "Roboto",
                lineHeight: 17,
                color: "#000"
              },
              numberOfLines: 1
            },
            children: [
              {
                type: "literal",
                value: "Title"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "rightWrapper",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            alignItems: "flex-end",
            justifyContent: "center"
          }
        },
        children: [
          {
            type: "touchableOpacity",
            styleName: "button",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {},
            children: [
              {
                type: "icon",
                styleName: "rightIcon",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  style: {
                    backgroundColor: "transparent",
                    color: "#FFFFFF",
                    fontSize: 32
                  },
                  name: "ios-add",
                  type: "Ionicons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
