// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/List2.ts
export default {
  path: "/src/components/CupertinoList2.js",
  isDirty: false,
  props: {
    style: {
      height: 500,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flex: 1,
        backgroundColor: "#FFF",
        paddingTop: 8
      }
    },
    children: [
      {
        type: "flatList",
        styleName: "list1",
        import: {
          layerName: "FlatList",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "#FFF"
          },
          renderItem: {
            type: "view",
            styleName: "rowBgColor",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flexDirection: "row",
                padding: 16,
                paddingLeft: 0,
                height: 43,
                marginLeft: 16,
                borderBottomWidth: 1,
                borderBottomColor: "#CECED2",
                justifyContent: "space-between",
                alignItems: "center"
              }
            },
            children: [
              {
                type: "text",
                styleName: "rowText",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    color: "#212121",
                    fontSize: 17
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "List Row"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              },
              {
                type: "icon",
                styleName: "icon",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  name: "ios-arrow-forward",
                  style: {
                    color: "#212121",
                    opacity: 0.5,
                    fontSize: 24
                  },
                  type: "Ionicons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
