// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/List6.ts
export default {
  path: "/src/components/CupertinoList6.js",
  isDirty: false,
  props: {
    style: {
      height: 500,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flex: 1,
        backgroundColor: "#FFF"
      }
    },
    children: [
      {
        type: "flatList",
        styleName: "list1",
        import: {
          layerName: "FlatList",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "#FFF",
            height: 60
          },
          renderItem: {
            type: "view",
            styleName: "rowBgColor",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                height: 60,
                flexDirection: "row",
                alignItems: "center"
              }
            },
            children: [
              {
                type: "view",
                styleName: "contentBgColor",
                import: {
                  layerName: "View",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    flex: 1,
                    height: 60,
                    flexDirection: "row",
                    alignItems: "center",
                    justifyContent: "space-between",
                    paddingLeft: 16,
                    paddingRight: 16,
                    borderBottomColor: "#CECED2",
                    borderBottomWidth: 1
                  }
                },
                children: [
                  {
                    type: "view",
                    styleName: "wrapper",
                    import: {
                      layerName: "View",
                      packageName: "native",
                      isDefaultImport: false
                    },
                    props: {},
                    children: [
                      {
                        type: "text",
                        styleName: "rowText",
                        import: {
                          layerName: "Text",
                          packageName: "native",
                          isDefaultImport: false
                        },
                        props: {
                          style: {
                            fontSize: 17,
                            lineHeight: 24,
                            color: "#212121"
                          }
                        },
                        children: [
                          {
                            type: "literal",
                            value: "List Row"
                          }
                        ],
                        locked: false,
                        template: {
                          centerVertical: false,
                          centerHorizontal: false
                        }
                      },
                      {
                        type: "text",
                        styleName: "subText",
                        import: {
                          layerName: "Text",
                          packageName: "native",
                          isDefaultImport: false
                        },
                        props: {
                          style: {
                            fontSize: 15,
                            lineHeight: 20,
                            color: "#888"
                          }
                        },
                        children: [
                          {
                            type: "literal",
                            value: "SubText"
                          }
                        ],
                        locked: false,
                        template: {
                          centerVertical: false,
                          centerHorizontal: false
                        }
                      }
                    ],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  },
                  {
                    type: "icon",
                    styleName: "icon",
                    import: {
                      layerName: "Icon",
                      packageName: "VectorIcons",
                      isDefaultImport: false
                    },
                    props: {
                      style: {
                        opacity: 0.5,
                        color: "#212121",
                        fontSize: 24
                      },
                      name: "ios-arrow-forward",
                      type: "Ionicons"
                    },
                    children: [],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
