// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/Radio.ts
export default {
  path: "/src/components/CupertinoRadio.js",
  isDirty: false,
  props: {
    style: {
      height: 40,
      width: 40,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "TouchableOpacity",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        alignItems: "center",
        justifyContent: "center",
        borderRadius: 20,
        backgroundColor: "transparent"
      }
    },
    children: [
      {
        type: "icon",
        styleName: "radioIcon",
        import: {
          layerName: "Icon",
          packageName: "VectorIcons",
          isDefaultImport: true
        },
        props: {
          name: {
            type: "conditionalExpression",
            test: {
              type: "expression",
              expression: "this.props.selected"
            },
            consequent: {
              type: "literal",
              value: "ios-radio-button-on"
            },
            alternate: {
              type: "literal",
              value: "ios-radio-button-off"
            }
          },
          type: "Ionicons",
          style: {
            color: {
              type: "conditionalExpression",
              test: {
                type: "expression",
                expression: "this.props.selected"
              },
              consequent: {
                type: "literal",
                value: "#007AFF"
              },
              alternate: {
                type: "literal",
                value: "#ccc"
              }
            },
            fontSize: 23
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
