// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/SearchBarWithMic.ts
export default {
  path: "/src/components/CupertinoSearchBarWithMic.js",
  isDirty: false,
  props: {
    style: {
      height: 44,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flexDirection: "row",
        alignItems: "center",
        backgroundColor: "#000",
        padding: 8,
        paddingRight: 0
      }
    },
    children: [
      {
        type: "view",
        styleName: "inputBox",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "row",
            backgroundColor: "#404040",
            borderRadius: 20
          }
        },
        children: [
          {
            type: "icon",
            styleName: "inputLeftIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                alignSelf: "center",
                paddingRight: 5,
                paddingLeft: 5,
                color: "#FFF",
                fontSize: 20
              },
              name: "magnify",
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "textInput",
            styleName: "inputStyle",
            import: {
              layerName: "TextInput",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                height: 32,
                flex: 1,
                alignSelf: "flex-start",
                fontSize: 15,
                fontFamily: "Roboto",
                lineHeight: 15,
                color: "#FFF"
              },
              placeholder: "Search",
              placeholderTextColor: "rgba(255,255,255,0.7)"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "touchableOpacity",
            styleName: "recordButton",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                width: 24,
                height: 24,
                backgroundColor: "rgba(0,0,0,0.4)",
                borderRadius: 13,
                justifyContent: "center",
                alignItems: "center",
                alignSelf: "center",
                marginRight: 8
              }
            },
            children: [
              {
                type: "icon",
                styleName: "inputRightIcon",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  style: {
                    alignSelf: "center",
                    opacity: 0.7,
                    paddingRight: 5,
                    paddingLeft: 5,
                    color: "#FFF",
                    fontSize: 20
                  },
                  name: "ios-mic",
                  type: "Ionicons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "rightIconButton",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            alignItems: "center",
            padding: 8
          }
        },
        children: [
          {
            type: "text",
            styleName: "rightButtonText",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                fontSize: 15,
                fontFamily: "Roboto",
                color: "#FFF"
              }
            },
            children: [
              {
                type: "literal",
                value: "Cancel"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
