// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/SearchBar.ts
export default {
  path: "/src/components/CupertinoSearchBar.js",
  isDirty: false,
  props: {
    style: {
      height: 44,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flexDirection: "row",
        alignItems: "center",
        backgroundColor: "#EFEFF4",
        padding: 8
      }
    },
    children: [
      {
        type: "touchableOpacity",
        styleName: "leftIconButton",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            padding: 8
          }
        },
        children: [
          {
            type: "icon",
            styleName: "leftIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                backgroundColor: "transparent",
                opacity: 0.6,
                color: "#007AFF",
                fontSize: 24
              },
              name: "subdirectory-arrow-right",
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "inputBox",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "row",
            backgroundColor: "#CECED2",
            borderRadius: 5
          }
        },
        children: [
          {
            type: "icon",
            styleName: "inputLeftIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                color: "#000",
                fontSize: 20,
                alignSelf: "center",
                paddingLeft: 5,
                paddingRight: 5
              },
              name: "magnify",
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "textInput",
            styleName: "inputStyle",
            import: {
              layerName: "TextInput",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                height: 32,
                alignSelf: "flex-start",
                fontSize: 15,
                fontFamily: "Roboto",
                lineHeight: 15,
                color: "#000",
                flex: 1
              },
              placeholder: "Search",
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "icon",
            styleName: "inputRightIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                color: "#000",
                fontSize: 20,
                alignSelf: "center",
                paddingLeft: 5,
                paddingRight: 5
              },
              name: "close-circle",
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "rightIconButton",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            alignItems: "center",
            padding: 8
          }
        },
        children: [
          {
            type: "icon",
            styleName: "rightIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                backgroundColor: "transparent",
                opacity: 0.6,
                color: "#007AFF",
                fontSize: 24
              },
              name: "upload",
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
