// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/SegmentWithTwoTabs.ts
export default {
  path: "/src/components/CupertinoSegmentWithTwoTabs.js",
  isDirty: false,
  props: {
    style: {
      height: 56,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flexDirection: "row",
        alignItems: "center",
        backgroundColor: "#FFF"
      }
    },
    children: [
      {
        type: "view",
        styleName: "textWrapper",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            height: 29,
            flex: 1,
            paddingLeft: 30,
            paddingRight: 30,
            flexDirection: "row"
          }
        },
        children: [
          {
            type: "touchableOpacity",
            styleName: "segmentTextWrapperLeft",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flex: 1,
                alignItems: "center",
                backgroundColor: "#007AFF",
                padding: 6,
                borderWidth: 1,
                borderColor: "#007AFF",
                borderBottomLeftRadius: 5,
                borderTopLeftRadius: 5
              }
            },
            children: [
              {
                type: "text",
                styleName: "titleLeft",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    fontSize: 13,
                    fontFamily: "Roboto",
                    color: "#FFFFFF"
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "Puppies"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "touchableOpacity",
            styleName: "segmentTextWrapperRight",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flex: 1,
                alignItems: "center",
                backgroundColor: "#FFFFFF",
                padding: 6,
                borderWidth: 1,
                borderColor: "#007AFF",
                borderBottomRightRadius: 5,
                borderTopRightRadius: 5
              }
            },
            children: [
              {
                type: "text",
                styleName: "titleRight",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    fontSize: 13,
                    fontFamily: "Roboto",
                    color: "#007AFF"
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "Cubs"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
