// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/Stepper.ts
export default {
  path: "/src/components/CupertinoStepper.js",
  isDirty: false,
  props: {
    style: {
      height: 29,
      width: 94,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flexDirection: "row",
        alignItems: "center",
        backgroundColor: "#FFF"
      }
    },
    children: [
      {
        type: "touchableOpacity",
        styleName: "leftStepper",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            alignItems: "center",
            backgroundColor: {
              type: "conditionalExpression",
              test: {
                type: "expression",
                expression: "this.props.decrement"
              },
              consequent: {
                type: "literal",
                value: "rgba(0, 122, 255,0.1)"
              },
              alternate: {
                type: "literal",
                value: "#FFFFFF"
              }
            },
            borderWidth: 1,
            borderColor: "#007AFF",
            borderBottomLeftRadius: 3,
            borderTopLeftRadius: 3,
            borderRightWidth: 0
          }
        },
        children: [
          {
            type: "icon",
            styleName: "leftIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                fontSize: 30,
                color: "#007AFF"
              },
              name: "minus",
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "rightStepper",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            alignItems: "center",
            backgroundColor: {
              type: "conditionalExpression",
              test: {
                type: "expression",
                expression: "this.props.increment"
              },
              consequent: {
                type: "literal",
                value: "rgba(0, 122, 255,0.1)"
              },
              alternate: {
                type: "literal",
                value: "#FFFFFF"
              }
            },
            borderWidth: 1,
            borderColor: "#007AFF",
            borderBottomRightRadius: 3,
            borderTopRightRadius: 3
          }
        },
        children: [
          {
            type: "icon",
            styleName: "rightIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                fontSize: 30,
                color: "#007AFF"
              },
              name: "plus",
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
