// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/Symbols/Toolbar.ts
export default {
  path: "/src/components/CupertinoToolbar.js",
  isDirty: false,
  props: {
    style: {
      height: 44,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        padding: 8,
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "space-between",
        backgroundColor: "rgba(255,255,255,0.5)"
      }
    },
    children: [
      {
        type: "touchableOpacity",
        styleName: "buttonWrapper1",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "column",
            alignItems: "center",
            minWidth: 30,
            maxWidth: 36,
            height: 38,
            borderRadius: 4,
            justifyContent: "center",
            opacity: 1
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon1",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                backgroundColor: "transparent",
                opacity: 0.8,
                color: "#007AFF",
                fontSize: 24
              },
              name: "ios-navigate",
              type: "Ionicons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "buttonWrapper2",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "column",
            alignItems: "center",
            minWidth: 30,
            maxWidth: 36,
            height: 38,
            borderRadius: 4,
            justifyContent: "center",
            opacity: 1
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon2",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                backgroundColor: "transparent",
                opacity: 0.8,
                color: "#007AFF",
                fontSize: 24
              },
              name: "television",
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "buttonWrapper3",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "column",
            alignItems: "center",
            minWidth: 30,
            maxWidth: 36,
            height: 38,
            borderRadius: 4,
            justifyContent: "center",
            opacity: 1
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon3",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                backgroundColor: "transparent",
                opacity: 0.8,
                color: "#007AFF",
                fontSize: 24
              },
              name: "ios-share",
              type: "Ionicons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "buttonWrapper4",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            flexDirection: "column",
            alignItems: "center",
            minWidth: 30,
            maxWidth: 36,
            height: 38,
            borderRadius: 4,
            justifyContent: "center",
            opacity: 1,
            backgroundColor: {
              type: "conditionalExpression",
              test: {
                type: "expression",
                expression: "this.props.active"
              },
              consequent: {
                type: "literal",
                value: "rgba(0, 122, 255,0.1)"
              },
              alternate: {
                type: "literal",
                value: "transparent"
              }
            }
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon4",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              style: {
                backgroundColor: "transparent",
                opacity: 0.8,
                color: "#007AFF",
                fontSize: 24
              },
              name: {
                type: "conditionalExpression",
                test: {
                  type: "expression",
                  expression: "this.props.active"
                },
                consequent: {
                  type: "literal",
                  value: "ios-information-circle"
                },
                alternate: {
                  type: "literal",
                  value: "ios-information-circle-outline"
                }
              },
              type: "Ionicons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
