// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/index.ts
import SharedComponentLibrary from "../../SharedComponentLibrary";
import BasicFooter from "./Symbols/BasicFooter";
import BasicTab from "./Symbols/BasicTab";
import ButtonViolet from "./Symbols/ButtonViolet";
import ButtonPrimary from "./Symbols/ButtonPrimary";
import ButtonPink from "./Symbols/ButtonPink";
import ButtonDanger from "./Symbols/ButtonDanger";
import ButtonDark from "./Symbols/ButtonDark";
import ButtonHamburger from "./Symbols/ButtonHamburger";
import ButtonTransparentHamburger from "./Symbols/ButtonTransparentHamburger";
import ButtonShare from "./Symbols/ButtonShare";
import ButtonWithVioletText from "./Symbols/ButtonWithVioletText";
import ButtonGrey from "./Symbols/ButtonGrey";
import ButtonLight from "./Symbols/ButtonLight";
import ButtonWithShadow from "./Symbols/ButtonWithShadow";
import ButtonSuccess from "./Symbols/ButtonSuccess";
import ButtonPurple from "./Symbols/ButtonPurple";
import CardBasic from "./Symbols/CardBasic";
import CardWithButtons from "./Symbols/CardWithButtons";
import CardWithImageAndTitle from "./Symbols/CardWithImageAndTitle";
import Card12 from "./Symbols/Card12";
import CardWithVerticalButtons from "./Symbols/CardWithVerticalButtons";
import CardWithContent from "./Symbols/CardWithContent";
import Card3 from "./Symbols/Card3";
import CardWithContentAndActionButtons from "./Symbols/CardWithContentAndActionButtons";
import Card5 from "./Symbols/Card5";
import Card6 from "./Symbols/Card6";
import CardWithoutImage from "./Symbols/CardWithoutImage";
import CardWithRightButtons from "./Symbols/CardWithRightButtons";
import CardWithTextOverImage from "./Symbols/CardWithTextOverImage";
import Carousel from "./Symbols/Carousel";
import Checkbox from "./Symbols/Checkbox";
import CheckboxWithLabel from "./Symbols/CheckboxWithLabel";
import ChipBasic from "./Symbols/ChipBasic";
import ChipWithCloseButton from "./Symbols/ChipWithCloseButton";
import ChipWithImage from "./Symbols/ChipWithImage";
import ChipWithImageAndCloseButton from "./Symbols/ChipWithImageAndCloseButton";
import DisabledTextbox from "./Symbols/DisabledTextbox";
import FixedLabelTextbox from "./Symbols/FixedLabelTextbox";
import Header1 from "./Symbols/Header1";
import Header2 from "./Symbols/Header2";
import Header3 from "./Symbols/Header3";
import Header4 from "./Symbols/Header4";
import HelperTextBox from "./Symbols/HelperTextBox";
import IconButtonsFooter from "./Symbols/IconButtonsFooter";
import IconTextButtonsFooter from "./Symbols/IconTextButtonsFooter";
import IconTextbox from "./Symbols/IconTextbox";
// import List1 from "./Symbols/List1";
// import List10 from "./Symbols/List10";
// import List11 from "./Symbols/List11";
// import List12 from "./Symbols/List12";
// import List13 from "./Symbols/List13";
// import List14 from "./Symbols/List14";
// import List15 from "./Symbols/List15";
// import List2 from "./Symbols/List2";
// import List3 from "./Symbols/List3";
// import List4 from "./Symbols/List4";
// import List5 from "./Symbols/List5";
// import List6 from "./Symbols/List6";
// import List7 from "./Symbols/List7";
// import List8 from "./Symbols/List8";
// import List9 from "./Symbols/List9";
import MapView from "./Symbols/MapView";
import MessageTextbox from "./Symbols/MessageTextbox";
// import Picker from "./Symbols/Picker";
import Radio from "./Symbols/Radio";
import RightIconTextbox from "./Symbols/RightIconTextbox";
import SearchBar from "./Symbols/SearchBar";
import SearchBarWithBackground from "./Symbols/SearchBarWithBackground";
import Slider from "./Symbols/Slider";
import Spinner from "./Symbols/Spinner";
import StackedLabelTextbox from "./Symbols/StackedLabelTextbox";
import Switch from "./Symbols/Switch";
import Toast1 from "./Symbols/Toast1";
import Toast2 from "./Symbols/Toast2";
import Toast3 from "./Symbols/Toast3";
import Toast4 from "./Symbols/Toast4";
import UnderlineTextbox from "./Symbols/UnderlineTextbox";

// Material Component Definitions Imports
import ButtonVioletComponentDefinition from "./ComponentDefinitions/Button/ButtonViolet";
import ButtonPrimaryComponentDefinition from "./ComponentDefinitions/Button/ButtonPrimary";
import ButtonPinkComponentDefinition from "./ComponentDefinitions/Button/ButtonPink";
import ButtonDangerComponentDefinition from "./ComponentDefinitions/Button/ButtonDanger";
import ButtonDarkComponentDefinition from "./ComponentDefinitions/Button/ButtonDark";
import ButtonHamburgerComponentDefinition from "./ComponentDefinitions/Button/ButtonHamburger";
import ButtonTransparentHamburgerComponentDefinition from "./ComponentDefinitions/Button/ButtonTransparentHamburger";
import ButtonShareComponentDefinition from "./ComponentDefinitions/Button/ButtonShare";
import ButtonWithVioletTextComponentDefinition from "./ComponentDefinitions/Button/ButtonWithVioletText";
import ButtonGreyComponentDefinition from "./ComponentDefinitions/Button/ButtonGrey";
import ButtonLightComponentDefinition from "./ComponentDefinitions/Button/ButtonLight";
import ButtonWithShadowComponentDefinition from "./ComponentDefinitions/Button/ButtonWithShadow";
import ButtonSuccessComponentDefinition from "./ComponentDefinitions/Button/ButtonSuccess";
import ButtonPurpleComponentDefinition from "./ComponentDefinitions/Button/ButtonPurple";
import CardBasicComponentDefinition from "./ComponentDefinitions/Card/CardBasic";
import CardWithButtonsComponentDefinition from "./ComponentDefinitions/Card/CardWithButtons";
import CardWithImageAndTitleComponentDefinition from "./ComponentDefinitions/Card/CardWithImageAndTitle";
import Card12ComponentDefinition from "./ComponentDefinitions/Card/Card12";
import CardWithVerticalButtonsComponentDefinition from "./ComponentDefinitions/Card/CardWithVerticalButtons";
import CardWithContentComponentDefinition from "./ComponentDefinitions/Card/CardWithContent";
import Card3ComponentDefinition from "./ComponentDefinitions/Card/Card3";
// tslint:disable-next-line: max-line-length
import CardWithContentAndActionButtonsComponentDefinition from "./ComponentDefinitions/Card/CardWithContentAndActionButtons";
import Card5ComponentDefinition from "./ComponentDefinitions/Card/Card5";
import Card6ComponentDefinition from "./ComponentDefinitions/Card/Card6";
import CardWithoutImageComponentDefinition from "./ComponentDefinitions/Card/CardWithoutImage";
import CardWithRightButtonsComponentDefinition from "./ComponentDefinitions/Card/CardWithRightButtons";
import CardWithTextOverImageComponentDefinition from "./ComponentDefinitions/Card/CardWithTextOverImage";
// import CarouselComponentDefinition from "./ComponentDefinitions/Carousel/Carousel";
import CheckboxComponentDefinition from "./ComponentDefinitions/Checkbox/Checkbox";
import CheckboxWithLabelComponentDefinition from "./ComponentDefinitions/Checkbox/CheckboxWithLabel";
import ChipBasicComponentDefinition from "./ComponentDefinitions/Chips/ChipBasic";
import ChipWithCloseButtonComponentDefinition from "./ComponentDefinitions/Chips/ChipWithCloseButton";
import ChipWithImageComponentDefinition from "./ComponentDefinitions/Chips/ChipWithImage";
import ChipWithImageAndCloseButtonComponentDefinition from "./ComponentDefinitions/Chips/ChipWithImageAndCloseButton";
import BasicFooterComponentDefinition from "./ComponentDefinitions/Footer/BasicFooter";
import IconButtonsFooterComponentDefinition from "./ComponentDefinitions/Footer/IconButtonsFooter";
import IconTextButtonsFooterComponentDefinition from "./ComponentDefinitions/Footer/IconTextButtonsFooter";
import DisabledTextboxComponentDefinition from "./ComponentDefinitions/FormInputs/DisabledTextbox";
import FixedLabelTextboxComponentDefinition from "./ComponentDefinitions/FormInputs/FixedLabelTextbox";
import HelperTextBoxComponentDefinition from "./ComponentDefinitions/FormInputs/HelperTextBox";
import IconTextboxComponentDefinition from "./ComponentDefinitions/FormInputs/IconTextbox";
// import MessageTextboxComponentDefinition from "./ComponentDefinitions/FormInputs/MessageTextbox";
import RightIconTextboxComponentDefinition from "./ComponentDefinitions/FormInputs/RightIconTextbox";
import StackedLabelTextboxComponentDefinition from "./ComponentDefinitions/FormInputs/StackedLabelTextbox";
import UnderlineTextboxComponentDefinition from "./ComponentDefinitions/FormInputs/UnderlineTextbox";
import Header1ComponentDefinition from "./ComponentDefinitions/Header/Header1";
import Header2ComponentDefinition from "./ComponentDefinitions/Header/Header2";
import Header3ComponentDefinition from "./ComponentDefinitions/Header/Header3";
import Header4ComponentDefinition from "./ComponentDefinitions/Header/Header4";
// import List1ComponentDefinition from "./ComponentDefinitions/List/List1";
// import List10ComponentDefinition from "./ComponentDefinitions/List/List10";
// import List11ComponentDefinition from "./ComponentDefinitions/List/List11";
// import List12ComponentDefinition from "./ComponentDefinitions/List/List12";
// import List13ComponentDefinition from "./ComponentDefinitions/List/List13";
// import List14ComponentDefinition from "./ComponentDefinitions/List/List14";
// import List15ComponentDefinition from "./ComponentDefinitions/List/List15";
// import List2ComponentDefinition from "./ComponentDefinitions/List/List2";
// import List3ComponentDefinition from "./ComponentDefinitions/List/List3";
// import List4ComponentDefinition from "./ComponentDefinitions/List/List4";
// import List5ComponentDefinition from "./ComponentDefinitions/List/List5";
// import List6ComponentDefinition from "./ComponentDefinitions/List/List6";
// import List7ComponentDefinition from "./ComponentDefinitions/List/List7";
// import List8ComponentDefinition from "./ComponentDefinitions/List/List8";
// import List9ComponentDefinition from "./ComponentDefinitions/List/List9";
import MapViewComponentDefinition from "./ComponentDefinitions/MapView/MapView";
// import PickerComponentDefinition from "./ComponentDefinitions/Picker/Picker";
import RadioComponentDefinition from "./ComponentDefinitions/Radio/Radio";
import SearchBarComponentDefinition from "./ComponentDefinitions/SearchBar/SearchBar";
import SearchBarWithBackgroundComponentDefinition from "./ComponentDefinitions/SearchBar/SearchBarWithBackground";
import SliderComponentDefinition from "./ComponentDefinitions/Slider/Slider";
import SpinnerComponentDefinition from "./ComponentDefinitions/Spinner/Spinner";
import SwitchComponentDefinition from "./ComponentDefinitions/Switch/Switch";
// import BasicTabComponentDefinition from "./ComponentDefinitions/Tabs/BasicTab";
import Toast1ComponentDefinition from "./ComponentDefinitions/Toast/Toast1";
import Toast2ComponentDefinition from "./ComponentDefinitions/Toast/Toast2";
import Toast3ComponentDefinition from "./ComponentDefinitions/Toast/Toast3";
import Toast4ComponentDefinition from "./ComponentDefinitions/Toast/Toast4";

import { entries } from "lodash";

export class MaterialComponentLibrary extends SharedComponentLibrary {
  constructor() {
    super();
    this.name = "Material";
    this.symbols = new Map(
      entries({
        MaterialBasicFooter: BasicFooter,
        MaterialBasicTab: BasicTab,
        MaterialButtonViolet: ButtonViolet,
        MaterialButtonPrimary: ButtonPrimary,
        MaterialButtonPink: ButtonPink,
        MaterialButtonDanger: ButtonDanger,
        MaterialButtonDark: ButtonDark,
        MaterialButtonHamburger: ButtonHamburger,
        MaterialButtonTransparentHamburger: ButtonTransparentHamburger,
        MaterialButtonShare: ButtonShare,
        MaterialButtonWithVioletText: ButtonWithVioletText,
        MaterialButtonGrey: ButtonGrey,
        MaterialButtonLight: ButtonLight,
        MaterialButtonWithShadow: ButtonWithShadow,
        MaterialButtonSuccess: ButtonSuccess,
        MaterialButtonPurple: ButtonPurple,
        MaterialCardBasic: CardBasic,
        MaterialCardWithButtons: CardWithButtons,
        MaterialCardWithImageAndTitle: CardWithImageAndTitle,
        MaterialCard12: Card12,
        MaterialCardWithVerticalButtons: CardWithVerticalButtons,
        MaterialCardWithContent: CardWithContent,
        MaterialCard3: Card3,
        MaterialCardWithContentAndActionButtons: CardWithContentAndActionButtons,
        MaterialCard5: Card5,
        MaterialCard6: Card6,
        MaterialCardWithoutImage: CardWithoutImage,
        MaterialCardWithRightButtons: CardWithRightButtons,
        MaterialCardWithTextOverImage: CardWithTextOverImage,
        MaterialCarousel: Carousel,
        MaterialCheckbox: Checkbox,
        MaterialCheckboxWithLabel: CheckboxWithLabel,
        MaterialChipBasic: ChipBasic,
        MaterialChipWithCloseButton: ChipWithCloseButton,
        MaterialChipWithImage: ChipWithImage,
        MaterialChipWithImageAndCloseButton: ChipWithImageAndCloseButton,
        MaterialDisabledTextbox: DisabledTextbox,
        MaterialFixedLabelTextbox: FixedLabelTextbox,
        MaterialHeader1: Header1,
        MaterialHeader2: Header2,
        MaterialHeader3: Header3,
        MaterialHeader4: Header4,
        MaterialHelperTextBox: HelperTextBox,
        MaterialIconButtonsFooter: IconButtonsFooter,
        MaterialIconTextButtonsFooter: IconTextButtonsFooter,
        MaterialIconTextbox: IconTextbox,
        // MaterialList1: List1,
        // MaterialList10: List10,
        // MaterialList11: List11,
        // MaterialList12: List12,
        // MaterialList13: List13,
        // MaterialList14: List14,
        // MaterialList15: List15,
        // MaterialList2: List2,
        // MaterialList3: List3,
        // MaterialList4: List4,
        // MaterialList5: List5,
        // MaterialList6: List6,
        // MaterialList7: List7,
        // MaterialList8: List8,
        // MaterialList9: List9,
        MaterialMapView: MapView,
        // MaterialMessageTextbox: MessageTextbox,
        // MaterialPicker: Picker,
        MaterialRadio: Radio,
        MaterialRightIconTextbox: RightIconTextbox,
        MaterialSearchBar: SearchBar,
        MaterialSearchBarWithBackground: SearchBarWithBackground,
        MaterialSlider: Slider,
        MaterialSpinner: Spinner,
        MaterialStackedLabelTextbox: StackedLabelTextbox,
        MaterialSwitch: Switch,
        MaterialToast1: Toast1,
        MaterialToast2: Toast2,
        MaterialToast3: Toast3,
        MaterialToast4: Toast4,
        MaterialUnderlineTextbox: UnderlineTextbox
      })
    );

    this.templates = {
      Button: [
        ButtonVioletComponentDefinition,
        ButtonPrimaryComponentDefinition,
        ButtonPinkComponentDefinition,
        ButtonDangerComponentDefinition,
        ButtonDarkComponentDefinition,
        ButtonHamburgerComponentDefinition,
        ButtonTransparentHamburgerComponentDefinition,
        ButtonShareComponentDefinition,
        ButtonWithVioletTextComponentDefinition,
        ButtonGreyComponentDefinition,
        ButtonLightComponentDefinition,
        ButtonWithShadowComponentDefinition,
        ButtonSuccessComponentDefinition,
        ButtonPurpleComponentDefinition
      ],
      Card: [
        CardBasicComponentDefinition,
        CardWithButtonsComponentDefinition,
        CardWithImageAndTitleComponentDefinition,
        Card12ComponentDefinition,
        CardWithVerticalButtonsComponentDefinition,
        CardWithContentComponentDefinition,
        Card3ComponentDefinition,
        CardWithContentAndActionButtonsComponentDefinition,
        Card5ComponentDefinition,
        Card6ComponentDefinition,
        CardWithoutImageComponentDefinition,
        CardWithRightButtonsComponentDefinition,
        CardWithTextOverImageComponentDefinition
      ],
      // Carousel: [CarouselComponentDefinition],
      Checkbox: [
        CheckboxComponentDefinition,
        CheckboxWithLabelComponentDefinition
      ],
      Chips: [
        ChipBasicComponentDefinition,
        ChipWithCloseButtonComponentDefinition,
        ChipWithImageComponentDefinition,
        ChipWithImageAndCloseButtonComponentDefinition
      ],
      Footer: [
        BasicFooterComponentDefinition,
        IconButtonsFooterComponentDefinition,
        IconTextButtonsFooterComponentDefinition
      ],
      FormInputs: [
        DisabledTextboxComponentDefinition,
        FixedLabelTextboxComponentDefinition,
        HelperTextBoxComponentDefinition,
        IconTextboxComponentDefinition,
        // MessageTextboxComponentDefinition,
        RightIconTextboxComponentDefinition,
        StackedLabelTextboxComponentDefinition,
        UnderlineTextboxComponentDefinition
      ],
      Header: [
        Header1ComponentDefinition,
        Header2ComponentDefinition,
        Header3ComponentDefinition,
        Header4ComponentDefinition
      ],
      // List: [
      //   List1ComponentDefinition,
      //   List10ComponentDefinition,
      //   List11ComponentDefinition,
      //   List12ComponentDefinition,
      //   List13ComponentDefinition,
      //   List14ComponentDefinition,
      //   List15ComponentDefinition,
      //   List2ComponentDefinition,
      //   List3ComponentDefinition,
      //   List4ComponentDefinition,
      //   List5ComponentDefinition,
      //   List6ComponentDefinition,
      //   List7ComponentDefinition,
      //   List8ComponentDefinition,
      //   List9ComponentDefinition
      // ],
      MapView: [MapViewComponentDefinition],
      // Picker: [PickerComponentDefinition],
      Radio: [RadioComponentDefinition],
      SearchBar: [
        SearchBarComponentDefinition,
        SearchBarWithBackgroundComponentDefinition
      ],
      Slider: [SliderComponentDefinition],
      Spinner: [SpinnerComponentDefinition],
      Switch: [SwitchComponentDefinition],
      // Tabs: [BasicTabComponentDefinition],
      Toast: [
        Toast1ComponentDefinition,
        Toast2ComponentDefinition,
        Toast3ComponentDefinition,
        Toast4ComponentDefinition
      ]
    };
  }
  getSize() {
    return this.symbols.size;
  }
}

export const materialComponentLibrary = new MaterialComponentLibrary();
