// @builderx-core/builderx/src/form-builder/lib/FormBuilder/Editors/LayoutEditor/index.tsx
import * as React from "react";
import { FormRow, FormCol, LayoutButton } from "components-x";
import { Layer } from "domain-x";
import { IEvent } from "common-interfaces";
import { Session } from "session";

export default class LayoutEditor extends React.PureComponent<
  {
    key: string;
    layer: Layer;
    session: Session;
    onEvent: (event: IEvent) => void;
  },
  any
> {
  constructor(props: any) {
    super(props);
    this.state = {
      showEditor: true
    };
  }

  render() {
    let isDoable = this.props.layer.splitViewIsDoable;
    // const showLayoutOptions = this.props.layer.showLayoutOptions();
    // if (showLayoutOptions) {
    return (
      <React.Fragment>
        <FormRow>
          <FormCol first>
            <LayoutButton
              disabled={!isDoable}
              onClick={(e: any) => {
                this.props.onEvent({
                  type: "splitView",
                  payload: {
                    direction: "Horizontal"
                  },
                  target: this.props.layer,
                  originalEvent: e
                });
              }}
              data-tip="Horizontal Split"
            />
          </FormCol>
          <FormCol>
            <LayoutButton
              vertical
              disabled={!isDoable}
              onClick={(e: any) => {
                this.props.onEvent({
                  type: "splitView",
                  payload: {
                    direction: "Vertical"
                  },
                  target: this.props.layer,
                  originalEvent: e
                });
              }}
              data-tip="Vertical Split"
            />
          </FormCol>
        </FormRow>
      </React.Fragment>
    );
    // } else {
    //   return null;
    // }
  }
}
