// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/FormInputs/disabledTextbox.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { StyleSheet, View, TextInput } from "react-native";
export default class DisabledTextBox extends Component {
  static containerStyle = {
    height: 43,
    width: 360
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <TextInput
          placeholder="Disabled Textbox"
          editable={false}
          style={styles.inputStyle}
        />
        <Icon
          name="information-outline"
          style={styles.iconStyle}
          type="MaterialCommunityIcons"
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    borderBottomWidth: 1,
    borderColor: "#D9D5DC",
    backgroundColor: "transparent",
    flexDirection: "row",
    alignItems: "center"
  },
  inputStyle: {
    color: "#000",
    paddingRight: 5,
    fontFamily: "Roboto",
    fontSize: 16,
    alignSelf: "stretch",
    flex: 1,
    lineHeight: 16,
    paddingTop: 16,
    paddingBottom: 8
  },
  iconStyle: {
    color: "#384850",
    fontSize: 24,
    paddingRight: 8,
    opacity: 0.7
  }
});
