// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/FormInputs/fixedLabelTextbox.js
import React, { Component } from "react";
import { StyleSheet, View, TextInput, Text } from "react-native";
export default class FixedLabel extends Component {
  static containerStyle = {
    height: 43,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Text style={styles.label}>FixedLabel</Text>
        <TextInput style={styles.inputStyle} />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  label: {
    fontFamily: "Roboto",
    fontSize: 16,
    lineHeight: 16,
    paddingTop: 16,
    paddingBottom: 8,
    color: "#000",
    opacity: 0.5,
    alignSelf: "flex-start"
  },
  root: {
    borderBottomWidth: 1,
    borderColor: "#D9D5DC",
    backgroundColor: "transparent",
    flexDirection: "row",
    paddingLeft: 16
  },
  inputStyle: {
    color: "#000",
    paddingRight: 5,
    fontFamily: "Roboto",
    fontSize: 16,
    alignSelf: "stretch",
    flex: 1,
    lineHeight: 16,
    paddingTop: 14,
    paddingBottom: 8,
    paddingLeft: 30
  }
});
