// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/FormInputs/helperTextBox.js
import React, { Component } from "react";
import { StyleSheet, View, TextInput, Text } from "react-native";
export default class HelperTextBox extends Component {
  static containerStyle = {
    height: 90,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Text style={styles.label}>StackedLabel</Text>
        <TextInput placeholder="Input" style={styles.inputStyle} />
        <Text style={styles.helper}>Helper text</Text>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  label: {
    fontFamily: "Roboto",
    fontSize: 12,
    textAlign: "left",
    color: "#000",
    opacity: 0.6,
    paddingTop: 16
  },
  helper: {
    fontFamily: "Roboto",
    fontSize: 12,
    textAlign: "left",
    color: "#000",
    opacity: 0.6,
    paddingTop: 8
  },
  root: {
    backgroundColor: "transparent"
  },
  inputStyle: {
    borderBottomWidth: 1,
    borderColor: "#D9D5DC",
    color: "#000",
    fontFamily: "Roboto",
    fontSize: 16,
    alignSelf: "stretch",
    flex: 1,
    lineHeight: 16,
    paddingTop: 8,
    paddingBottom: 8,
    width: "100%"
  }
});
