// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/FormInputs/messageTextbox.js
import React, { Component } from "react";
import { StyleSheet, View, TextInput, Text } from "react-native";
export default class MessageTextBox extends Component {
  static containerStyle = {
    width: 375,
    height: 90
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Text
          style={[
            styles.label,
            {
              color: this.props.error
                ? "red"
                : this.props.success
                ? "green"
                : "rgba(0,0,0,0.6)"
            }
          ]}
        >
          Label
        </Text>
        <TextInput
          placeholder="Input"
          style={[
            styles.inputStyle,
            {
              borderBottomColor: this.props.error
                ? "red"
                : this.props.success
                ? "green"
                : "#D9D5DC"
            }
          ]}
        />
        {this.props.error ? (
          <Text
            style={[
              styles.helper1,
              {
                color: this.props.error ? "red" : "transparent"
              }
            ]}
          >
            Error message
          </Text>
        ) : null}
        {this.props.success ? (
          <Text
            style={[
              styles.helper2,
              {
                color: this.props.success ? "green" : "transparent"
              }
            ]}
          >
            Success message
          </Text>
        ) : null}
      </View>
    );
  }
}
const styles = StyleSheet.create({
  label: {
    fontFamily: "Roboto",
    fontSize: 12,
    textAlign: "left",
    color: "#000",
    paddingTop: 16
  },
  helper1: {
    fontFamily: "Roboto",
    fontSize: 12,
    textAlign: "left",
    paddingTop: 8
  },
  helper2: {
    fontFamily: "Roboto",
    fontSize: 12,
    textAlign: "left",
    paddingTop: 8
  },
  root: {
    backgroundColor: "transparent"
  },
  inputStyle: {
    borderBottomWidth: 1,
    borderColor: "#D9D5DC",
    color: "#000",
    fontFamily: "Roboto",
    fontSize: 16,
    alignSelf: "stretch",
    flex: 1,
    lineHeight: 16,
    paddingTop: 8,
    paddingBottom: 8,
    width: "100%"
  }
});
