// @builderx-core/builderx/src/web-app/assets/Fonts/AdditionalFontList.ts
import { SystemFont } from "common-interfaces";
import urljoin from "url-join";

export function getAdditionalFonts(appAssestsURL: string): Array<SystemFont> {
  const appPath: string = urljoin(appAssestsURL, "fonts");
  const verdanaAppPath = urljoin(appPath, "Verdana");
  const calibriAppPath = urljoin(appPath, "Calibri");
  const arialAppPath = urljoin(appPath, "Arial");
  const comicSansAppPath = urljoin(appPath, "Comic Sans MS");
  const courierAppPath = urljoin(appPath, "Courier Regular");
  const timesNewRomanAppPath = urljoin(appPath, "TimesNewRoman");
  const helveticaAppPath = urljoin(appPath, "Helvetica");
  const georgiaAppPath = urljoin(appPath, "Georgia");
  const impactAppPath = urljoin(appPath, "Impact");
  const trebuchetMSAppPath = urljoin(appPath, "Trebuchet MS");

  return [
    // {
    //   path: urljoin(arialAppPath, "Arial-Narrow.ttf"),
    //   postscriptName: "arial-narrow",
    //   family: "Arial",
    //   style: "Narrow"
    // },
    {
      path: urljoin(arialAppPath, "Arial.ttf"),
      postscriptName: "arial-regular",
      family: "Arial",
      style: "Regular",
      type: "system"
    },

    ///

    {
      path: urljoin(calibriAppPath, "Calibri Regular.ttf"),
      postscriptName: "calibri-regular",
      family: "Calibri",
      style: "regular",
      type: "system"
    },
    {
      path: urljoin(calibriAppPath, "Calibri Light.ttf"),
      postscriptName: "calibri-light",
      family: "Calibri",
      style: "light",
      type: "system"
    },
    {
      path: urljoin(calibriAppPath, "Calibri Light Italic.ttf"),
      postscriptName: "calibri-light-italic",
      family: "Calibri",
      style: "light italic",
      type: "system"
    },
    {
      path: urljoin(calibriAppPath, "Calibri Bold.ttf"),
      postscriptName: "calibri-bold",
      family: "Calibri",
      style: "bold",
      type: "system"
    },
    {
      path: urljoin(calibriAppPath, "Calibri Bold Italic.ttf"),
      postscriptName: "calibri-bold-italic",
      family: "Calibri",
      style: "bold italic",
      type: "system"
    },

    {
      path: urljoin(calibriAppPath, "Calibri Italic.ttf"),
      postscriptName: "calibri-italic",
      family: "Calibri",
      style: "italic",
      type: "system"
    },

    {
      path: urljoin(comicSansAppPath, "comic-sans-ms.ttf"),
      postscriptName: "comic-sans-ms-regular",
      family: "Comic Sans MS",
      style: "Regular",
      type: "system"
    },

    {
      path: urljoin(courierAppPath, "Courier Regular.ttf"),
      postscriptName: "courier-regular",
      family: "Courier",
      style: "Regular",
      type: "system"
    },
    {
      path: urljoin(georgiaAppPath, "georgia.ttf"),
      postscriptName: "georgia-regular",
      family: "Georgia Regular",
      style: "Regular",
      type: "system"
    },
    {
      path: urljoin(helveticaAppPath, "Helvetica.ttf"),
      postscriptName: "helvetica-regular",
      family: "Helvetica",
      style: "Regular",
      type: "system"
    },
    {
      path: urljoin(helveticaAppPath, "Helvetica-Greek.ttf"),
      postscriptName: "helvetica-greek",
      family: "Helvetica",
      style: "Greek",
      type: "system"
    },
    {
      path: urljoin(helveticaAppPath, "Helvetica-Monospaced.ttf"),
      postscriptName: "helvetica-monospaced",
      family: "Helvetica",
      style: "Monospaced",
      type: "system"
    },
    {
      path: urljoin(helveticaAppPath, "Helvetica-Neue-eText.ttf"),
      postscriptName: "helvetica-neue-eText",
      family: "Helvetica",
      style: "Neue-eText",
      type: "system"
    },
    {
      path: urljoin(helveticaAppPath, "Helvetica-Neue.ttf"),
      postscriptName: "helvetica-neue",
      family: "Helvetica",
      style: "Neue",
      type: "system"
    },
    {
      path: urljoin(helveticaAppPath, "Helvetica-World.ttf"),
      postscriptName: "helvetica-world",
      family: "Helvetica",
      style: "World",
      type: "system"
    },
    {
      path: urljoin(impactAppPath, "impact.ttf"),
      postscriptName: "impact-regular",
      family: "Impact",
      style: "Regular",
      type: "system"
    },
    {
      path: urljoin(timesNewRomanAppPath, "Times-New-Roman.ttf"),
      postscriptName: "times-new-roman-regular",
      family: "Times New Roman",
      style: "Regular",
      type: "system"
    },
    {
      path: urljoin(trebuchetMSAppPath, "Trebuchet MS.ttf"),
      postscriptName: "trebuchet-ms-regular",
      family: "Trebuchet MS",
      style: "Regular",
      type: "system"
    },
    {
      path: urljoin(verdanaAppPath, "Verdana.ttf"),
      postscriptName: "verdana-regular",
      family: "Verdana",
      style: "Regular",
      type: "system"
    }
  ];
}

// AdditionalFonts.forEach(font => {
//   Utils.loadFont(font);
// });
