// @builderx-core/builderx/src/components-x/demo/Components/CodeEditor.tsx
import * as React from "react";
import { PureComponent } from "react";

import {
  Pane,
  Bar,
  Button,
  ButtonGroup,
  Icon,
  Text,
  WrapperDiv,
  Spinner,
  FormGroup
} from "../../lib";

class CodeEditor extends PureComponent<any, any> {
  render() {
    return (
      <Pane uiBackground={800} id="app-tutorial-code-editor">
        <Bar name="bx-codeEditor-btn" uiBackground={700} padder={true}>
          <ButtonGroup>
            <Button
              transparent={true}
              icon={
                <Icon
                  key="code-editor-icon"
                  name={this.props.size > 40 ? "downarrow" : "uparrow"}
                  size={12}
                  marginRight={true}
                />
              }
              caption={<Text uiSize="m">Code Editor</Text>}
            />
            <Text uiSize="m">{"index.js"}</Text>
            <Text>{"This Error"}</Text>
          </ButtonGroup>
        </Bar>
        {this.props.size > 40 && (
          <React.Fragment>
            {/* <MonacoEditor
              height={this.props.size}
              language="javascript"
              theme="vs-dark"
              value={this.getFileCode()}
              ref={(ref: any) => {
                if (ref) {
                  this.monacoRef = ref;
                }
              }}
              options={{}}
              onChange={this.onCodeChange}
              editorWillMount={this.editorWillMount}
              editorDidMount={this.editorDidMount}
            /> */}
            {false ? (
              <WrapperDiv
                style={{
                  width: "100%",
                  height: "100%",
                  background: "transparent",
                  position: "absolute",
                  top: 40,
                  zIndex: 2
                }}
                onClick={(e: any) => {
                  e.preventDefault();
                  e.stopPropagation();
                }}
              >
                <FormGroup
                  style={{
                    justifyContent: "center",
                    alignItems: "center",
                    display: "flex",
                    height: "100%"
                  }}
                >
                  <Spinner uiSize="l" />
                  <Text style={{ zIndex: 2, padding: 10 }}>
                    Generating Code. Please wait...
                  </Text>
                </FormGroup>
              </WrapperDiv>
            ) : null}
          </React.Fragment>
        )}
      </Pane>
    );
  }
}

export default CodeEditor;
