// @builderx-core/builderx/src/components-x/lib/core/Badge.tsx
import * as React from "react";
import { secondary } from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import Size from "./types/SizePropType";
import * as classNames from "classnames";
import createComponent from "../createComponent";

interface PropType extends Background, Size {}

const uiSizes = {
  l: 40,
  m: 30,
  s: 20,
  xs: 15
};

export default createComponent(
  `
    .badgeX {
      padding: .25em .4em;
      font-weight: 700;
      line-height: 1;
      text-align: center;
      flex-direction: row;
      align-items: center;
      justify-content: center;
      display: flex;
      white-space: nowrap;
      vertical-align: middle;
      border-radius: 2px;
      position: absolute;
      color: white;
      font-size: 10px;
    }

`,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiSize, uiBackground, ...remainingProps } = props;

    const className = classNames({
      badgeX: true
    });
    if (remainingProps.style) {
      delete remainingProps.style["height"];
      delete remainingProps.style["width"];
    }
    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? secondary[uiBackground] : secondary[500]
          // height: uiSize ? uiSizes[uiSize] : uiSizes["s"]
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
