// @builderx-core/builderx/src/components-x/lib/core/BarHeading.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import Size from "./types/SizePropType";
import createComponent from "../createComponent";
import * as classNames from "classnames";
import * as React from "react";

interface PropType extends Background, Size {}

const uiSizes = {
  l: 40,
  m: 35,
  s: 30
};

export default createComponent(
  `
  .barHeadingX {
    display: flex;
    font-size: 16px;
    justify-content: space-between;
    align-items: center;
  }
  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiSize, uiBackground, ...remainingProps } = props;

    const className = classNames({
      barHeadingX: true
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : "transparent",
          height: uiSize ? uiSizes[uiSize] : uiSizes["m"]
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
