// @builderx-core/builderx/src/components-x/lib/core/FormCol.tsx
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import Size from "./types/SizePropType";
import createComponent from "../createComponent";
import * as classNames from "classnames";
import * as React from "react";

interface PropType extends Size, Background {
  first?: boolean;
  contentLeft?: boolean;
  contentRight?: boolean;
  alignTop?: boolean;
  alignBottom?: boolean;
  flex?: boolean;
}
export default createComponent(
  `
  .formColX {
    display: block;
    width: 100%;
    padding: 0px 0px 0px 8px;
    align-items: center;
    justify-content: center;
    position: relative;
  }

  .formColX.first{
    padding: 0px;
  }

  .formColX.contentLeft {
    justify-content: flex-start;
  }

  .formColX.contentRight {
    justify-content: flex-end;
  }

  .formColX.alignTop {
    align-items: flex-start;
  }

  .formColX.alignBottom {
    align-items: flex-end;
  }

  .formColX.flex {
    display: flex;
  }
  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const {
      uiSize,
      uiBackground,
      first,
      contentLeft,
      contentRight,
      alignTop,
      alignBottom,
      flex,
      ...remainingProps
    } = props;

    const className = classNames({
      formColX: true,
      first: first ? true : false,
      contentLeft: contentLeft ? true : false,
      contentRight: contentRight ? true : false,
      alignTop: alignTop ? true : false,
      alignBottom: alignBottom ? true : false,
      flex: flex ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : "transparent",
          flex: uiSize ? uiSize : 1
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
