// @builderx-core/builderx/src/components-x/lib/core/Iframe.tsx
import * as React from "react";
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import Color from "./types/ColorPropType";
import createComponent from "../createComponent";
import * as classNames from "classnames";

interface PropType extends Background, Color {}
const Iframe = createComponent(
  `
  .iFrameX {
      width: 100%;
      height: 48vh;
      margin: 15px 0px;
      border-radius: 2px;

   }

  `,
  (props: PropType & React.HTMLProps<HTMLIFrameElement>) => {
    const { uiColor, uiBackground, ...remainingProps } = props;

    const className = classNames({
      iFrameX: true
    });

    return (
      <iframe
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : COLORS[700]
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);

export default class extends React.Component<any, any> {
  render() {
    const {} = this.props;
    return (
      <Iframe width="100%" height="100%" frameBorder="0" {...this.props} />
    );
  }
}
