// @builderx-core/builderx/src/components-x/lib/core/InputGroup.tsx
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import createComponent from "../createComponent";
import * as classNames from "classnames";
import * as React from "react";

interface PropType extends Background {
  disabled?: boolean;
}

export default createComponent(
  `
  .inputGroupX {
    display: flex;
  align-items: center;
  justify-content: center;
  transition: all 0.3s ease 0s;
  width: 100%;
  border-radius: 3px
  }

  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiBackground, disabled, ...remainingProps } = props;

    const className = classNames({
      inputGroupX: true,
      disabled: disabled ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          background: uiBackground ? COLORS[uiBackground] : "transparent",
          opacity: disabled !== undefined && disabled ? 0.5 : 1,
          cursor:
            disabled === undefined
              ? "pointer"
              : disabled
              ? "not-allowed"
              : "pointer",
          ...remainingProps.style
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
