// @builderx-core/builderx/src/components-x/lib/core/IOSSpinner.tsx
import COLORS from "./constants/COLORS";
import * as React from "react";
import createComponent from "../createComponent";
import * as classNames from "classnames";

import Size from "./types/SizePropType";
interface PropType extends Size {
  color?: string;
}

export default createComponent(
  `
  div.spinnerX {
    position: relative;
    display: inline-block;
  }

  div.spinnerX div {
    width: 6%;
    height: 16%;
    background: ${COLORS[500]};
    position: absolute;
    left: 49%;
    top: 43%;
    opacity: 0;
    -webkit-border-radius: 50px;
    -webkit-box-shadow: 0 0 3px rgba(0,0,0,0.2);
    -webkit-animation: fade 1s linear infinite;
  }

  @-webkit-keyframes fade {
    from {opacity: 1;}
    to {opacity: 0.25;}
  }

  div.spinnerX div.bar1 {
    -webkit-transform:rotate(0deg) translate(0, -130%);
    -webkit-animation-delay: 0s;
  }

  div.spinnerX div.bar2 {
    -webkit-transform:rotate(30deg) translate(0, -130%);
    -webkit-animation-delay: -0.9167s;
  }

  div.spinnerX div.bar3 {
    -webkit-transform:rotate(60deg) translate(0, -130%);
    -webkit-animation-delay: -0.833s;
  }
  div.spinnerX div.bar4 {
    -webkit-transform:rotate(90deg) translate(0, -130%);
    -webkit-animation-delay: -0.7497s;
  }
  div.spinnerX div.bar5 {
    -webkit-transform:rotate(120deg) translate(0, -130%);
    -webkit-animation-delay: -0.667s;
  }
  div.spinnerX div.bar6 {
    -webkit-transform:rotate(150deg) translate(0, -130%);
    -webkit-animation-delay: -0.5837s;
  }
  div.spinnerX div.bar7 {
    -webkit-transform:rotate(180deg) translate(0, -130%);
    -webkit-animation-delay: -0.5s;
  }
  div.spinnerX div.bar8 {
    -webkit-transform:rotate(210deg) translate(0, -130%);
    -webkit-animation-delay: -0.4167s;
  }
  div.spinnerX div.bar9 {
    -webkit-transform:rotate(240deg) translate(0, -130%);
    -webkit-animation-delay: -0.333s;
  }
  div.spinnerX div.bar10 {
    -webkit-transform:rotate(270deg) translate(0, -130%);
    -webkit-animation-delay: -0.2497s;
  }
  div.spinnerX div.bar11 {
    -webkit-transform:rotate(300deg) translate(0, -130%);
    -webkit-animation-delay: -0.167s;
  }
  div.spinnerX div.bar12 {
    -webkit-transform:rotate(330deg) translate(0, -130%);
    -webkit-animation-delay: -0.0833s;
  }

  `,
  React.forwardRef(
    (props: PropType & React.HTMLProps<HTMLDivElement>, ref: any) => {
      const { uiSize, color, ...remainingProps } = props;

      const height = uiSize === "l" ? 40 : 22;
      const width = uiSize === "l" ? 40 : 22;

      const className = classNames({
        spinnerX: true
      });
      return (
        <div
          ref={ref}
          className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
          style={{
            height: height,
            width: width
          }}
          {...remainingProps}
        >
          <div
            className="bar1"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar2"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar3"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar4"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar5"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar6"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar7"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar8"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar9"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar10"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar11"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
          <div
            className="bar12"
            style={{
              background: color ? color : COLORS[500]
            }}
          />
        </div>
      );
    }
  )
);
