// @builderx-core/builderx/src/components-x/lib/core/LayerAfter.tsx
import * as React from "react";
import * as classNames from "classnames";
import { secondary } from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import createComponent from "../createComponent";

interface PropType extends Background {
  open?: boolean;
  selected?: boolean;
  hover?: boolean;
  dragHoverInside?: boolean;
  dragHoverAfter?: boolean;
  level?: number;
}

export default createComponent(
  `
  .layerAfterX {
    height: 2px !important;
  }

  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const {
      uiBackground,
      dragHoverInside,
      dragHoverAfter,
      hover,
      selected,
      level,
      ...remainingProps
    } = props;

    const className = classNames({
      layerAfterX: true,
      dragHoverInside: dragHoverInside ? true : false,
      dragHoverAfter: dragHoverAfter ? true : false,
      hover: hover ? true : false,
      selected: selected ? true : false
    });
    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          borderBottom: dragHoverAfter ? `2px solid ${secondary[600]}` : "",
          marginLeft: 10 + (level ? level * 17 : 0),
          marginRight: 2
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      >
        <div
          style={{
            position: "relative",
            width: 10,
            height: 10,
            marginLeft: -10,
            top: -4,
            background: "transparent",
            border: dragHoverAfter ? `2px solid ${secondary[600]}` : undefined,
            zIndex: 10,
            borderRadius: 10
          }}
        />
      </div>
    );
  }
);
