// @builderx-core/builderx/src/components-x/lib/core/Layer.tsx
import * as React from "react";
import * as classNames from "classnames";
import COLORS, { secondary } from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import createComponent from "../createComponent";

interface PropType extends Background {
  borderBottom?: boolean;
  open?: boolean;
  dragHoverParent?: boolean;
}

export default createComponent(
  `
  .layerX {
  }
  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const {
      uiBackground,
      borderBottom,
      dragHoverParent,
      ...remainingProps
    } = props;

    const className = classNames({
      layerX: true
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: open
            ? COLORS[800]
            : uiBackground
            ? COLORS[uiBackground]
            : "transparent",
          marginRight: 2,
          marginLeft: 2,
          boxShadow: dragHoverParent
            ? `0 0 0 2px ${secondary[600]}`
            : undefined,
          borderRadius: dragHoverParent ? 4 : 0
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
