// @builderx-core/builderx/src/components-x/lib/core/Logo.tsx
import styled from "styled-components";
import Size from "./types/SizePropType";
import * as classNames from "classnames";
import createComponent from "../createComponent";
import * as React from "react";

interface PropType extends Size {}
const uiSizes = {
  xl: 120,
  l: 98,
  s: 30
};

const margins = {
  xl: 10,
  l: 10,
  s: 5
};
export default createComponent(
  `
  .logoX {
    user-select: none;
    -webkit-app-region: no-drag;
  }

  `,
  (
    props: PropType &
      React.DetailedHTMLProps<
        React.ImgHTMLAttributes<HTMLImageElement>,
        HTMLImageElement
      >
  ) => {
    const { uiSize, ...remainingProps } = props;

    const className = classNames({
      logoX: true
    });

    return (
      <img
        {...remainingProps}
        style={{
          ...remainingProps.style,

          height: uiSize ? uiSizes[uiSize] + "px" : uiSizes["s"] + "px",
          marginBottom: uiSize ? margins[uiSize] + "px" : margins["s"] + "px"
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);

// export default styled.img<PropType & React.HTMLProps<HTMLInputElement>>`
//   height: ${(props: any) =>
//     props.uiSize ? uiSizes[props.uiSize] : uiSizes["s"]}px;
//   margin-bottom: ${(props: any) =>
//     props.uiSize ? margins[props.uiSize] : margins["s"]}px;
// `;
