// @builderx-core/builderx/src/components-x/lib/core/ModalContent.tsx
import * as React from "react";
import * as classNames from "classnames";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import Size from "./types/SizePropType";
import createComponent from "../createComponent";

interface PropType extends Background, Size {
  maxHeight?: boolean;
}

const uiSizes = {
  xl: "1600px",
  l: "980px",
  ms: "600px",
  m: "500px",
  s: "430px",
  xs: "320px"
};

export default createComponent(
  `
  @keyframes modalMove{
    from {
     transform: translate3d(0, -110%, 0);
     visibility: visible;
   }
   to {
     transform: translate3d(0, 0, 0);
   } }
  .modalContentX{
    animation: modalMove 200ms linear;
    width: 100%;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    overflow: hidden;
    z-index: 300000;
    margin: auto;
    height: fit-content;
    max-height:auto;
    overflow: visible;
  }
  .modalContentX.maxHeight{
    max-height:90%;
  }
  `,
  (
    props: PropType &
      React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLDivElement>,
        HTMLDivElement
      >
  ) => {
    const {
      uiBackground,
      uiSize,
      maxHeight,
      className,
      ...remainingProps
    } = props;

    const className2 = classNames({
      modalContentX: true,
      maxHeight: maxHeight ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : "transparent",
          maxWidth: uiSize ? uiSizes[uiSize] : "auto"
        }}
        className={className2 + " " + className}
      />
    );
  }
);
