// @builderx-core/builderx/src/components-x/lib/core/ProgressBar.tsx
import * as React from "react";
import * as classNames from "classnames";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import createComponent from "../createComponent";

interface PropType extends Background {
  width?: string;
  uiSize?: number;
  progressColor?: string;
}

const Bar = createComponent(
  `
.BarX {
 width: 100%;
  height: 10px;
  border-radius: 5px;
}`,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiBackground, width, ...remainingProps } = props;
    const className = classNames({
      BarX: true
    });
    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          backgroundColor: uiBackground ? COLORS[uiBackground] : COLORS[400]
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
const Progress = createComponent(
  `
.progressX {
 height: 10px;
border-radius: 5px;
}
`,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { width, progressColor, ...remainingProps } = props;
    const className = classNames({
      progressX: true
    });
    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          backgroundColor: progressColor ? progressColor : "#3F8EEF",
          width: width ? width : "10%"
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);

// Changes for progress bar
// tslint:disable-next-line: max-line-length
// background-image: -webkit-linear-gradient(-45deg, rgba(255,255,255,.15) 25%, transparent 25%, transparent 50%, rgba(255,255,255,.15) 50%, rgba(255,255,255,.15) 75%, transparent 75%, transparent);
// tslint:disable-next-line: max-line-length
//   background-image: -o-linear-gradient(-45deg, rgba(255,255,255,.15) 25%, transparent 25%, transparent 50%, rgba(255,255,255,.15) 50%, rgba(255,255,255,.15) 75%, transparent 75%, transparent);
// tslint:disable-next-line: max-line-length
//   background-image: linear-gradient(-45deg, rgba(255,255,255,.15) 25%, transparent 25%, transparent 50%, rgba(255,255,255,.15) 50%, rgba(255,255,255,.15) 75%, transparent 75%, transparent);

//   @-webkit-keyframes progress-bar-stripes {
//     from  { background-position: 0 0; }
//     to    { background-position: 10px 0; }
//   };
//   @keyframes progress-bar-stripes {
//     from  { background-position: 0 0; }
//     to    { background-position: 10px 0; }
//   };
//   -webkit-animation: progress-bar-stripes 2s linear infinite;
//   -o-animation: progress-bar-stripes 2s linear infinite;
//   animation: progress-bar-stripes 2s linear infinite;
export default class extends React.Component<PropType, any> {
  render() {
    const { children, ...otherProps } = this.props;
    return (
      <Bar>
        <Progress {...otherProps} />
      </Bar>
    );
  }
}
