// @builderx-core/builderx/src/components-x/lib/core/QrCodeWrapper.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import createComponent from "../createComponent";
import * as classNames from "classnames";
import * as React from "react";

interface PropType extends Background {
  size?: number;
}
export default createComponent(
  `
  .qrCodeWrapperX {
      display: flex;
      align-items: center;
      justify-content: center;

   }

  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiBackground, size, ...remainingProps } = props;

    const className = classNames({
      qrCodeWrapperX: true
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : "white",
          height: size ? size + "px" : "160px",
          width: size ? size + "px" : "160px"
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);

// export default styled.div<PropType & React.HTMLProps<HTMLDivElement>>`
//   display: flex;
//   align-items: center;
//   justify-content: center;
//   background: ${(props: PropType) =>
//     props.uiBackground ? COLORS[props.uiBackground] : "white"};
//   height: ${(props: PropType) => (props.size ? `${props.size}px` : "160px")};
//   width: ${(props: PropType) => (props.size ? `${props.size}px` : "160px")};
// `;
