// @builderx-core/builderx/src/components-x/lib/core/RadioButtonBox.tsx
import styled from "styled-components";
import Color from "./types/ColorPropType";
// import COLORS from "./constants/COLORS";
import Size from "./types/SizePropType";
import createComponent from "../createComponent";
import * as classNames from "classnames";
import * as React from "react";

interface PropType extends Size, Color {
  paddingTop?: number;
}
const uiSizes = {
  l: 16,
  m: 13,
  s: 12,
  xs: 11
};
export default createComponent(
  `
  .radioButtonBoxX {
    display: flex;
      justify-content: return "flex-start";
      align-items: center;

   }

  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiSize, uiColor, paddingTop, ...remainingProps } = props;

    const className = classNames({
      radioButtonBoxX: true
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          flexGrow: uiSize ? uiSizes[uiSize] : uiSizes["s"]
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
