// @builderx-core/builderx/src/components-x/lib/core/RenderCompiledGrid.tsx
import * as React from "react";
import COLORS from "./constants/COLORS";

interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}
interface GridI {
  width: number;
  height: number;
}

export default class Grid extends React.Component<
  {
    rect: Rect;
    grid: GridI;
    index?: number;
    id: string;
  },
  any
> {
  render() {
    return (
      <div
        style={{
          position: "absolute",
          pointerEvents: "none",
          top: this.props.rect.y,
          left: this.props.rect.x,
          height: this.props.rect.height,
          width: this.props.rect.width
        }}
        className="grid-render-builderx"
      >
        <svg width="100%" height="100%" xmlns="http://www.w3.org/2000/svg">
          <defs>
            <pattern
              id={`grid_${this.props.id}_${this.props.index}`}
              width={this.props.grid.width}
              height={this.props.grid.height}
              patternUnits="userSpaceOnUse"
            >
              <rect
                width={this.props.grid.width}
                height={this.props.grid.height}
                fill="none"
              />
              <path
                d={`M ${this.props.grid.width} 0 L 0 0 0 ${
                  this.props.grid.height
                }`}
                fill="none"
                stroke={COLORS["A100"]}
                strokeWidth={1}
              />
            </pattern>
          </defs>

          <rect
            width="100%"
            height="100%"
            fill={`url(#grid_${this.props.id}_${this.props.index})`}
          />
        </svg>
      </div>
    );
  }
}
