// @builderx-core/builderx/src/components-x/lib/core/SectionFooter.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import * as classNames from "classnames";
import createComponent from "../createComponent";
import * as React from "react";

interface PropType extends Background {
  padder?: string;
}

const padders = {
  l: "40px",
  m: "30px",
  s: "20px",
  xs: "10px"
};

export default createComponent(
  `
  .sectionFooterX {
    align-items: center;
      display: flex;
      bottom: 0px;
      width: 100%;
      position: absolute;

   }



  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiBackground, padder, ...remainingProps } = props;

    const className = classNames({
      sectionFooterX: true
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : "transparent",
          padding: padder ? padders[padder] : padders[""]
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
