// @builderx-core/builderx/src/components-x/lib/core/SectionBody.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import * as React from "react";
import * as classNames from "classnames";
import createComponent from "../createComponent";

interface PropType extends Background {
  padder?: string;
  paddingTop?: number;
  marginBottom?: number;
}

const padders = {
  l: "40px",
  m: "30px",
  s: "20px",
  xs: "10px"
};
export default createComponent(
  `
  .sectionBodyX {
    display: flex;
  align-content: flex-start;
  flex-wrap: wrap;
  flex: 1;
   }

  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const {
      paddingTop,
      marginBottom,
      uiBackground,
      padder,
      ...remainingProps
    } = props;

    const className = classNames({
      sectionBodyX: true
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : "transparent",
          padding: padder ? padders[padder] : padders[""],
          paddingTop: paddingTop ? paddingTop : "",
          marginBottom: marginBottom ? marginBottom : 35
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
