// @builderx-core/builderx/src/components-x/lib/core/SectionHeader.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import * as React from "react";
import * as classNames from "classnames";
import createComponent from "../createComponent";
import Size from "./types/SizePropType";

interface PropType extends Background, Size {
  padder?: string;
  paddingTop?: number;
  paddingBottom?: number;
}

const uiSizes = {
  l: "200px",
  m: "100px",
  s: "50px"
};
const padders = {
  l: "40px",
  m: "30px",
  s: "20px",
  xs: "10px"
};
export default createComponent(
  `
  .sectionHeaderX {
    align-items: center;
     display: flex;

   }

  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const {
      paddingTop,
      paddingBottom,
      uiBackground,
      padder,
      uiSize,
      ...remainingProps
    } = props;

    const className = classNames({
      sectionHeaderX: true
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : "transparent",
          padding: padder ? padders[padder] : padders[""],
          height: uiSize ? uiSizes[uiSize] : uiSizes["s"],
          paddingBottom: paddingBottom ? paddingBottom : ""
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
