// @builderx-core/builderx/src/components-x/lib/core/SubTitle.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Size from "./types/SizePropType";
import * as React from "react";
import * as classNames from "classnames";
import createComponent from "../createComponent";

interface PropType extends Size {
  marginBottom?: Boolean;
  uiColor?: number;
  marginTop?: boolean;
}
const uiSizes = {
  l: 21,
  m: 16,
  s: 14,
  xs: 12
};
export default createComponent(
  `
  .subTitleX {
    letter-spacing: -0.27px;
    font-weight: 400;
    margin-bottom:10px;
    line-height: 1.5em;
   }
   .subTitleX.marginBottom{
    margin-bottom:50px
   }


  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const {
      uiColor,
      uiSize,
      marginBottom,
      marginTop,
      ...remainingProps
    } = props;

    const className = classNames({
      subTitleX: true,
      marginBottom: marginBottom ? true : false,
      marginTop: marginTop ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          fontSize: uiSize ? uiSizes[uiSize] + "px" : uiSizes["l"] + "px",
          color: uiColor ? COLORS[uiColor] : COLORS["A700"],
          lineHeight:
            `calc(uiSize?uiSizes[uiSize]+"px":uiSizes["s"]+"px")` + "3px"
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
