// @builderx-core/builderx/src/components-x/lib/core/TabHeader.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import Size from "./types/SizePropType";
import * as React from "react";
import * as classNames from "classnames";
import createComponent from "../createComponent";

interface PropType extends Background, Size {
  vertical?: boolean;
}

const uiSizes = {
  l: 40,
  m: 35,
  s: 26,
  xs: 18
};
export default createComponent(
  `
  .tabHeaderX {
    border-bottom:0px;
    border-right:0px;
    display:flex;
    flex-direction:row;
    cursor:pointer
   }
   .tabHeaderX.vertical{
    border-bottom:0px;
    border-right:0px;
    display:block;
    flex-direction:column
   }


  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiBackground, vertical, uiSize, height, ...remainingProps } = props;

    const className = classNames({
      tabHeaderX: true,
      vertical: vertical ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : COLORS[700],
          height: uiSize
            ? uiSizes[uiSize]
            : vertical
            ? "100%"
            : height
            ? height
            : uiSizes["s"]
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
