// @builderx-core/builderx/src/components-x/lib/core/Title.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Size from "./types/SizePropType";
import createComponent from "../createComponent";
import * as classNames from "classnames";
import * as React from "react";

interface PropType extends Size {
  marginTop?: number;
  marginBottom?: boolean;
  uiColor?: number;
  smooth?: boolean;
}
const uiSizes = {
  l: 32,
  m: 24,
  s: 18,
  xs: 14
};
export default createComponent(
  `
  .titleX {
    letter-spacing: -0.27px;
    -webkit-font-smoothing:"";
    font-weight: 400;
    margin-bottom:10px;
    flex:1;
  }
  .titleX.smooth {
    -webkit-font-smoothing:antialiased;
  }
  .titleX.marginBottom {
    margin-bottom:40px
  }
  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const {
      uiSize,
      marginTop,
      marginBottom,
      uiColor,
      smooth,
      ...remainingProps
    } = props;

    const className = classNames({
      barX: true,
      marginBottom: marginBottom ? true : false,
      smooth: smooth ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          fontSize: uiSize ? uiSizes[uiSize] + "px" : uiSizes["l"] + "px",
          color: uiColor ? COLORS[uiColor] : COLORS["A700"],
          marginTop: marginTop ? marginTop + "10px" : "0px"
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
