// @builderx-core/builderx/src/components-x/lib/core/Window.tsx
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import SizePropType from "./types/SizePropType";
import * as React from "react";
import * as classNames from "classnames";
import createComponent from "../createComponent";

interface PropType extends Background, SizePropType {
  maxHeight?: boolean;
}

const uiSizes = {
  l: "980px",
  m: "500px",
  ms: "600px",
  s: "430px",
  ss: "370px",
  xs: "320px"
};
export default createComponent(
  `
  .windowX {
    margin: auto;
    border-radius: 3px;
    display: table;
    z-index: 1000;
    overflow: hidden;
    box-shadow: 0px 0px 60px 0px rgba(1, 2, 2, 0.5);
  }
  .windowX.maxHeight{
    overflow: scroll;
  }
  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiSize, uiBackground, maxHeight, ...remainingProps } = props;

    const className = classNames({
      windowX: true,
      maxHeight: maxHeight ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : COLORS[800],
          width: uiSize ? uiSizes[uiSize] : "100%"
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
