// @builderx-core/builderx/src/components-x/lib/core/WindowBody.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import Size from "./types/SizePropType";
import * as React from "react";
import * as classNames from "classnames";
import createComponent from "../createComponent";

interface PropType extends Background, Size {
  padder?: boolean;
  maxHeight?: boolean;
}

const uiSizes = {
  l: "900px",
  ms: "580px",
  m: "520px",
  s: "200px"
};
export default createComponent(
  `
  .windowBodyX {
    display: flex;
    width: 100%;
    position: relative;
    overflow:hidden;
    max-height:auto;

  }
  .windowBodyX.maxHeight{
    overflow:scroll;
    max-height:650px
  }
  .windowBodyX.padder {
    padding:30px 0px
  }
  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const {
      uiSize,
      uiBackground,
      padder,
      maxHeight,
      ...remainingProps
    } = props;

    const className = classNames({
      windowBodyX: true,
      maxHeight: maxHeight ? true : false,
      padder: padder ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : "transparent",
          minHeight: uiSize ? uiSizes[uiSize] : uiSizes[""]
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
