// @builderx-core/builderx/src/components-x/lib/core/WindowFooter.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import Size from "./types/SizePropType";
import * as React from "react";
import * as classNames from "classnames";
import createComponent from "../createComponent";

interface PropType extends Background, Size {
  padder?: boolean;
  noBorder?: boolean;
}

const uiSizes = {
  l: 40,
  m: 35,
  s: 30,
  xs: 10
};
export default createComponent(
  `
  .windowFooterX {
  width:100%;
  }
  .borderTop{
    border-top: 1px solid ${COLORS[700]};
  }
  .windowFooterX.padder {
    padding:15px 10px
  }
  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiSize, uiBackground, padder, noBorder, ...remainingProps } = props;

    const className = classNames({
      windowFooterX: true,
      borderTop: noBorder ? false : true,
      padder: padder ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : COLORS[600],
          height: uiSize ? uiSizes[uiSize] + "px" : "auto"
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
