// @builderx-core/builderx/src/components-x/lib/core/WindowHeader.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Background from "./types/BackgroundPropType";
import Size from "./types/SizePropType";
import * as React from "react";
import * as classNames from "classnames";
import createComponent from "../createComponent";

interface PropType extends Background, Size {}

const uiSizes = {
  l: 40,
  m: 35,
  s: 30,
  xs: 18
};
export default createComponent(
  `
  .windowHeaderX {
    border-bottom: none;
    display:flex;

  }

  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const {
      uiSize,
      uiBackground,

      ...remainingProps
    } = props;

    const className = classNames({
      windowHeaderX: true
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          background: uiBackground ? COLORS[uiBackground] : COLORS[600],
          height: uiSize ? uiSizes[uiSize] + "px" : "auto"
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
