// @builderx-core/builderx/src/components-x/lib/core/WindowTitle.tsx
import styled from "styled-components";
import COLORS from "./constants/COLORS";
import Size from "./types/SizePropType";
import * as React from "react";
import * as classNames from "classnames";
import createComponent from "../createComponent";

interface PropType extends Size {
  marginTop?: number;
  marginBottom?: boolean;
}
const uiSizes = {
  l: 32,
  m: 22,
  s: 16,
  xs: 14
};
export default createComponent(
  `
  .windowTitleX {
    color: ${COLORS[50]};
    letter-spacing: -0.27px;
     font-weight: 300;
     flex:1;
     margin-bottom:10px

  }
  .windowTitleX.marginBottom{
    margin-bottom:50px
  }

  `,
  (props: PropType & React.HTMLProps<HTMLDivElement>) => {
    const { uiSize, marginBottom, marginTop, ...remainingProps } = props;

    const className = classNames({
      windowTitleX: true,
      marginBottom: marginBottom ? true : false
    });

    return (
      <div
        {...remainingProps}
        style={{
          ...remainingProps.style,
          fontSize: uiSize ? uiSizes[uiSize] + "px" : uiSizes["l"] + "px",
          marginTop: marginTop ? marginTop + "px" : "10px"
        }}
        className={
          className +
          (remainingProps.className ? " " + remainingProps.className : "")
        }
      />
    );
  }
);
