// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/SearchBar/searchBar.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import {
  View,
  StyleSheet,
  Text,
  TouchableOpacity,
  TextInput
} from "react-native";
export default class SearchBar extends Component {
  static containerStyle = {
    height: 56,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <View style={styles.rect1}>
          <TouchableOpacity style={styles.leftIconButton}>
            <Icon
              name="arrow-left"
              style={styles.leftIcon}
              type="MaterialCommunityIcons"
            />
          </TouchableOpacity>
          <TextInput placeholder="Search" style={styles.inputStyle} />
          <TouchableOpacity style={styles.rightIconButton}>
            <Icon
              name="close"
              style={styles.rightIcon}
              type="MaterialCommunityIcons"
            />
          </TouchableOpacity>
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    backgroundColor: "#3F51B5",
    padding: 4,
    shadowColor: "#111",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.2,
    shadowRadius: 1.2,
    elevation: 3
  },
  rect1: {
    flexDirection: "row",
    backgroundColor: "#FFFFFF",
    alignItems: "center",
    position: "absolute",
    top: 4,
    bottom: 4,
    left: 4,
    right: 4,
    borderRadius: 2
  },
  leftIcon: {
    backgroundColor: "transparent",
    color: "#000",
    fontFamily: "Roboto",
    fontSize: 24,
    opacity: 0.6
  },
  leftIconButton: {
    padding: 11,
    position: "absolute",
    top: 5,
    left: 5
  },
  inputStyle: {
    height: 48,
    color: "#000",
    paddingRight: 5,
    fontFamily: "Roboto",
    fontSize: 16,
    alignSelf: "flex-start",
    width: "70%",
    lineHeight: 16,
    position: "absolute",
    left: 72,
    top: 4
  },
  rightIconButton: {
    padding: 11,
    position: "absolute",
    top: 5,
    right: 5,
    alignItems: "center"
  },
  rightIcon: {
    backgroundColor: "transparent",
    color: "#000",
    fontFamily: "Roboto",
    fontSize: 24,
    opacity: 0.6
  }
});
