// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Checkbox/checkboxWithLabel.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { TouchableOpacity, StyleSheet, Text } from "react-native";
export default class CheckBoxWithLabel extends Component {
  static containerStyle = {
    height: 40,
    width: 90
  };
  render() {
    return (
      <TouchableOpacity style={[styles.root, this.props.style]}>
        <Icon
          name={
            this.props.checked ? "checkbox-marked" : "checkbox-blank-outline"
          }
          style={styles.checkIcon}
          type="MaterialCommunityIcons"
        />
        <Text style={styles.checkLabel}>{this.props.label || "Label"}</Text>
      </TouchableOpacity>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    alignItems: "center",
    justifyContent: "center",
    borderRadius: 20,
    backgroundColor: "transparent",
    flexDirection: "row"
  },
  checkIcon: {
    color: "#3F51B5",
    fontFamily: "Roboto",
    fontSize: 28,
    lineHeight: 28
  },
  checkLabel: { marginLeft: 2, fontSize: 16, color: "rgba(0,0,0,0.87)" }
});
