// @builderx-core/builderx/src/components-x/lib/core/DeviceFrames/iPhoneX.tsx
import * as React from "react";

export default class OverLayImage extends React.Component<
  { deviceOrientation: string; top: number; left: number },
  any
> {
  render() {
    const { deviceOrientation, left, top } = this.props;
    const computedLeft =
      deviceOrientation === "landscape" ? left + 211 : left - 8;
    const computedTop = deviceOrientation === "landscape" ? top - 225 : top - 8;
    const computedTransform =
      deviceOrientation === "landscape" ? `rotate(270deg)` : "none";

    const styleObject = {
      position: "absolute",
      pointerEvents: "none",
      zIndex: 1,
      left: computedLeft,
      top: computedTop,
      transform: computedTransform
    };
    return (
      <div style={styleObject as any}>
        <svg
          width="388px"
          height="824px"
          viewBox="-2 -2 388 824"
          version="1.1"
          xmlnsXlink="http://www.w3.org/1999/xlink"
        >
          <path
            d={`M46,6 C23.909,6 6,23.909 6,46 L6,778 C6,800.091 23.909,818 46,818 L341,818 C363.091,818 381,800.091 381,
            778 L381,46 C381,23.909 363.091,6 341,6 L46,6 Z M112.04,37 C99.337,37 89.04,26.703 89.04,13.961
            L89.046,10.037 C89.0345531,9.99058934 89.0205314,9.94485191 89.004,9.9 C88.334,8.086 86.78,6.674
            84.892,6.216 L84,6 L303,6 L302.108,6.216 C300.210771,6.68271966 298.67094,8.06429185 298.002,9.9
            C297.970221,9.98647005 297.953971,10.0778752 297.954,10.17 L297.961,14 C297.961,26.703 287.663,37
            274.961,37 L112.039,37 L112.04,37 Z M0,0 L388,0 L388,824 L0,824 L0,0 Z M128.5,800 C127.119288,800
            126,801.119288 126,802.5 C126,803.880712 127.119288,805 128.5,805 L258.5,805 C259.880712,805 261,803.880712
            261,802.5 C261,801.119288 259.880712,800 258.5,800 L128.5,800 Z`}
            fill="#171A21"
            fillRule="evenodd"
            id="Shape"
          />
        </svg>
      </div>
    );
  }
}
