// @builderx-core/builderx/src/form-builder/lib/FormBuilder/Editors/MarginPaddingEditor/PaddingBox.tsx
import * as React from "react";

import {
  PaddingLayout,
  PaddingLock,
  PaddingTopSelector as TopSelector,
  PaddingLeftSelector as LeftSelector,
  PaddingRightSelector as RightSelector,
  PaddingBottomSelector as BottomSelector,
  FormGroup,
  FormRow,
  FormCol,
  Text,
  Icon,
  Slider,
  InputGroupInput,
  WrapperDiv
} from "components-x";
import { BoxModel } from "domain-x";
import { Session } from "session";
import { debounce } from "lodash";

export default class extends React.Component<
  {
    title: string;
    session: Session;
    currentValue: number;
    padding: BoxModel;
    onChangeEnd: (e: any, eventName: string) => void;
    onClickLockIcon: (e: any) => void;
    updateEditorState: (prop: string) => void;
  },
  any
> {
  render() {
    return (
      <FormGroup noPadding style={{ display: "table-caption" }}>
        <FormCol first>
          <Text uiSize="m">
            {this.props.title ? this.props.title : "Padding Editor Title"}
          </Text>
        </FormCol>
        <FormCol first uiSize={4} style={{ paddingTop: 12 }}>
          <WrapperDiv uiBackground={700} style={{ padding: 8 }}>
            <PaddingLayout>
              <TopSelector
                selected={this.props.padding.top}
                onClick={(e: any) => {
                  this.props.updateEditorState("paddingTop");
                }}
              />
              <LeftSelector
                selected={this.props.padding.left}
                onClick={(e: any) => {
                  this.props.updateEditorState("paddingLeft");
                }}
              />
              <RightSelector
                selected={this.props.padding.right}
                onClick={(e: any) => {
                  this.props.updateEditorState("paddingRight");
                }}
              />
              <BottomSelector
                selected={this.props.padding.bottom}
                onClick={(e: any) => {
                  this.props.updateEditorState("paddingBottom");
                }}
              />
              <PaddingLock>
                <Icon
                  name={this.props.padding.lock ? "lock" : "unlock"}
                  size={16}
                  onClick={this.props.onClickLockIcon}
                />
              </PaddingLock>
            </PaddingLayout>
          </WrapperDiv>
        </FormCol>
        <FormCol first>
          <FormRow style={{ paddingLeft: 0, paddingRight: 0 }}>
            <FormCol first uiSize={3}>
              <Slider
                min={0}
                max={100}
                value={this.props.currentValue ? this.props.currentValue : 0}
                onChange={(e: any) => {
                  this.props.onChangeEnd(e, "onChange");
                }}
                onMouseUp={(e: any) => {
                  this.props.onChangeEnd(e, "onMouseUp");
                }}
                onMouseDown={(e: any) => {
                  this.props.onChangeEnd(e, "onMouseDown");
                }}
              />
            </FormCol>
            <FormCol uiSize={2}>
              <InputGroupInput
                undo={this.props.session.commandManager.undo}
                redo={this.props.session.commandManager.redo}
                type="number"
                validator={e => !isNaN(parseInt(e.target.value, 10))}
                value={this.props.currentValue ? this.props.currentValue : 0}
                // onChange={(e: any) => {
                //   this.props.onChangeEnd(e, "onChange");
                // }}

                // onFocus={(e: any) => this.props.onChangeEnd(e, "onMouseDown")}
                // onBlur={(e: any) => this.props.onChangeEnd(e, "onMouseUp")}
                onChange={debounce(
                  (e: any) => this.props.onChangeEnd(e, "onMouseUp"),
                  600
                )}
              />
            </FormCol>
          </FormRow>
        </FormCol>
      </FormGroup>
    );
  }
}
