// @builderx-core/builderx/src/form-builder/lib/FormBuilder/Editors/Flexbox/ColumnElement/index.tsx
import * as React from "react";
import { Component } from "react";
import { Button, Icon, ButtonGroup, FormRow, FormCol } from "components-x";
import { Layer } from "domain-x";
import { IEvent } from "common-interfaces";
import { Session } from "session";
type propType = {
  layer: Layer;
  session: Session;
  onEvent: (event: IEvent) => void;
};
class ColumnElement extends Component<propType, any> {
  onUpdate = (key: string) => {
    this.forceUpdate();
  };
  componentDidMount() {
    this.props.layer.props.style.on(
      ["flexDirection", "alignItems", "justifyContent"],
      this.onUpdate
    );
  }
  componentDidUpdate(prevProps: propType) {
    prevProps.layer.props.style.off(
      ["flexDirection", "alignItems", "justifyContent"],
      this.onUpdate
    );
    this.props.layer.props.style.on(
      ["flexDirection", "alignItems", "justifyContent"],
      this.onUpdate
    );
  }
  componentWillUnmount() {
    this.props.layer.props.style.off(
      ["flexDirection", "alignItems", "justifyContent"],
      this.onUpdate
    );
  }

  onClick(event: any, key: string, value: string | undefined) {
    this.props.onEvent({
      payload: {
        key: key,
        value: value
      },
      originalEvent: event,
      target: this.props.layer,
      type: "setStyleAttribute"
    });
  }
  render() {
    const layerProps = this.props.layer.props;
    const alignItems =
      layerProps.getResolvedStyleAttribute("alignItems") || "flex-start";
    const justifyContent = layerProps.getResolvedStyleAttribute(
      "justifyContent"
    );
    const flexDirection = layerProps.getResolvedStyleAttribute("flexDirection");
    const isDisabled = !this.props.layer.setFlexPropertyIsDoable; // not of doable
    return (
      <>
        <FormRow>
          <FormCol first>
            <ButtonGroup uiSize="xs">
              <Button
                data-cypress-id="vertical-justify-flex-start-button"
                active={justifyContent === "flex-start" ? true : false}
                flexible
                borderRight
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value =
                    justifyContent === "flex-start" ? undefined : "flex-start";
                  this.onClick(e, "justifyContent", value);
                }}
                icon={
                  <Icon
                    size={14}
                    name={
                      flexDirection === "column-reverse"
                        ? "flexlayout-vertical-end"
                        : "flexlayout-vertical-start"
                    }
                  />
                }
                data-tip="Align layers to top"
                data-for="tooltip-right"
              />
              <Button
                data-cypress-id="vertical-justify-center-button"
                active={justifyContent === "center" ? true : false}
                flexible
                borderRight
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value =
                    justifyContent === "center" ? undefined : "center";
                  this.onClick(e, "justifyContent", value);
                }}
                icon={
                  <Icon size={14} name={"flexlayout-vertical-justify-center"} />
                }
                data-tip="Align layers to center"
              />
              <Button
                data-cypress-id="vertical-justify-flex-end-button"
                active={justifyContent === "flex-end" ? true : false}
                flexible
                borderRight
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value =
                    justifyContent === "flex-end" ? undefined : "flex-end";
                  this.onClick(e, "justifyContent", value);
                }}
                icon={
                  <Icon
                    size={14}
                    name={
                      flexDirection === "column-reverse"
                        ? "flexlayout-vertical-start"
                        : "flexlayout-vertical-end"
                    }
                  />
                }
                data-tip="Align layers to bottom"
              />
              <Button
                data-cypress-id="vertical-justify-space-between-button"
                active={justifyContent === "space-between" ? true : false}
                flexible
                borderRight
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value =
                    justifyContent === "space-between"
                      ? undefined
                      : "space-between";
                  this.onClick(e, "justifyContent", value);
                }}
                icon={
                  <Icon size={14} name={"flexlayout-justify-spacebetween"} />
                }
                data-tip="Distribute layers according to space between them"
              />
              <Button
                data-cypress-id="vertical-justify-space-around-button"
                active={justifyContent === "space-around" ? true : false}
                flexible
                borderRight
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value =
                    justifyContent === "space-around"
                      ? undefined
                      : "space-around";
                  this.onClick(e, "justifyContent", value);
                }}
                icon={
                  <Icon size={14} name={"flexlayout-justify-spacearound"} />
                }
                data-tip="Distribute layers according to space around them"
                data-for="tooltip-left"
              />
            </ButtonGroup>
          </FormCol>
        </FormRow>
        <FormRow>
          <FormCol first>
            <ButtonGroup uiSize="xs">
              <Button
                data-cypress-id="vertical-align-flex-start-button"
                active={alignItems === "flex-start" ? true : false}
                flexible
                borderRight
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value =
                    alignItems === "flex-start" ? undefined : "flex-start";
                  this.onClick(e, "alignItems", value);
                }}
                icon={
                  <Icon
                    size={14}
                    name={
                      flexDirection === "column"
                        ? "flexlayout-align-start"
                        : "flexlayout-reverse-alignstart"
                    }
                  />
                }
                data-tip="Align layers to left"
                data-for="tooltip-right"
                // tooltip={
                //   <TooltipText position="right">
                //     Align layers to left
                //   </TooltipText>
                // }
              />
              <Button
                data-cypress-id="vertical-align-center-button"
                active={alignItems === "center" ? true : false}
                flexible
                borderRight
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value = alignItems === "center" ? undefined : "center";
                  this.onClick(e, "alignItems", value);
                }}
                icon={
                  <Icon
                    size={14}
                    name={
                      flexDirection === "column"
                        ? "flexlayout-align-center"
                        : "flexlayout-reverse-aligncenter"
                    }
                  />
                }
                data-tip="Align layers to middle"
                // tooltip={<TooltipText>Align layers to middle</TooltipText>}
              />
              <Button
                data-cypress-id="vertical-align-flex-end-button"
                active={alignItems === "flex-end" ? true : false}
                flexible
                borderRight
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value =
                    alignItems === "flex-end" ? undefined : "flex-end";
                  this.onClick(e, "alignItems", value);
                }}
                icon={
                  <Icon
                    size={14}
                    name={
                      flexDirection === "column"
                        ? "flexlayout-align-end"
                        : "flexlayout-reverse-alignend"
                    }
                  />
                }
                data-tip="Align layers to right"
                // tooltip={<TooltipText>Align layers to right</TooltipText>}
              />
              <Button
                data-cypress-id="vertical-align-stretch-button"
                active={alignItems === "stretch" ? true : false}
                flexible
                borderRight
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value = alignItems === "stretch" ? undefined : "stretch";
                  this.onClick(e, "alignItems", value);
                }}
                icon={<Icon size={14} name={"flexlayout-align-stretch"} />}
                data-tip="Distribute layers to stretch vertically"
                // tooltip={
                //   <TooltipText>
                //     Distribute layers to stretch vertically
                //   </TooltipText>
                // }
              />
              <Button
                data-cypress-id="vertical-align-baseline-button"
                active={alignItems === "baseline" ? true : false}
                flexible
                uiBackground={700}
                disabled={isDisabled}
                onClick={(e: any) => {
                  let value =
                    alignItems === "baseline" ? undefined : "baseline";
                  this.onClick(e, "alignItems", value);
                }}
                icon={
                  <Icon
                    size={14}
                    name={
                      flexDirection === "column"
                        ? "flexlayout-align-align-baseline"
                        : "flexlayout-reverse-alignbaseline"
                    }
                  />
                }
                data-tip="Distribute layers to baseline vertically"
                data-for="tooltip-left"
                // tooltip={
                //   <TooltipText position="left">
                //     Distribute layers to baseline vertically
                //   </TooltipText>
                // }
              />
            </ButtonGroup>
          </FormCol>
        </FormRow>
      </>
    );
  }
}

export default ColumnElement;
