// @builderx-core/builderx/src/components-x/demo/Editors/TextPropEditor/Colorpicker.tsx
import * as React from "react";
import { ColorPickerBox, ColorPickerContainer } from "components-x";
// import onClickOutside from "react-onclickoutside";
import GradientEditor from "./GradientEditor";

class Colorpicker extends React.Component<any, any> {
  constructor(props: any) {
    super(props);
    this.state = {
      showPicker: false,
      display: "right"
    };
  }
  handleClickOutside = (evt: any) => {
    this.setState({ showPicker: false });
  };

  handleColorClick = (e: any) => {
    e.stopPropagation();
    if (window.innerWidth - e.clientX < 150) {
      this.setState({ display: "right" });
    } else {
      this.setState({ display: "left" });
    }
    this.setState({
      showPicker: true
    });
  };
  render() {
    return (
      <ColorPickerBox alignRight={this.props.alignRight}>
        <div
          style={{
            padding: 1,
            height: "100%",
            cursor: this.props.disabled ? "not-allowed" : "default"
          }}
        >
          <div
            style={{
              height: "100%",
              background: "black",
              pointerEvents: this.props.disabled ? "none" : "auto"
            }}
            onClick={e => this.handleColorClick(e)}
          />
        </div>
        {!this.state.showPicker ? null : (
          <ColorPickerContainer width={195} position={this.state.display}>
            <div
              style={{
                color: "#fff",
                width: "100%",
                textAlign: "center",
                backgroundColor: "#353940"
              }}
            >
              <GradientEditor color={"#000000"} enableAlpha={true} />
            </div>
            <div className="closePicker" />
          </ColorPickerContainer>
        )}
      </ColorPickerBox>
    );
  }
}

export default Colorpicker;
