// @builderx-core/builderx/src/components-x/demo/Editors/TouchableOpacityEditor/index.tsx
import * as React from "react";
import {
  COLORS,
  PropertyGroup,
  PropertyHeader,
  FormRow,
  FormCol,
  Checkbox,
  CheckboxTick,
  CheckboxLabel
} from "components-x";

export default class TouchableOpacityEditor extends React.Component<any, any> {
  render() {
    let propertyHeaderStyle = {
      paddingTop: 0,
      paddingBottom: 0,
      height: 25,
      borderTopLeftRadius: 2,
      borderTopRightRadius: 2,
      borderBottomLeftRadius: 2,
      borderBottomRightRadius: 2,
      backgroundColor: COLORS[700]
    };

    return (
      <PropertyGroup uiBackground={700} style={{ borderRadius: 2 }}>
        <PropertyHeader uiBackground={700} style={propertyHeaderStyle}>
          <FormRow>
            <FormCol first>
              <Checkbox style={{ color: "#FFF" }}>
                <CheckboxTick
                  noMargin
                  uiColor={50}
                  noBorder={true}
                  checked={false}
                  onChange={() => {
                    //
                  }}
                />
                <CheckboxLabel
                  style={{
                    color: COLORS[300]
                  }}
                >
                  Clickable
                </CheckboxLabel>
              </Checkbox>
            </FormCol>
          </FormRow>
        </PropertyHeader>
      </PropertyGroup>
    );
  }
}
