// @builderx-core/builderx/src/domain-x/lib/services/ThirdPartyComponents/Tab/TabContent.ts
import { ComponentDefinition } from "common-interfaces";

const tabContentSchema: ComponentDefinition = {
  version: "0.0.1",
  template: {
    type: "thirdParty",
    import: {
      layerName: "TabContent",
      packageName: "@builderx/tab-view",
      isDefaultImport: false,
      asName: ""
    },
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    },
    props: {
      title: "Tab1",
      style: {
        flex: 1
      }
    },
    children: [
      {
        type: "view",
        children: [],
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false,
          asName: ""
        },
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "#ff4081"
          }
        }
      }
    ]
  },
  thumbnail: "",
  name: "TabContent",
  controlsProperty: [],
  componentRenderDefinition: `class ThirdPartyFlatList extends React.Component {    constructor() {      super();
      const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 });
      this.state = { dataSource: ds.cloneWithRows([\"row 1\", \"row 2\"]) };    }    render() {
         const props = {        ...this.props,        style:{ ...this.props.style},
         dataSource:this.state.dataSource,          renderRow:rowData => <Text>{rowData}</Text>      };
         return (        <ListView          {...props}        />      );    }  }`
};

export default tabContentSchema;
