// @builderx-core/builderx/src/sketch-import/lib/src/utils/index.ts
import { replace, upperFirst, camelCase } from "lodash";
import { getNameId } from "./NameMapsStore";

export function getValidName(name: string, mapName: string): string {
  let symbolName = camelCase(name);
  symbolName = replace(symbolName, /[^a-zA-Z0-9]/g, "");
  let firstMatch = symbolName.match("[a-zA-z]");
  let firstChar = firstMatch ? firstMatch.toString() : null;
  let firstCharIndex = firstChar ? symbolName.indexOf(firstChar) : -1;
  symbolName =
    firstCharIndex !== -1 ? symbolName.substring(firstCharIndex) : "Screen";
  symbolName = upperFirst(symbolName);
  if (mapName !== "artboard") {
    let serialNo = getNameId(symbolName, mapName);
    if (serialNo !== 0) {
      symbolName += serialNo;
    }
  }
  return symbolName;
}
let deviceList = {
  Android: "Nexus 5",
  "Pixel 2": "Google Pixel 2",
  "Pixel 2 XL": "Google Pixel 2 XL",
  "Galaxy S8": "Samsung Galaxy S8",
  Pixel: "Google Pixel",
  "iPhone 8": "iPhone 8",
  "iPhone 8 Plus": "iPhone 8 Plus",
  "iPhone SE": "iPhone SE",
  "iPhone X": "iPhone X",
  iPad: "The iPad 2017",
  "iPad Pro": "12.9 iPad Pro 2017"
};

export function isValidDeviceToExport(name: string) {
  return name in deviceList;
}
export function getMappedDevice(name: string) {
  return deviceList[name];
}
