// @builderx-core/builderx/src/components-x/demo/Examples/InlineStyleTexts.tsx
import * as React from "react";

import { Text } from "../../lib/core";

export default class InlineStyleTexts extends React.Component {
  constructor(props: any) {
    super(props);
    this.state = {};
  }
  public render() {
    return (
      <div>
        <Text>Normal Text Component</Text>

        <Text>Session</Text>
        <p>
          <br />
          <h4>1. CodeEditor</h4>
          <br />
          <Text style={{ zIndex: 2, padding: 10 }}>
            Generating Code. Please wait...
          </Text>
          <br />
          <br />
          <h4>2. Panels/LeftPanel</h4>
          <br />
          <Text
            uiSize="m"
            style={{
              paddingLeft: 8,
              whiteSpace: "nowrap",
              overflow: "hidden",
              textOverflow: "ellipsis"
            }}
          >
            Layers / FileName
          </Text>
          <br />
          <br />
          <h4>
            3. StartScreen - Tutorials | RecentProjectListItemName | Docs |
            SampleAppListItem
          </h4>
          <br />
          <Text
            inline
            uiSize="m"
            style={{
              width: "100%",
              paddingLeft: 5,
              paddingRight: 5
            }}
          >
            Tutorials
          </Text>
          <Text
            inline
            uiSize="m"
            style={{
              width: "100%",
              paddingLeft: 5,
              paddingRight: 5
            }}
          >
            Docs
          </Text>
          <Text
            inline
            uiSize="m"
            style={{
              width: "100%",
              paddingLeft: 5,
              paddingRight: 5
            }}
          >
            RecentProjectListItem Name
          </Text>
          <Text
            inline
            uiSize="m"
            style={{
              width: "100%",
              paddingLeft: 5,
              paddingRight: 5
            }}
          >
            SampleApp List Item
          </Text>{" "}
          <br />
          <br />
          <h4>5. StartScreen - Recents|Sample Apps</h4>
          <br />
          <Text uiSize="l" style={{ paddingLeft: 3 }}>
            Recents
          </Text>
          <Text uiSize="l" style={{ paddingLeft: 3 }}>
            Sample Apps
          </Text>
          <br />
          <br />
          <br />
          <h4>StageToolBar ProjectName</h4>
          <Text
            inline
            uiColor={300}
            uiSize="m"
            style={{
              paddingLeft: 6,
              fontWeight: 500,
              maxWidth: 250
            }}
          >
            Project-Name
          </Text>
          <br />
          <br />
        </p>
      </div>
    );
  }
}
