// @builderx-core/builderx/src/components-x/demo/Examples/LeftPanel.tsx
import * as React from "react";

import {
  PropertyGroup,
  PropertyHeader,
  Scrollbars,
  Bar,
  BarHeading,
  WrapperDiv,
  Icon,
  BarAction,
  MenuDropdown,
  OptionMenu,
  OptionMenuTitle,
  Button,
  Option,
  Text
} from "../../lib/core";
import LayerPanel from "../Components/LayerPanel";

export default class LeftPanel extends React.Component<any, any> {
  constructor(props: any) {
    super(props);
    this.state = {};
  }
  public render() {
    return (
      <PropertyGroup
        noBorder
        style={{ height: this.props.showCodeEditor ? "100%" : "95%" }}
      >
        <PropertyHeader
          uiSize="m"
          bordered={true}
          style={{
            paddingBottom: 18,
            paddingTop: 30
          }}
          open={true}
        >
          <Bar uiSize="m" uiBackground={800} style={{ width: "100%" }}>
            <BarHeading uiSize="m">
              <WrapperDiv
                style={{
                  display: "flex",
                  alignItems: "center",
                  overflow: "hidden"
                }}
              >
                <Icon name={"icon_layer"} size={14} />
                <Text
                  uiSize="m"
                  style={{
                    paddingLeft: 8,
                    whiteSpace: "nowrap",
                    overflow: "hidden",
                    textOverflow: "ellipsis"
                  }}
                >
                  Layers / Untitled.jsx
                </Text>
              </WrapperDiv>
              <BarAction uiSize="m">
                <MenuDropdown>
                  <Option>
                    <OptionMenu transparent>
                      <Icon
                        name="cog"
                        size={16}
                        onClick={(e: any) => e.stopPropagation()}
                      />
                    </OptionMenu>
                    <MenuDropdown child>
                      <Option onClick={() => alert("sdfsdf")} key="rename file">
                        <OptionMenu>
                          <OptionMenuTitle>Rename</OptionMenuTitle>
                        </OptionMenu>
                      </Option>

                      <Option
                        onClick={() => {
                          alert("sdfsdf");
                        }}
                        key="duplicate artboard"
                      >
                        <OptionMenu>
                          <OptionMenuTitle>Duplicate Artboard</OptionMenuTitle>
                        </OptionMenu>
                      </Option>

                      <Option
                        onClick={() => {
                          alert("sdfdsf");
                        }}
                        key="remove artboard"
                      >
                        <OptionMenu>
                          <OptionMenuTitle>Remove Artboard</OptionMenuTitle>
                        </OptionMenu>
                      </Option>
                    </MenuDropdown>
                  </Option>
                </MenuDropdown>
                <Button
                  transparent={true}
                  icon={
                    <Icon
                      name={this.props.open ? "uparrow" : "downarrow"}
                      size={10}
                    />
                  }
                  uiSize="xs"
                  key="arrow"
                />
              </BarAction>
            </BarHeading>
          </Bar>
        </PropertyHeader>
        <Scrollbars
          autoHide
          style={{
            height: "calc(98% - 62px)",
            display: "flex"
          }}
        >
          <LayerPanel />
        </Scrollbars>
      </PropertyGroup>
    );
  }
}
