// @builderx-core/builderx/src/components-x/demo/Examples/SelectX.tsx
import * as React from "react";
import { SelectX } from "../../lib/core";
import { toLower } from "lodash";

export default class SelectXExample extends React.Component<any, any> {
  constructor(props: any) {
    super(props);
    this.state = {
      selected: {
        caption: "Second",
        id: 2
      },
      options: [
        {
          caption: "First",
          id: 1
        },
        {
          caption: "Second",
          id: 2
        },
        {
          caption: "Third",
          id: 3
        },
        {
          caption: "Fourth",
          id: 4
        }
      ]
    };
  }
  onChange = (e: any) => {
    if (e) {
      this.setState({ selected: e });
    }
  };
  public render() {
    return (
      <div>
        <SelectX
          options={this.state.options}
          value={this.state.selected}
          renderOption={(option: any) => {
            return option.caption;
          }}
          renderValue={(option: any) => option}
          extractValue={(option: any) => option.caption}
          onChange={this.onChange}
          placeholder={"Style"}
          tooltip="Border Style"
          filter={(inputValue: string) =>
            this.state.options.filter((option: any) => {
              return (
                !inputValue ||
                toLower(option.caption).includes(toLower(inputValue))
              );
            })
          }
          onOpenStateChange={(isOpen: boolean) => {
            // console.log("*** isOpen", isOpen)
          }}
          onTextChange={(input: string) => {
            // console.log("*** text change", input);
          }}
          onHoverOption={(option: any) => {
            // console.log("*** onHoverOption", option)
          }}
        />
        {/* <SelectX
          options={this.state.selectedProps.options}
          value={
            typeof this.state.selectedProps.selected === "string"
              ? this.state.selectedProps.selected
              : "solid"
          }
          renderOption={(option: any) => option.caption}
          onChange={this.onChange}
          placeholder={"Style"}
          tooltip="Border Style"
          filter={(inputValue: string) =>
            this.state.selectedProps.options.filter((option: any) => {
              return (
                !inputValue ||
                toLower(option.caption).includes(toLower(inputValue))
              );
            })
          }
        /> */}
      </div>
    );
  }
}
