// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/Footer/footer2.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, TouchableOpacity, Text, StyleSheet } from "react-native";
export default class cupertinoFooter2 extends Component {
  static containerStyle = {
    height: 49,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <TouchableOpacity style={styles.btnWrapper1}>
          <Icon
            style={[
              styles.icon,
              {
                color: this.props.active ? "#007AFF" : "#616161"
              }
            ]}
            type="MaterialCommunityIcons"
            name="av-timer"
          />
          <Text
            style={[
              styles.btn2Caption,
              {
                color: this.props.active ? "#007AFF" : "#9E9E9E"
              }
            ]}
          >
            Dashboard
          </Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.btnWrapper3}>
          <Icon
            type="MaterialCommunityIcons"
            style={styles.icon2}
            name="library-books"
          />
          <Text style={styles.btn3Caption}>Orderbook</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.btnWrapper4}>
          <Icon
            type="MaterialCommunityIcons"
            style={styles.icon3}
            name="poll"
          />
          <Text style={styles.btn4Caption}>Stats</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.btnWrapper2}>
          <Icon
            type="MaterialCommunityIcons"
            style={styles.icon1}
            name="credit-card"
          />
          <Text style={styles.btn2Caption}>Payment</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.btnWrapper5}>
          <Icon style={styles.icon4} name="ios-heart-empty" type="Ionicons" />
          <Text style={styles.btn5Caption}>Favorite</Text>
        </TouchableOpacity>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    width: "100%",
    flexDirection: "row",
    backgroundColor: "rgba(255,255,255,1)",
    justifyContent: "space-between"
  },
  btnWrapper1: {
    flex: 1,
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center"
  },
  icon: {
    backgroundColor: "transparent",
    opacity: 0.8,
    fontSize: 24,
    color: "#616161"
  },
  btn2Caption: {
    backgroundColor: "transparent",
    paddingTop: 4,
    fontSize: 12,
    fontFamily: "Roboto",
    color: "#9E9E9E"
  },
  btnWrapper2: {
    flex: 1,
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center"
  },
  icon1: {
    backgroundColor: "transparent",
    opacity: 0.8,
    fontSize: 24,
    color: "#616161"
  },
  btn2Caption: {
    backgroundColor: "transparent",
    paddingTop: 4,
    fontSize: 12,
    fontFamily: "Roboto",
    color: "#9E9E9E"
  },
  btnWrapper3: {
    flex: 1,
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center"
  },
  icon2: {
    backgroundColor: "transparent",
    opacity: 0.8,
    fontSize: 24,
    color: "#616161"
  },
  btn3Caption: {
    backgroundColor: "transparent",
    paddingTop: 4,
    fontSize: 12,
    fontFamily: "Roboto",
    color: "#9E9E9E"
  },
  btnWrapper4: {
    flex: 1,
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center"
  },
  icon3: {
    backgroundColor: "transparent",
    opacity: 0.8,
    fontSize: 24,
    color: "#616161"
  },
  btn4Caption: {
    backgroundColor: "transparent",
    paddingTop: 4,
    fontSize: 12,
    fontFamily: "Roboto",
    color: "#9E9E9E"
  },
  btnWrapper5: {
    flex: 1,
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center"
  },
  icon4: {
    backgroundColor: "transparent",
    opacity: 0.8,
    fontSize: 24,
    color: "#616161"
  },
  btn5Caption: {
    backgroundColor: "transparent",
    paddingTop: 4,
    fontSize: 12,
    fontFamily: "Roboto",
    color: "#9E9E9E"
  }
});
