// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/Footer/footer1.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, StyleSheet, Text, TouchableOpacity } from "react-native";
export default class CupertinoFooter extends Component {
  static containerStyle = {
    height: 49,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <TouchableOpacity style={styles.btnWrapper1}>
          <Icon
            name="av-timer"
            type="MaterialCommunityIcons"
            style={[
              styles.icon,
              { color: this.props.active ? "#007AFF" : "#616161" }
            ]}
          />
          <Text
            style={[
              styles.btn1Caption,
              { color: this.props.active ? "#007AFF" : "#9E9E9E" }
            ]}
          >
            Dashboard
          </Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.btnWrapper2}>
          <Icon
            type="MaterialCommunityIcons"
            name="library-books"
            style={styles.icon1}
          />
          <Text style={styles.btn2Caption}>Orderbook</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.btnWrapper3}>
          <Icon
            type="MaterialCommunityIcons"
            name="poll"
            style={styles.icon2}
          />
          <Text style={styles.btn3Caption}>Stats</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.btnWrapper4}>
          <Icon
            type="MaterialCommunityIcons"
            name="credit-card"
            style={styles.icon3}
          />
          <Text style={styles.btn4Caption}>Payment</Text>
        </TouchableOpacity>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    backgroundColor: "rgba(255,255,255,1)",
    flexDirection: "row",
    width: "100%"
  },
  btnWrapper1: {
    flex: 1,
    alignItems: "center",
    flexDirection: "column",
    justifyContent: "center"
  },
  icon: {
    backgroundColor: "transparent",
    color: "#616161",
    fontSize: 24,
    opacity: 0.8
  },
  btn1Caption: {
    fontFamily: "Roboto",
    fontSize: 12,
    color: "#9E9E9E",
    backgroundColor: "transparent",
    paddingTop: 4
  },
  btnWrapper2: {
    flex: 1,
    alignItems: "center",
    flexDirection: "column",
    justifyContent: "center"
  },
  icon1: {
    backgroundColor: "transparent",
    color: "#616161",
    fontSize: 24,
    opacity: 0.8
  },
  btn2Caption: {
    fontFamily: "Roboto",
    fontSize: 12,
    color: "#9E9E9E",
    backgroundColor: "transparent",
    paddingTop: 4
  },
  btnWrapper3: {
    flex: 1,
    alignItems: "center",
    flexDirection: "column",
    justifyContent: "center"
  },
  icon2: {
    backgroundColor: "transparent",
    color: "#616161",
    fontSize: 24,
    opacity: 0.8
  },
  btn3Caption: {
    fontFamily: "Roboto",
    fontSize: 12,
    color: "#9E9E9E",
    backgroundColor: "transparent",
    paddingTop: 4
  },
  btnWrapper4: {
    flex: 1,
    alignItems: "center",
    flexDirection: "column",
    justifyContent: "center"
  },
  icon3: {
    backgroundColor: "transparent",
    color: "#616161",
    fontSize: 24,
    opacity: 0.8
  },
  btn4Caption: {
    fontFamily: "Roboto",
    fontSize: 12,
    color: "#9E9E9E",
    backgroundColor: "transparent",
    paddingTop: 4
  }
});
