// @builderx-core/builderx/src/code-engine/flutter/layers/FileCodeGen.ts
import { FlutterContext } from "../types";
import LayerCodeGen from "../LayerCodeGen";

export default class FileCodeGen extends LayerCodeGen {
  context: FlutterContext;
  constructor(context: FlutterContext) {
    super();
    this.context = context;
  }

  generateImports() {
    let importCode = "";
    for (var i = 0; i < this.context.imports.length; i++) {
      importCode += this.context.imports[i] + "\n";
    }
    return importCode;
  }
  // generateStyleNames() {
  //   return this.generateStyleComponentCode(this.context.styles as any);
  // }

  generateCustomPaintClasses() {
    let classes = "";
    for (let i = 0; i < this.context.customPaintClasses.length; i++) {
      classes += this.context.customPaintClasses[i] + "\n";
    }
    return classes;
  }

  generate(json: any): string {
    let className = json.path;
    className = className.substring(
      className.lastIndexOf("/") + 1,
      className.lastIndexOf(".")
    );

    const layersCode = this.context.baseCodeGen.generate(json.layer);

    return `import 'package:flutter/material.dart';
        ${this.generateImports()}

    class ${className} extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    debugShowCheckedModeBanner: false;
    return Scaffold(
      body: ${layersCode}
    );
  }
}

${this.generateCustomPaintClasses()}`;
  }
}
