// @builderx-core/builderx/src/components-x/lib/core/constants/COLORS.ts
export const GREY = {
  0: "#FFFFFF",
  50: "#FBFBFB",
  100: "#f1f2f4",
  200: "#dde0e4",
  300: "#babdc6",
  400: "#6F727D",
  500: "#353940",
  600: "#2e3139",
  700: "#252931",
  800: "#1e2129",
  850: "#1c1e25",
  900: "#171a21",
  A100: "#DBDBDB",
  A200: "#D3D3D3",
  A400: "#BEBEBE",
  A700: "#848383"
};

export const INDIGO = {
  50: "#E8EAF6",
  100: "#C5CAE9",
  200: "#9FA8DA",
  300: "#7986CB",
  400: "#5C6BC0",
  500: "#3F51B5",
  600: "#3949AB",
  700: "#303F9F",
  800: "#283593",
  900: "#1A237E",
  A100: "#8C9EFF",
  A200: "#536DFE",
  A400: "#3D5AFE",
  A700: "#304FFE"
};

export const TEAL = {
  50: "#E0F2F1",
  100: "#B2DFDB",
  200: "#80CBC4",
  300: "#4DB6AC",
  400: "#26A69A",
  500: "#009688",
  600: "#00897B",
  700: "#00796B",
  800: "#00695C",
  900: "#004D40",
  A100: "#A7FFEB",
  A200: "#64FFDA",
  A400: "#1DE9B6",
  A700: "#00BFA5"
};

export const BLUEGREY = {
  50: "#ECEFF1",
  100: "#CFD8DC",
  200: "#B0BEC5",
  300: "#90A4AE",
  400: "#78909C",
  500: "#607D8B",
  600: "#546E7A",
  700: "#455A64",
  800: "#37474F",
  900: "#263238",
  A100: "#B8B8B8",
  A200: "#A0A0A0",
  A400: "#656565",
  A700: "#3F3F3F"
};

// export const GOLD = {
//   50: "#FCF6ED",
//   100: "#F7E5CA",
//   200: "#F1D3A7",
//   300: "#EECB95",
//   400: "#ECC283",
//   500: "#E9B972",
//   600: "#E6B160",
//   700: "#E3A84F",
//   800: "#E19F3D",
//   900: "#DD9527",

//   A100: "#FFE5BB",
//   A200: "#F8CE8E",
//   A400: "#F5BD75",
//   A700: "#F4A133"
// };

export const AQUA = {
  50: "#E5F7F9",
  100: "#BDEAF1",
  200: "#91DCE8",
  300: "#65CEDF",
  400: "#44C4D8",
  500: "#25B9D1",
  600: "#1FB2CC",
  700: "#1AAAC6",
  800: "#15A2C0",
  900: "#0C93B5",

  A100: "#E1F8FF",
  A200: "#AEEDFF",
  A400: "#7BE2FF",
  A700: "#62DCFF"
};

export default GREY;
export const primary = GREY;
export const secondary = AQUA;
export const DANGER = "#f15455";

export const getUtilityColor = (index: number) => {
  return [
    "#2bcbba",
    "#2ecc71",
    "#3498db",
    "#9b59b6",
    "#34495e",
    "#0fb9b1",
    "#20bf6b",
    "#2d98da",
    "#2c3e50",
    "#fed330",
    "#fa8231",
    "#fc5c65",
    "#95a5a6",
    "#f7b731",
    "#eb3b5a",
    "#7f8c8d",
    "#1e272e",
    "#00d8d6",
    "#a55eea",
    "#ef5777"
  ][index % 20];
};
