// @builderx-core/builderx/src/code-engine/react-native/specialLayers/CenterCodeGen.ts
import { Context, ExtraParams } from "../types";
import { addImportToContext } from "../utils";
import LayerCodeGen from "../LayerCodeGen";

export default class CenterCodeGen extends LayerCodeGen {
  context: Context;
  constructor(context: Context) {
    super();
    this.context = context;
  }

  generateCenterProps(template: any) {
    if (template.centerVertical && !template.centerHorizontal) {
      return "vertical";
    } else if (!template.centerVertical && template.centerHorizontal) {
      return "horizontal";
    }
    return "";
  }

  generate(json: any, extraParams: ExtraParams): string {
    addImportToContext(this.context, {
      import: {
        isDefaultImport: false,
        layerName: "Center",
        packageName: "@builderx/utils"
      }
    });

    const propsCode = this.generateCenterProps(json.template);

    delete json.template.centerHorizontal;
    delete json.template.centerVertical;

    return `<Center ${propsCode} >
      ${this.context.baseCodeGen.generate(json, extraParams)}
    </Center>`;
  }
}
