// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Header/header1.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, StyleSheet, Text, TouchableOpacity } from "react-native";
export default class Header1 extends Component {
  static containerStyle = {
    height: 56,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <TouchableOpacity style={styles.leftIconButton}>
          <Icon
            name="menu"
            style={styles.leftIcon}
            type="MaterialCommunityIcons"
          />
        </TouchableOpacity>
        <View style={styles.textWrapper}>
          <Text numberOfLines={1} style={styles.title}>
            Title
          </Text>
        </View>
        <TouchableOpacity style={styles.rightIconButton}>
          <Icon
            name="dots-vertical"
            style={styles.rightIcon}
            type="MaterialCommunityIcons"
          />
        </TouchableOpacity>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    backgroundColor: "#3F51B5",
    flexDirection: "row",
    alignItems: "center",
    padding: 4,
    justifyContent: "space-between",
    shadowColor: "#111",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.2,
    shadowRadius: 1.2,
    elevation: 3
  },
  leftIconButton: {
    padding: 11,
    position: "absolute",
    top: 5,
    left: 5
  },
  rightIconButton: {
    padding: 11,
    position: "absolute",
    top: 5,
    right: 5,
    alignItems: "center"
  },
  title: {
    fontFamily: "Roboto",
    fontSize: 18,
    color: "#FFFFFF",
    backgroundColor: "transparent",
    lineHeight: 18,
    fontWeight: "600"
  },
  textWrapper: {
    position: "absolute",
    left: 72,
    bottom: 19
  },
  leftIcon: {
    backgroundColor: "transparent",
    color: "#FFFFFF",
    fontFamily: "Roboto",
    fontSize: 24
  },
  rightIcon: {
    backgroundColor: "transparent",
    color: "#FFFFFF",
    fontFamily: "Roboto",
    fontSize: 24
  }
});
