// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/Stepper/stepper.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, TouchableOpacity, StyleSheet } from "react-native";
export default class CupertinoStepper1 extends Component {
  static containerStyle = {
    height: 29,
    width: 94
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <TouchableOpacity
          style={[
            styles.leftStepper,
            {
              backgroundColor: this.props.decrement
                ? "rgba(0, 122, 255,0.1)"
                : "#FFFFFF"
            }
          ]}
        >
          <Icon
            style={styles.leftIcon}
            name="minus"
            type="MaterialCommunityIcons"
          />
        </TouchableOpacity>
        <TouchableOpacity
          style={[
            styles.rightStepper,
            {
              backgroundColor: this.props.increment
                ? "rgba(0, 122, 255,0.1)"
                : "#FFFFFF"
            }
          ]}
        >
          <Icon
            style={styles.rightIcon}
            name="plus"
            type="MaterialCommunityIcons"
          />
        </TouchableOpacity>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: "#FFF"
  },
  leftStepper: {
    flex: 1,
    alignItems: "center",
    backgroundColor: "#FFFFFF",
    borderWidth: 1,
    borderColor: "#007AFF",
    borderBottomLeftRadius: 3,
    borderTopLeftRadius: 3,
    borderRightWidth: 0
  },
  leftIcon: {
    fontSize: 30,
    color: "#007AFF"
  },
  rightStepper: {
    flex: 1,
    alignItems: "center",
    backgroundColor: "#FFFFFF",
    borderWidth: 1,
    borderColor: "#007AFF",
    borderBottomRightRadius: 3,
    borderTopRightRadius: 3
  },
  rightIcon: {
    fontSize: 30,
    color: "#007AFF"
  }
});
