// @builderx-core/builderx/src/index.ts
// import "./realtime-io/demo/";
// import "./web-app/demo/";
// import "./domain-x/demo/";
// import "./measurement-lines/demo/";
// import "./smart-flex/demo/";
// import "./dialog-x/demo/";
// import "./session/demo/";
// import "./electron-app/demo/";
// import "./components-x/demo/";
// import "./model/demo";
// import "./sketch-import/demo";
// import "./icon-modal/demo";
// import "./yogax/demo";
// export { ISessionPlugin } from "./common-interfaces";
export { default as BuilderXComponent } from "./web-app/demo/BuilderXComponent";
export { default as loadFonts } from "./components-x/lib/core/helpers/loadFont";
export { generateSVGPath, getPointsArrayFromPath } from "conversions";
import * as Interfaces from "./common-interfaces";
export type ICodeEngine = Interfaces.ICodeEngine;
export type ISessionPlugin = Interfaces.ISessionPlugin;

export { Session, applyPatchInBuilderX } from "session";
import * as bxConstants from "bx-constants";
// import AuthPlugin from "src/session-plugins/lib/plugins/Auth";
export { bxConstants };
export { designerAction } from "@builderx-core/designer";
export { action, addPostCommandActionCallback } from "model";
export { Alert } from "dialog-x";
export { Commands } from "session";
import * as ComponentsX from "components-x";
export { ComponentsX };
export { Utils } from "utils";

// export  ISessionPlugin;
// export const ISessionPlugin1 = ISessionPlugin;
