// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/BasicTab.ts
export default {
  path: "/src/components/MaterialBasicTab.js",
  isDirty: false,
  props: {
    style: {
      width: 375,
      height: 500,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flex: 1,
        backgroundColor: "transparent"
      }
    },
    children: [
      {
        type: "tabView",
        styleName: "tabView1",
        import: {
          layerName: "TabView",
          packageName: "@builderx/tab-view",
          isDefaultImport: false
        },
        props: {
          style: {
            width: 375,
            height: 500
          },
          tabIndicatorStyle: {
            backgroundColor: "#ffffff",
            elevation: 3,
            shadowColor: "#000",
            shadowOffset: {
              width: 0,
              height: 2
            },
            shadowOpacity: 0.2,
            shadowRadius: 1.2
          },
          tabBarStyle: {
            backgroundColor: "#3F51B5"
          },
          activeTabIndex: 0
        },
        children: [
          {
            type: "tabContent",
            styleName: "tabContent1",
            import: {
              layerName: "TabContent",
              packageName: "@builderx/tab-view",
              isDefaultImport: false
            },
            props: {
              title: "Tab",
              style: {
                flex: 1
              }
            },
            children: [
              {
                type: "view",
                styleName: "rect1",
                import: {
                  layerName: "View",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    flex: 1,
                    backgroundColor: "#eeeeee"
                  }
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "tabContent",
            styleName: "tabContent2",
            import: {
              layerName: "TabContent",
              packageName: "@builderx/tab-view",
              isDefaultImport: false
            },
            props: {
              title: "Tab1",
              style: {
                flex: 1
              }
            },
            children: [
              {
                type: "view",
                styleName: "rect2",
                import: {
                  layerName: "View",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    flex: 1,
                    backgroundColor: "#eeeeee"
                  }
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
