// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/BasicFooter.ts
export default {
  path: "/src/components/MaterialBasicFooter.js",
  isDirty: false,
  props: {
    style: {
      height: 56,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "#3f51b5",
        flexDirection: "row",
        alignItems: "center",
        shadowColor: "#111",
        shadowOffset: {
          width: 0,
          height: -2
        },
        shadowOpacity: 0.2,
        shadowRadius: 1.2,
        elevation: 3
      }
    },
    children: [
      {
        type: "touchableOpacity",
        styleName: "btnWrapper1",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            paddingTop: 8,
            paddingBottom: 6,
            paddingHorizontal: 12,
            minWidth: 80,
            maxWidth: 168,
            alignItems: "center",
            flexDirection: "column"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon1",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              name: "television",
              style: {
                backgroundColor: "transparent",
                color: "#FFFFFF",
                fontSize: 24,
                opacity: 0.8
              },
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "btn1Text",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                color: "#FFFFFF",
                opacity: 0.8,
                fontFamily: "Roboto"
              }
            },
            children: [
              {
                type: "literal",
                value: "Movies"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "activebtnWrapper",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            paddingTop: 6,
            paddingBottom: 10,
            paddingHorizontal: 12,
            minWidth: 80,
            maxWidth: 168,
            alignItems: "center",
            flexDirection: "column"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "activeIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              name: "music-note",
              style: {
                backgroundColor: "transparent",
                color: "#FFFFFF",
                fontSize: 24,
                opacity: 1
              },
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "activeText",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                color: "#FFFFFF",
                fontSize: 14,
                opacity: 1,
                fontFamily: "Roboto",
                paddingTop: 4
              }
            },
            children: [
              {
                type: "literal",
                value: "Music"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "btnWrapper2",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            paddingTop: 8,
            paddingBottom: 6,
            paddingHorizontal: 12,
            minWidth: 80,
            maxWidth: 168,
            alignItems: "center",
            flexDirection: "column"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon2",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              name: "book",
              style: {
                backgroundColor: "transparent",
                color: "#FFFFFF",
                fontSize: 24,
                opacity: 0.8
              },
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "btn2Text",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                color: "#FFFFFF",
                opacity: 0.8,
                fontFamily: "Roboto"
              }
            },
            children: [
              {
                type: "literal",
                value: "Books"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "btnWrapper3",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            paddingTop: 8,
            paddingBottom: 6,
            paddingHorizontal: 12,
            minWidth: 80,
            maxWidth: 168,
            alignItems: "center",
            flexDirection: "column"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "icon3",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              name: "calendar-text",
              style: {
                backgroundColor: "transparent",
                color: "#FFFFFF",
                fontSize: 24,
                opacity: 0.8
              },
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "text",
            styleName: "btn3Text",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                color: "#FFFFFF",
                opacity: 0.8,
                fontFamily: "Roboto"
              }
            },
            children: [
              {
                type: "literal",
                value: "Calendar"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
