// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/ButtonTransparentHamburger.ts
export default {
  path: "/src/components/MaterialButtonTransparentHamburger.js",
  isDirty: false,
  props: {
    style: {
      height: 36,
      width: 36,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "TouchableOpacity",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "transparent",
        justifyContent: "center",
        alignItems: "center",
        flexDirection: "row",
        borderRadius: 2
      }
    },
    children: [
      {
        type: "icon",
        styleName: "caption",
        import: {
          layerName: "Icon",
          packageName: "VectorIcons",
          isDefaultImport: true
        },
        props: {
          name: "menu",
          style: {
            color: "#3F51B5",
            fontSize: 24
          },
          type: "MaterialCommunityIcons"
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
