// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/ButtonGrey.ts
export default {
  path: "/src/components/MaterialButtonGrey.js",
  isDirty: false,
  props: {
    style: {
      height: 36,
      width: 100,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "TouchableOpacity",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "#999999",
        justifyContent: "center",
        alignItems: "center",
        flexDirection: "row",
        borderRadius: 2,
        shadowColor: "#000",
        shadowOffset: {
          width: 0,
          height: 1
        },
        shadowOpacity: 0.35,
        shadowRadius: 5,
        elevation: 2,
        minWidth: 88,
        paddingLeft: 16,
        paddingRight: 16
      }
    },
    children: [
      {
        type: "text",
        styleName: "caption",
        import: {
          layerName: "Text",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            color: "#000000",
            fontSize: 14,
            fontFamily: "Roboto",

          }
        },
        children: [
          {
            type: "literal",
            value: "BUTTON"
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
