// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/ButtonWithVioletText.ts
export default {
  path: "/src/components/MaterialButtonWithVioletText.js",
  isDirty: false,
  props: {
    style: {
      height: 36,
      width: 100,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "TouchableOpacity",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "transparent",
        justifyContent: "center",
        alignItems: "center",
        flexDirection: "row",
        minWidth: 88,
        paddingLeft: 16,
        paddingRight: 16
      }
    },
    children: [
      {
        type: "text",
        styleName: "caption",
        import: {
          layerName: "Text",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            color: "#3F51B5",
            fontSize: 14,
            fontFamily: "Roboto",

          }
        },
        children: [
          {
            type: "literal",
            value: "BUTTON"
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
