// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/ButtonShare.ts
export default {
  path: "/src/components/MaterialButtonShare.js",
  isDirty: false,
  props: {
    style: {
      height: 56,
      width: 56,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "TouchableOpacity",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "#3F51B5",
        alignItems: "center",
        justifyContent: "center",
        borderRadius: 28,
        shadowColor: "#111",
        shadowOffset: {
          width: 0,
          height: 2
        },
        shadowOpacity: 0.2,
        shadowRadius: 1.2,
        elevation: 2,
        minWidth: 40,
        minHeight: 40
      }
    },
    children: [
      {
        type: "icon",
        styleName: "icon",
        import: {
          layerName: "Icon",
          packageName: "VectorIcons",
          isDefaultImport: true
        },
        props: {
          name: "share-variant",
          style: {
            color: "#fff",
            fontSize: 24,
            alignSelf: "center"
          },
          type: "MaterialCommunityIcons"
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
