// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/CardWithContent.ts
export default {
  path: "/src/components/MaterialCardWithContent.js",
  isDirty: false,
  props: {
    style: {
      height: 352,
      width: 359,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        borderWidth: 1,
        borderRadius: 2,
        borderColor: "#CCC",
        flexWrap: "nowrap",
        backgroundColor: "#FFF",
        shadowColor: "#000",
        shadowOffset: {
          width: -2,
          height: 2
        },
        shadowOpacity: 0.1,
        shadowRadius: 1.5,
        elevation: 3,
        overflow: "hidden"
      }
    },
    children: [
      {
        type: "view",
        styleName: "cardItem1Style",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flexDirection: "row",
            alignItems: "center",
            padding: 16,
            height: 72
          }
        },
        children: [
          {
            type: "view",
            styleName: "headerStyle",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flex: 1,
                flexDirection: "row",
                alignItems: "center"
              }
            },
            children: [
              {
                type: "image",
                styleName: "leftImage",
                import: {
                  layerName: "Image",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  source: {
                    uri: "src/assets/cardImage.png"
                  },
                  style: {
                    height: 40,
                    width: 40,
                    borderRadius: 20
                  }
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              },
              {
                type: "view",
                styleName: "headerContent",
                import: {
                  layerName: "View",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    paddingLeft: 16,
                    justifyContent: "center"
                  }
                },
                children: [
                  {
                    type: "text",
                    styleName: "textStyle",
                    import: {
                      layerName: "Text",
                      packageName: "native",
                      isDefaultImport: false
                    },
                    props: {
                      style: {
                        fontFamily: "Roboto",
                        fontSize: 16,
                        color: "#000",
                        lineHeight: 20
                      }
                    },
                    children: [
                      {
                        type: "literal",
                        value: "Title"
                      }
                    ],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  },
                  {
                    type: "text",
                    styleName: "noteTextStyle",
                    import: {
                      layerName: "Text",
                      packageName: "native",
                      isDefaultImport: false
                    },
                    props: {
                      style: {
                        fontFamily: "Roboto",
                        fontSize: 14,
                        color: "#000",
                        lineHeight: 16,
                        opacity: 0.5
                      }
                    },
                    children: [
                      {
                        type: "literal",
                        value: "Subhead"
                      }
                    ],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "image",
        styleName: "cardItemImagePlace",
        import: {
          layerName: "Image",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          source: {
            uri: "src/assets/cardImage.png"
          },
          style: {
            backgroundColor: "#ccc",
            flex: 1,
            minHeight: 210
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "body",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            padding: 16
          }
        },
        children: [
          {
            type: "text",
            styleName: "bodyText",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                lineHeight: 20,
                fontSize: 14,
                color: "#424242"
              }
            },
            children: [
              {
                type: "literal",
                value:
                  "BuilderX is a screen design tool which codes React Native for you."
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
