// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/CardWithVerticalButtons.ts
export default {
  path: "/src/components/MaterialCardWithVerticalButtons.js",
  isDirty: false,
  props: {
    style: {
      height: 272,
      width: 359,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        borderWidth: 1,
        borderRadius: 2,
        borderColor: "#CCC",
        flexWrap: "nowrap",
        backgroundColor: "#FFF",
        shadowColor: "#000",
        shadowOffset: {
          width: -2,
          height: 2
        },
        shadowOpacity: 0.1,
        shadowRadius: 1.5,
        elevation: 3,
        flexDirection: "row",
        padding: 16,
        justifyContent: "space-between",
        overflow: "hidden"
      }
    },
    children: [
      {
        type: "image",
        styleName: "cardItemImagePlace",
        import: {
          layerName: "Image",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          source: {
            uri: "src/assets/cardImage.png"
          },
          style: {
            backgroundColor: "#ccc",
            minHeight: 240,
            minWidth: 240,
            flex: 1
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "body",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            padding: 8,
            paddingLeft: 16
          }
        },
        children: [
          {
            type: "touchableOpacity",
            styleName: "topBtn",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 8,
                marginTop: 8,
                marginBottom: 8
              }
            },
            children: [
              {
                type: "icon",
                styleName: "icon1",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  name: "heart",
                  style: {
                    fontSize: 24,
                    color: "#000",
                    opacity: 0.5
                  },
                  type: "MaterialCommunityIcons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "touchableOpacity",
            styleName: "centerBtn",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 8,
                marginTop: 8,
                marginBottom: 8
              }
            },
            children: [
              {
                type: "icon",
                styleName: "icon2",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  name: "book",
                  style: {
                    fontSize: 24,
                    color: "#000",
                    opacity: 0.5
                  },
                  type: "MaterialCommunityIcons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "touchableOpacity",
            styleName: "bottomBtn",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 8,
                marginTop: 8,
                marginBottom: 8
              }
            },
            children: [
              {
                type: "icon",
                styleName: "icon3",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  name: "share",
                  style: {
                    fontSize: 24,
                    color: "#000",
                    opacity: 0.5
                  },
                  type: "MaterialCommunityIcons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
