// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/CardBasic.ts
export default {
  path: "/src/components/MaterialCardBasic.js",
  isDirty: false,
  props: {
    style: {
      height: 282,
      width: 359,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        borderWidth: 1,
        borderRadius: 2,
        borderColor: "#CCC",
        flexWrap: "nowrap",
        backgroundColor: "#FFF",
        shadowColor: "#000",
        shadowOffset: {
          width: -2,
          height: 2
        },
        shadowOpacity: 0.1,
        shadowRadius: 1.5,
        elevation: 3,
        overflow: "hidden"
      }
    },
    children: [
      {
        type: "image",
        styleName: "cardItemImagePlace",
        import: {
          layerName: "Image",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          source: {
            uri: "src/assets/cardImage.png"
          },
          style: {
            backgroundColor: "#ccc",
            minHeight: 210,
            flex: 1
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "body",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            padding: 16
          }
        },
        children: [
          {
            type: "text",
            styleName: "bodyText",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                lineHeight: 20,
                fontSize: 14,
                color: "#424242"
              }
            },
            children: [
              {
                type: "literal",
                value:
                  "BuilderX is a screen design tool which codes React Native for you."
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
