// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/CardWithImageAndTitle.ts
export default {
  path: "/src/components/MaterialCardWithImageAndTitle.js",
  isDirty: false,
  props: {
    style: {
      height: 166,
      width: 359,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        borderWidth: 1,
        borderRadius: 2,
        borderColor: "#CCC",
        flexWrap: "nowrap",
        backgroundColor: "#FFF",
        shadowColor: "#000",
        shadowOffset: {
          width: -2,
          height: 2
        },
        shadowOpacity: 0.1,
        shadowRadius: 1.5,
        elevation: 3,
        overflow: "hidden"
      }
    },
    children: [
      {
        type: "view",
        styleName: "cardBody",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flexDirection: "row",
            justifyContent: "space-between"
          }
        },
        children: [
          {
            type: "view",
            styleName: "bodyContent",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 16,
                paddingTop: 24,
                flex: 1
              }
            },
            children: [
              {
                type: "text",
                styleName: "titleStyle",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    fontFamily: "Roboto",
                    fontSize: 24,
                    color: "#000",
                    paddingBottom: 12
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "Title goes here"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              },
              {
                type: "text",
                styleName: "subtitleStyle",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    fontFamily: "Roboto",
                    fontSize: 14,
                    color: "#000",
                    lineHeight: 16,
                    opacity: 0.5
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "Subtitle here"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "image",
            styleName: "cardItemImagePlace",
            import: {
              layerName: "Image",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              source: {
                uri: "src/assets/cardImage.png"
              },
              style: {
                backgroundColor: "#ccc",
                height: 80,
                width: 80,
                margin: 16
              }
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "actionBody",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            padding: 8,
            flexDirection: "row"
          }
        },
        children: [
          {
            type: "touchableOpacity",
            styleName: "actionButton1",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 8,
                height: 36
              }
            },
            children: [
              {
                type: "text",
                styleName: "actionText1",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    fontSize: 14,
                    color: "#000",
                    opacity: 0.9
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "ACTION 1"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "touchableOpacity",
            styleName: "actionButton2",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 8,
                height: 36
              }
            },
            children: [
              {
                type: "text",
                styleName: "actionText2",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    fontSize: 14,
                    color: "#000",
                    opacity: 0.9
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "ACTION 2"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
