// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/Carousel.ts
export default {
  path: "/src/components/MaterialCarousel.js",
  isDirty: false,
  props: {
    style: {
      height: 400,
      width: 330,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flex: 1
      }
    },
    children: [
      {
        type: "swiper",
        styleName: "carousel1",
        import: {
          layerName: "Swiper",
          packageName: "@builderx/react-native-swiper",
          isDefaultImport: true
        },
        props: {
          style: {
            height: 400,
            width: 330
          },
          index: 0
        },
        children: [
          {
            type: "view",
            styleName: "rect1",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flex: 1,
                backgroundColor: "#5C6BC0",
                justifyContent: "center",
                alignItems: "center"
              }
            },
            children: [
              {
                type: "text",
                styleName: "text1",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    color: "#FFF",
                    fontSize: 18,
                    fontWeight: "700"
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "Page 1"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "view",
            styleName: "rect2",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flex: 1,
                backgroundColor: "#2196F3",
                justifyContent: "center",
                alignItems: "center"
              }
            },
            children: [
              {
                type: "text",
                styleName: "text2",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    color: "#FFF",
                    fontSize: 18,
                    fontWeight: "700"
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "Page 2"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "view",
            styleName: "rect3",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flex: 1,
                backgroundColor: "#009688",
                justifyContent: "center",
                alignItems: "center"
              }
            },
            children: [
              {
                type: "text",
                styleName: "text3",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    color: "#FFF",
                    fontSize: 18,
                    fontWeight: "700"
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "Page 3"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
