// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/Checkbox.ts
export default {
  path: "/src/components/MaterialCheckbox.js",
  isDirty: false,
  props: {
    style: {
      height: 40,
      width: 40,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "TouchableOpacity",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        alignItems: "center",
        justifyContent: "center",
        borderRadius: 20,
        backgroundColor: "transparent"
      }
    },
    children: [
      {
        type: "icon",
        styleName: "checkIcon",
        import: {
          layerName: "Icon",
          packageName: "VectorIcons",
          isDefaultImport: true
        },
        props: {
          name: {
            type: "conditionalExpression",
            test: {
              type: "expression",
              expression: "this.props.checked"
            },
            consequent: {
              type: "literal",
              value: "checkbox-marked"
            },
            alternate: {
              type: "literal",
              value: "checkbox-blank-outline"
            }
          },
          style: {
            color: "#3F51B5",
            fontSize: 28,
            lineHeight: 28
          },
          type: "MaterialCommunityIcons"
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
