// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/ChipBasic.ts
export default {
  path: "/src/components/MaterialChipBasic.js",
  isDirty: false,
  props: {
    style: {
      width: 150,
      height: 32,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        alignItems: "center",
        justifyContent: "center",
        backgroundColor: "rgb(230,230,230)",
        borderRadius: 50,
        paddingLeft: 12,
        paddingRight: 12
      }
    },
    children: [
      {
        type: "text",
        styleName: "chipText",
        import: {
          layerName: "Text",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            fontSize: 13,
            color: "rgba(0,0,0,0.87)"
          }
        },
        children: [
          {
            type: "literal",
            value: "Example Chip"
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
