// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/ChipWithCloseButton.ts
export default {
  path: "/src/components/MaterialChipWithCloseButton.js",
  isDirty: false,
  props: {
    style: {
      height: 32,
      width: 150,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "space-between",
        backgroundColor: "#E0E0E0",
        paddingLeft: 12,
        borderRadius: 50
      }
    },
    children: [
      {
        type: "text",
        styleName: "chipText",
        import: {
          layerName: "Text",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            fontSize: 13,
            color: "rgba(0,0,0,0.87)"
          }
        },
        children: [
          {
            type: "literal",
            value: "Example Chip"
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "icon",
        styleName: "iconStyle",
        import: {
          layerName: "Icon",
          packageName: "VectorIcons",
          isDefaultImport: true
        },
        props: {
          style: {
            color: "#9E9E9E",
            fontSize: 24,
            marginLeft: 4,
            marginRight: 4
          },
          name: "close-circle",
          type: "MaterialCommunityIcons"
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
