// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/ChipWithImage.ts
export default {
  path: "/src/components/MaterialChipWithImage.js",
  isDirty: false,
  props: {
    style: {
      width: 150,
      height: 32,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        alignItems: "center",
        justifyContent: "space-between",
        backgroundColor: "rgb(230,230,230)",
        borderRadius: 50,
        flexDirection: "row"
      }
    },
    children: [
      {
        type: "image",
        styleName: "leftImage",
        import: {
          layerName: "Image",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          source: {
            uri: "src/assets/cardImage.png"
          },
          style: {
            height: 32,
            width: 32,
            backgroundColor: "#CCC",
            borderRadius: 16
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "text",
        styleName: "chipText",
        import: {
          layerName: "Text",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            fontSize: 13,
            color: "rgba(0,0,0,0.87)",
            paddingLeft: 8,
            paddingRight: 12
          }
        },
        children: [
          {
            type: "literal",
            value: "Example Chip"
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
