// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/DisabledTextbox.ts
export default {
  path: "/src/components/MaterialDisabledTextbox.js",
  isDirty: false,
  props: {
    style: {
      height: 43,
      width: 360,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        borderBottomWidth: 1,
        borderColor: "#D9D5DC",
        backgroundColor: "transparent",
        flexDirection: "row",
        alignItems: "center"
      }
    },
    children: [
      {
        type: "textInput",
        styleName: "inputStyle",
        import: {
          layerName: "TextInput",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          placeholder: "Disabled Textbox",
          editable: false,
          style: {
            color: "#000",
            paddingRight: 5,
            fontFamily: "Roboto",
            fontSize: 16,
            alignSelf: "stretch",
            flex: 1,
            lineHeight: 16,
            paddingTop: 16,
            paddingBottom: 8
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "icon",
        styleName: "iconStyle",
        import: {
          layerName: "Icon",
          packageName: "VectorIcons",
          isDefaultImport: true
        },
        props: {
          name: "information-outline",
          style: {
            color: "#384850",
            fontFamily: "Roboto",
            fontSize: 24,
            paddingRight: 8,
            opacity: 0.7
          },
          type: "MaterialCommunityIcons"
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
