// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/FixedLabelTextbox.ts
export default {
  path: "/src/components/MaterialFixedLabelTextbox.js",
  isDirty: false,
  props: {
    style: {
      height: 43,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        borderBottomWidth: 1,
        borderColor: "#D9D5DC",
        backgroundColor: "transparent",
        flexDirection: "row",
        paddingLeft: 16
      }
    },
    children: [
      {
        type: "text",
        styleName: "label",
        import: {
          layerName: "Text",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            fontFamily: "Roboto",
            fontSize: 16,
            lineHeight: 16,
            paddingTop: 16,
            paddingBottom: 8,
            color: "#000",
            opacity: 0.5,
            alignSelf: "flex-start"
          }
        },
        children: [
          {
            type: "literal",
            value: "FixedLabel"
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "textInput",
        styleName: "inputStyle",
        import: {
          layerName: "TextInput",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            color: "#000",
            paddingRight: 5,
            fontFamily: "Roboto",
            fontSize: 16,
            alignSelf: "stretch",
            flex: 1,
            lineHeight: 16,
            paddingTop: 14,
            paddingBottom: 8,
            paddingLeft: 30
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
