// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/Header1.ts
export default {
  path: "/src/components/MaterialHeader1.js",
  isDirty: false,
  props: {
    style: {
      height: 56,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "#3F51B5",
        flexDirection: "row",
        alignItems: "center",
        padding: 4,
        justifyContent: "space-between",
        shadowColor: "#111",
        shadowOffset: {
          width: 0,
          height: 2
        },
        shadowOpacity: 0.2,
        shadowRadius: 1.2,
        elevation: 3
      }
    },
    children: [
      {
        type: "touchableOpacity",
        styleName: "leftIconButton",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            padding: 11,
            position: "absolute",
            top: 5,
            left: 5
          }
        },
        children: [
          {
            type: "icon",
            styleName: "leftIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              name: "menu",
              style: {
                backgroundColor: "transparent",
                color: "#FFFFFF",
                fontSize: 24
              },
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "view",
        styleName: "textWrapper",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            position: "absolute",
            left: 72,
            bottom: 19
          }
        },
        children: [
          {
            type: "text",
            styleName: "title",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              numberOfLines: 1,
              style: {
                fontFamily: "Roboto",
                fontSize: 18,
                color: "#FFFFFF",
                backgroundColor: "transparent",
                lineHeight: 18,
                fontWeight: "600"
              }
            },
            children: [
              {
                type: "literal",
                value: "Title"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "rightIconButton",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            padding: 11,
            position: "absolute",
            top: 5,
            right: 5,
            alignItems: "center"
          }
        },
        children: [
          {
            type: "icon",
            styleName: "rightIcon",
            import: {
              layerName: "Icon",
              packageName: "VectorIcons",
              isDefaultImport: true
            },
            props: {
              name: "dots-vertical",
              style: {
                backgroundColor: "transparent",
                color: "#FFFFFF",
                fontSize: 24
              },
              type: "MaterialCommunityIcons"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
