// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/List14.ts
export default {
  path: "/src/components/MaterialList14.js",
  isDirty: false,
  props: {
    style: {
      height: 500,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flex: 1,
        backgroundColor: "#FFF",
        paddingTop: 8
      }
    },
    children: [
      {
        type: "flatList",
        styleName: "list",
        import: {
          layerName: "FlatList",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "#FFF"
          },
          renderItem: {
            type: "view",
            styleName: "rowBgColor",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 16,
                paddingLeft: 0,
                height: 88,
                marginLeft: 16
              }
            },
            children: [
              {
                type: "text",
                styleName: "rowTitle",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    color: "#212121",
                    fontSize: 16,
                    paddingBottom: 8
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "Two-line Item"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              },
              {
                type: "text",
                styleName: "subText",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    color: "#9E9E9E",
                    fontSize: 13,
                    lineHeight: 16
                  }
                },
                children: [
                  {
                    type: "literal",
                    value:
                      "BuilderX is a screen design tool which codes React Native for you."
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          ListHeaderComponent: {
            type: "view",
            styleName: "listHeaderColor",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 16,
                backgroundColor: "#fff"
              }
            },
            children: [
              {
                type: "text",
                styleName: "listHeaderText",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    color: "#616161"
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "List Header"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          ListFooterComponent: {
            type: "view",
            styleName: "listFooterColor",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 16,
                backgroundColor: "#fff"
              }
            },
            children: [
              {
                type: "text",
                styleName: "listFooterText",
                import: {
                  layerName: "Text",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    color: "#616161"
                  }
                },
                children: [
                  {
                    type: "literal",
                    value: "List Footer"
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
