// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/List3.ts
export default {
  path: "/src/components/MaterialList3.js",
  isDirty: false,
  props: {
    style: {
      height: 500,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flex: 1,
        backgroundColor: "#FFF",
        paddingTop: 8
      }
    },
    children: [
      {
        type: "flatList",
        styleName: "list",
        import: {
          layerName: "FlatList",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "#FFF"
          },
          renderItem: {
            type: "view",
            styleName: "rowBgColor",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flexDirection: "row",
                paddingLeft: 16,
                alignItems: "center",
                height: 48
              }
            },
            children: [
              {
                type: "icon",
                styleName: "iconStyle",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  name: "checkbox-blank",
                  style: {
                    fontSize: 32,
                    color: "#616161"
                  },
                  type: "MaterialCommunityIcons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              },
              {
                type: "view",
                styleName: "contentColor",
                import: {
                  layerName: "View",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    position: "absolute",
                    left: 72,
                    right: 0,
                    flex: 1,
                    paddingRight: 16,
                    height: 48,
                    justifyContent: "center",
                    borderBottomWidth: 0.5,
                    borderColor: "#BDBDBD"
                  }
                },
                children: [
                  {
                    type: "text",
                    styleName: "rowText",
                    import: {
                      layerName: "Text",
                      packageName: "native",
                      isDefaultImport: false
                    },
                    props: {
                      style: {
                        color: "#212121",
                        fontSize: 16
                      }
                    },
                    children: [
                      {
                        type: "literal",
                        value: "Single-line Item"
                      }
                    ],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
