// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/List8.ts
export default {
  path: "/src/components/MaterialList8.js",
  isDirty: false,
  props: {
    style: {
      height: 500,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        flex: 1,
        backgroundColor: "#FFF",
        paddingTop: 8
      }
    },
    children: [
      {
        type: "flatList",
        styleName: "list",
        import: {
          layerName: "FlatList",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "#FFF"
          },
          renderItem: {
            type: "view",
            styleName: "rowBgColor",
            import: {
              layerName: "View",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                flexDirection: "row",
                paddingLeft: 16,
                alignItems: "center",
                height: 72
              }
            },
            children: [
              {
                type: "image",
                styleName: "avatarImageStyle",
                import: {
                  layerName: "Image",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  source: {
                    uri: "src/assets/cardImage.png"
                  },
                  style: {
                    width: 40,
                    height: 40,
                    borderRadius: 20,
                    backgroundColor: "#CCCCCC"
                  }
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              },
              {
                type: "view",
                styleName: "contentColor",
                import: {
                  layerName: "View",
                  packageName: "native",
                  isDefaultImport: false
                },
                props: {
                  style: {
                    position: "absolute",
                    left: 72,
                    right: 0,
                    paddingRight: 16,
                    height: 72,
                    justifyContent: "center",
                    borderBottomWidth: 0.5,
                    borderColor: "#BDBDBD"
                  }
                },
                children: [
                  {
                    type: "text",
                    styleName: "rowPrimaryText",
                    import: {
                      layerName: "Text",
                      packageName: "native",
                      isDefaultImport: false
                    },
                    props: {
                      style: {
                        color: "#212121",
                        fontSize: 16,
                        paddingBottom: 8
                      }
                    },
                    children: [
                      {
                        type: "literal",
                        value: "Two-line Item"
                      }
                    ],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  },
                  {
                    type: "text",
                    styleName: "rowSecondaryText",
                    import: {
                      layerName: "Text",
                      packageName: "native",
                      isDefaultImport: false
                    },
                    props: {
                      style: {
                        color: "#9E9E9E",
                        fontSize: 13
                      },
                      numberOfLines: 1
                    },
                    children: [
                      {
                        type: "literal",
                        value: "Secondary text"
                      }
                    ],
                    locked: false,
                    template: {
                      centerVertical: false,
                      centerHorizontal: false
                    }
                  }
                ],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
