// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/MapView.ts
export default {
  path: "/src/components/MaterialMapView.js",
  isDirty: false,
  props: {
    style: {
      width: 147,
      height: 212,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "white"
      }
    },
    children: [
      {
        type: "mapView",
        styleName: "MapView1",
        import: {
          layerName: "MapView",
          packageName: "expo",
          isDefaultImport: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "rgb(230,230,230)"
          },
          customMapStyle: ""
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
