// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/Picker.ts
export default {
  path: "/src/components/MaterialPicker.js",
  isDirty: false,
  props: {
    style: {
      width: 150,
      height: 174,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "white"
      }
    },
    children: [
      {
        type: "picker",
        styleName: "itemPicker2",
        import: {
          layerName: "Picker",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            width: 150,
            height: 174
          },
          enabled: true,
          itemStyle: {}
        },
        children: [
          {
            type: "pickerItem",
            styleName: "pickerItem",
            import: {
              layerName: "Picker.Item",
              packageName: "",
              isDefaultImport: false
            },
            props: {
              label: "Lorem Ipsum Dolor Sit",
              value: "one"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "pickerItem",
            styleName: "pickerItem",
            import: {
              layerName: "Picker.Item",
              packageName: "",
              isDefaultImport: false
            },
            props: {
              label: "Lorem Ipsum Dolor Sit",
              value: "two"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "pickerItem",
            styleName: "pickerItem",
            import: {
              layerName: "Picker.Item",
              packageName: "",
              isDefaultImport: false
            },
            props: {
              label: "Lorem Ipsum Dolor Sit",
              value: "three"
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
