// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/SearchBar.ts
export default {
  path: "/src/components/MaterialSearchBar.js",
  isDirty: false,
  props: {
    style: {
      height: 56,
      width: 375,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "#3F51B5",
        padding: 4,
        shadowColor: "#111",
        shadowOffset: {
          width: 0,
          height: 2
        },
        shadowOpacity: 0.2,
        shadowRadius: 1.2,
        elevation: 3
      }
    },
    children: [
      {
        type: "view",
        styleName: "rect1",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            flexDirection: "row",
            backgroundColor: "#FFFFFF",
            alignItems: "center",
            position: "absolute",
            top: 4,
            bottom: 4,
            left: 4,
            right: 4,
            borderRadius: 2
          }
        },
        children: [
          {
            type: "touchableOpacity",
            styleName: "leftIconButton",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 11,
                position: "absolute",
                top: 5,
                left: 5
              }
            },
            children: [
              {
                type: "icon",
                styleName: "leftIcon",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  name: "arrow-left",
                  style: {
                    backgroundColor: "transparent",
                    color: "#000",
                    fontSize: 24,
                    opacity: 0.6
                  },
                  type: "MaterialCommunityIcons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "textInput",
            styleName: "inputStyle",
            import: {
              layerName: "TextInput",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              placeholder: "Search",
              style: {
                height: 48,
                color: "#000",
                paddingRight: 5,
                fontFamily: "Roboto",
                fontSize: 16,
                alignSelf: "flex-start",
                width: "70%",
                lineHeight: 16,
                position: "absolute",
                left: 72,
                top: 4
              }
            },
            children: [],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          },
          {
            type: "touchableOpacity",
            styleName: "rightIconButton",
            import: {
              layerName: "TouchableOpacity",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                padding: 11,
                position: "absolute",
                top: 5,
                right: 5,
                alignItems: "center"
              }
            },
            children: [
              {
                type: "icon",
                styleName: "rightIcon",
                import: {
                  layerName: "Icon",
                  packageName: "VectorIcons",
                  isDefaultImport: true
                },
                props: {
                  name: "close",
                  style: {
                    backgroundColor: "transparent",
                    color: "#000",
                    fontSize: 24,
                    opacity: 0.6
                  },
                  type: "MaterialCommunityIcons"
                },
                children: [],
                locked: false,
                template: {
                  centerVertical: false,
                  centerHorizontal: false
                }
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
