// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/Slider.ts
export default {
  path: "/src/components/MaterialSlider.js",
  isDirty: false,
  props: {
    style: {
      width: 150,
      height: 30,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        backgroundColor: "transparent",
        justifyContent: "center"
      }
    },
    children: [
      {
        type: "slider",
        styleName: "slider",
        import: {
          layerName: "Slider",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          value: 50,
          style: {
            position: "absolute",
            height: 30,
            width: 150
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
