// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/Spinner.ts
export default {
  path: "/src/components/MaterialSpinner.js",
  isDirty: false,
  props: {
    style: {
      width: 22,
      height: 22,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        alignItems: "center",
        justifyContent: "center"
      }
    },
    children: [
      {
        type: "activityIndicator",
        styleName: "activityIndicator1",
        import: {
          layerName: "ActivityIndicator",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            width: 22,
            height: 22
          }
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
