// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/Switch.ts
export default {
  path: "/src/components/MaterialSwitch.js",
  isDirty: false,
  props: {
    style: {
      width: 45,
      height: 23,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        alignItems: "center",
        justifyContent: "center"
      }
    },
    children: [
      {
        type: "switch",
        styleName: "switch1",
        import: {
          layerName: "Switch",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            width: 45,
            height: 22
          },
          value: {
            type: "conditionalExpression",
            test: {
              type: "expression",
              expression: "this.props.value"
            },
            consequent: {
              type: "literal",
              value: true
            },
            alternate: {
              type: "literal",
              value: false
            }
          },
          thumbTintColor: {
            type: "conditionalExpression",
            test: {
              type: "expression",
              expression: "this.props.value"
            },
            consequent: {
              type: "literal",
              value: "#3F51B5"
            },
            alternate: {
              type: "literal",
              value: "#FFF"
            }
          },
          onTintColor: "rgba(63,81,181,0.6)",
          tintColor: "#9E9E9E"
        },
        children: [],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
