// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/Toast1.ts
export default {
  path: "/src/components/MaterialToast1.js",
  isDirty: false,
  props: {
    style: {
      width: 322,
      height: 48,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        minWidth: 288,
        justifyContent: "center",
        backgroundColor: "#323232",
        opacity: 1,
        paddingLeft: 24,
        paddingRight: 24,
        borderRadius: 2
      }
    },
    children: [
      {
        type: "text",
        styleName: "text1",
        import: {
          layerName: "Text",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            fontSize: 14,
            fontFamily: "Roboto",
            color: "rgba(255,255,255,1)"
          },
          numberOfLines: 1
        },
        children: [
          {
            type: "literal",
            value: "Toast Message"
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
