// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Material/Symbols/Toast3.ts
export default {
  path: "/src/components/MaterialToast3.js",
  isDirty: false,
  props: {
    style: {
      width: 322,
      height: 80,
      position: "absolute"
    }
  },
  scopeVariables: {
    deviceName: "iPhone X"
  },
  layer: {
    type: "root",
    styleName: "container",
    import: {
      layerName: "View",
      packageName: "native",
      isDefaultImport: false
    },
    props: {
      style: {
        minWidth: 288,
        justifyContent: "space-between",
        backgroundColor: "#323232",
        opacity: 1,
        padding: 24,
        borderRadius: 2,
        flexDirection: "row",
        alignItems: "center"
      }
    },
    children: [
      {
        type: "text",
        styleName: "text1",
        import: {
          layerName: "Text",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            fontSize: 14,
            fontFamily: "Roboto",
            color: "rgba(255,255,255,1)",
            lineHeight: 20
          }
        },
        children: [
          {
            type: "literal",
            value: "Multiline Text added to the toast of BuilderX"
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      },
      {
        type: "touchableOpacity",
        styleName: "button",
        import: {
          layerName: "TouchableOpacity",
          packageName: "native",
          isDefaultImport: false
        },
        props: {
          style: {
            marginLeft: 24
          }
        },
        children: [
          {
            type: "text",
            styleName: "buttonText",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false
            },
            props: {
              style: {
                color: "#4CAF50",
                fontSize: 14,
                fontFamily: "Roboto"
              }
            },
            children: [
              {
                type: "literal",
                value: "UNDO"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        }
      }
    ],
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    }
  }
};
