// @builderx-core/builderx/src/dialog-x/lib/share/ShareLink.tsx
import * as React from "react";
import {
  Window,
  WindowBody,
  Pane,
  FormRow,
  Button,
  InputGroupInput,
  Text,
  WindowFooter,
  FormCol,
  ButtonGroup,
  Title,
  FormGroup,
  RadioButtonBox,
  RadioButton,
  RadioButtonLabel
} from "components-x";
import { Utils } from "utils";
import urljoin from "url-join";

type propType = {
  resolvePromise: (val: boolean, data?: any) => void;
  closeDialog: () => void;
  copyURL: (url: string) => void;
  setConfirmButtonRef: (c: any) => void;
};
// type propType = any;
type stateType = {
  mode: string;
  shareUrl: string;
};
export default class DownloadShareModal extends React.Component<
  propType,
  stateType
> {
  inputRef: any;
  constructor(props: any) {
    super(props);
    this.state = {
      mode: "code",
      shareUrl: urljoin(
        Utils.getEnv("BUILDERX_APP_URL"),
        Utils.getEnv("PROJECT_ID"),
        "?mode=code"
      )
    };
  }

  componentDidMount() {
    if (this.inputRef && this.inputRef.value) {
      this.inputRef.focus();
      this.inputRef.select();
    }
  }

  copyToClipboard = () => {
    if (this.inputRef && this.inputRef.value) {
      this.inputRef.focus();
      this.inputRef.select();
      this.props.copyURL(this.inputRef.value);
    }
  };
  onSubmit = (e: any) => {
    e.preventDefault();
    this.props.resolvePromise(true, this.inputRef.value);
    this.props.closeDialog();
  };
  changeMode = (mode: string) => {
    let shareUrl = urljoin(
      Utils.getEnv("BUILDERX_APP_URL"),
      Utils.getEnv("PROJECT_ID"),
      mode === "design" ? "?mode=design" : "?mode=code"
    );

    this.setState({ mode, shareUrl });
  };
  render() {
    return (
      <Window style={{ display: "block" }}>
        <WindowBody>
          <Pane uiBackground={800} padder={"xs"}>
            <FormGroup noPaddingVertical>
              <FormRow noPadding alignTop>
                <FormCol uiSize={4} contentLeft alignTop style={{ padding: 8 }}>
                  <FormRow>
                    <Title style={{ marginBottom: 0 }} uiSize="s" uiColor={300}>
                      Share with Others
                    </Title>
                  </FormRow>
                  <FormRow style={{ width: 500 }}>
                    <Text>
                      All projects are public by default. If you need a private
                      project contact
                    </Text>
                  </FormRow>
                  <FormRow style={{ width: 500 }}>
                    <Text>
                      <a
                        href={`mailto:builderx-support@geekyants.com?Subject=Query regarding private project
                        `}
                        style={{ color: "white" }}
                      >
                        builderx-support@geekyants.com
                      </a>
                    </Text>
                  </FormRow>
                </FormCol>
              </FormRow>
              <FormRow marginTop={7} paddingBottom={16}>
                <FormCol contentLeft style={{ marginRight: -50 }}>
                  <Text bold uiSize="m">
                    Share:
                  </Text>
                </FormCol>
                <FormCol>
                  <RadioButtonBox>
                    <RadioButton
                      name="design"
                      value="design"
                      checked={this.state.mode === "design"}
                      onChange={(e: any) => this.changeMode(e.target.value)}
                    />
                    <RadioButtonLabel uiSize="l">Design</RadioButtonLabel>
                  </RadioButtonBox>
                </FormCol>
                <FormCol>
                  <RadioButtonBox>
                    <RadioButton
                      name="code"
                      value="code"
                      checked={this.state.mode === "code"}
                      onChange={(e: any) => this.changeMode(e.target.value)}
                    />
                    <RadioButtonLabel uiSize="l">
                      Design + Code
                    </RadioButtonLabel>
                  </RadioButtonBox>
                </FormCol>
              </FormRow>
              <FormRow marginLeft={-17} style={{ width: 500 }}>
                <InputGroupInput
                  ref={(input: any) => {
                    this.inputRef = input;
                  }}
                  value={this.state.shareUrl}
                  uiSize="l"
                  uiBackground={900}
                  style={{ marginLeft: 24 }}
                />

                <Button
                  marginLeft
                  marginRight
                  uiSize="s"
                  paddingLeft={26}
                  paddingRight={26}
                  borderRadius
                  onClick={this.copyToClipboard}
                >
                  <Text color="white" bold uiSize="xs">
                    Copy
                  </Text>
                </Button>
              </FormRow>
            </FormGroup>
          </Pane>
        </WindowBody>
        <WindowFooter>
          <Pane padder="xs" uiBackground={800}>
            <ButtonGroup contentRight={true} style={{ alignItems: "center" }}>
              <Button
                ref={(c: any) => this.props.setConfirmButtonRef(c)}
                marginLeft
                actionBtn
                uiBackground={600}
                caption="Done"
                uiSize="s"
                marginRight
                active
                paddingRight={20}
                paddingLeft={20}
                onClick={(e: any) => this.onSubmit(e)}
              />
            </ButtonGroup>
          </Pane>
        </WindowFooter>
      </Window>
    );
  }
}
