// @builderx-core/builderx/src/components-x/lib/core/GlobalStyle/index.tsx
// base-styles.js
import { createGlobalStyle } from "styled-components";
import reset from "styled-reset";
import COLORS, { DANGER } from "../constants/COLORS";
import fonts from "../fonts";
import spacerCSS from "./spacerCSS";
import borderCSS from "./borderCSS";

const GlobalStyle = createGlobalStyle`
  ${reset}
  ${fonts}
  ${spacerCSS}
  ${borderCSS}

  /* BAD HACK for React Native Web to remove scrollbars. Blame Sanket for this */

  div[class^="rn-WebkitOverflowScrolling-"], div[class*=" rn-WebkitOverflowScrolling-"] {
    overflow: hidden;
  }

  .highlight-filter {
    filter: invert(47%) sepia(100%) saturate(361%) hue-rotate(141deg) brightness(103%) contrast(94%);
  }

  .days-left {
    position: fixed;
    z-index: 9999;
    bottom: 0;
    left: 0;
    padding: 2px;
    border-top-right-radius: 3px;
    background: ${DANGER};
  }

  .export-layer-preview {
    background-image: linear-gradient(45deg, #DDD 25%, transparent 25%),
    linear-gradient(-45deg, #DDD 25%, transparent 25%),
    linear-gradient(45deg, transparent 75%, #DDD 75%),
    linear-gradient(-45deg, transparent 75%, #DDD 75%);
    background-size: 20px 20px;
    background-position: 0 0, 0 10px, 10px -10px, -10px 0px;
  }

  .ml-auto, .mx-auto {
    margin-left: auto !important;
  }

  .mr-auto, .mx-auto {
    margin-right: auto !important;
  }

  body  {
    overflow:hidden;
    background-color: ${COLORS[900]};
    *,
    *:before,
    *:after {
      -moz-box-sizing: border-box;
      -webkit-box-sizing: border-box;
      box-sizing: border-box;
    }
    font-family: Roboto;
    * {
      -webkit-touch-callout:none;
      -webkit-user-select:none;
      -moz-user-select:none;
      -ms-user-select:none;
      user-select:none;
    }
    * {
      -webkit-font-smoothing: antialiased;
      -moz-osx-font-smoothing: grayscale;
    }
    input, textarea, *[contenteditable="true"] /*.contenteditable?*/ {
      -webkit-touch-callout:default;
      -webkit-user-select:text;
      -moz-user-select:text;
      -ms-user-select:text;
      user-select:text;
    }

    ::-webkit-scrollbar {
      width: 1em;
      background: transparent;
    }
    ::-webkit-scrollbar-track {
      -webkit-box-shadow: inset 0 0 6px rgba(0,0,0,0.3);
    }
    ::-webkit-scrollbar-thumb {
      background-color: darkgrey;
      outline: 1px solid slategrey;
    }
  }
`;

export default GlobalStyle;
