// @builderx-core/builderx/src/bx-constants/DeviceStore/deviceDatabase.ts
import { Device } from "./Device";

export const devices: Array<Device> = [
  {
    name: "Samsung Galaxy S10e",
    model: "",
    os: "Android",
    dimension: { height: 162.5, width: 74.8 },
    resolution: { height: 3040, width: 1440 },
    pixelRatio: 4,
    ppi: 521
  },
  {
    name: "Samsung Galaxy S10",
    model: "",
    os: "Android",
    dimension: { height: 162.5, width: 74.8 },
    resolution: { height: 3040, width: 1440 },
    pixelRatio: 4,
    ppi: 521
  },
  {
    name: "Samsung Galaxy S10+",
    model: "",
    os: "Android",
    dimension: { height: 162.5, width: 74.8 },
    resolution: { height: 3040, width: 1440 },
    pixelRatio: 4,
    ppi: 521
  },
  {
    name: "Samsung Note 8",
    model: "",
    os: "Android",
    dimension: { height: 162.5, width: 74.8 },
    resolution: { height: 2960, width: 1440 },
    pixelRatio: 4,
    ppi: 521
  },
  {
    name: "Samsung Galaxy S8",
    model: "",
    os: "Android",
    dimension: { height: 148.9, width: 68.1 },
    resolution: { height: 2960, width: 1440 },
    pixelRatio: 4,
    ppi: 570
  },
  {
    name: "Samsung Galaxy S8 Plus",
    model: "",
    os: "Android",
    dimension: { height: 159.5, width: 73.4 },
    resolution: { height: 2960, width: 1440 },
    pixelRatio: 4,
    ppi: 529
  },
  {
    name: "Samsung Galaxy S7 Edge",
    model: "",
    os: "Android",
    dimension: { height: 150.9, width: 72.6 },
    resolution: { height: 2560, width: 1440 },
    pixelRatio: 4,
    ppi: 534
  },
  {
    name: "Samsung Galaxy S7",
    model: "",
    os: "Android",
    dimension: { height: 142.4, width: 69.6 },
    resolution: { height: 2560, width: 1440 },
    pixelRatio: 4,
    ppi: 577
  },
  {
    name: "Samsung Galaxy S6 Edge",
    model: "",
    os: "Android",
    dimension: { height: 142.1, width: 70.1 },
    resolution: { height: 2560, width: 1440 },
    pixelRatio: 4,
    ppi: 577
  },
  {
    name: "Samsung Galaxy S6",
    model: "",
    os: "Android",
    dimension: { height: 143.4, width: 70.5 },
    resolution: { height: 2560, width: 1440 },
    pixelRatio: 4,
    ppi: 577
  },
  {
    name: "Samsung Galaxy S6 Edge Plus",
    model: "",
    os: "Android",
    dimension: { height: 154.4, width: 75.8 },
    resolution: { height: 2560, width: 1440 },
    pixelRatio: 4,
    ppi: 518
  },
  {
    name: "iPhone 11 Pro Max",
    model: "iPhone 11 Pro Max",
    os: "iOS",
    dimension: { height: 143.6, width: 70.9 },
    resolution: { height: 2688, width: 1242 }, // This is Actual Height & Width
    pixelRatio: 3,
    ppi: 458
  },
  {
    name: "iPhone 11 Pro",
    model: "iPhone 11 Pro",
    os: "iOS",
    dimension: { height: 143.6, width: 70.9 },
    resolution: { height: 2436, width: 1125 }, // This is Actual Height & Width
    // resolution: { height: 2457, width: 1149 }, // Jugaad Height & Width for Frame
    pixelRatio: 3,
    ppi: 458
  },
  {
    name: "iPhone 11",
    model: "iPhone 11",
    os: "iOS",
    dimension: { height: 143.6, width: 70.9 },
    resolution: { height: 1792, width: 828 }, // This is Actual Height & Width
    pixelRatio: 2,
    ppi: 458
  },
  {
    name: "iPhone X",
    model: "iPhone X",
    os: "iOS",
    dimension: { height: 143.6, width: 70.9 },
    resolution: { height: 2436, width: 1125 }, // This is Actual Height & Width
    // resolution: { height: 2457, width: 1149 }, // Jugaad Height & Width for Frame
    pixelRatio: 3,
    ppi: 458
  },
  {
    name: "iPhone XS",
    model: "iPhone Xs",
    os: "iOS",
    dimension: { height: 143.6, width: 70.9 },
    resolution: { height: 2436, width: 1125 }, // This is Actual Height & Width
    pixelRatio: 3,
    ppi: 458
  },
  {
    name: "iPhone XS Max",
    model: "iPhone Xs Max",
    os: "iOS",
    dimension: { height: 157.5, width: 77.4 },
    resolution: { height: 2688, width: 1242 }, // This is Actual Height & Width
    pixelRatio: 3,
    ppi: 458
  },
  {
    name: "iPhone XR",
    model: "iPhone Xr",
    os: "iOS",
    dimension: { height: 150.9, width: 75.7 },
    resolution: { height: 1792, width: 828 }, // This is Actual Height & Width
    pixelRatio: 2,
    ppi: 326
  },
  {
    name: "iPhone 8 Plus",
    model: "",
    os: "iOS",
    dimension: { height: 158.4, width: 78.1 },
    resolution: { height: 1920, width: 1080 },
    pixelRatio: 3,
    ppi: 401
  },
  {
    name: "iPhone 8",
    model: "",
    os: "iOS",
    dimension: { height: 138.4, width: 67.3 },
    resolution: { height: 1334, width: 750 },
    pixelRatio: 2,
    ppi: 326
  },
  {
    name: "iPhone 7 Plus",
    model: "",
    os: "iOS",
    dimension: { height: 158.2, width: 77.9 },
    resolution: { height: 1920, width: 1080 },
    pixelRatio: 3,
    ppi: 401
  },
  {
    name: "iPhone 7",
    model: "",
    os: "iOS",
    dimension: { height: 138.3, width: 67.1 },
    resolution: { height: 1334, width: 750 },
    pixelRatio: 2,
    ppi: 326
  },
  {
    name: "iPhone SE",
    model: "",
    os: "iOS",
    dimension: { height: 123.8, width: 58.6 },
    resolution: { height: 1136, width: 640 },
    pixelRatio: 2,
    ppi: 326
  },
  {
    name: "iPhone 6s Plus",
    model: "",
    os: "iOS",
    dimension: { height: 158.2, width: 77.9 },
    resolution: { height: 1920, width: 1080 },
    pixelRatio: 3,
    ppi: 401
  },
  {
    name: "iPhone 6s",
    model: "",
    os: "iOS",
    dimension: { height: 138.3, width: 67.1 },
    resolution: { height: 1334, width: 750 },
    pixelRatio: 2,
    ppi: 326
  },
  {
    name: "iPhone 6 Plus",
    model: "",
    os: "iOS",
    dimension: { height: 158.1, width: 77.8 },
    resolution: { height: 1920, width: 1080 },
    pixelRatio: 3,
    ppi: 401
  },
  {
    name: "iPhone 6",
    model: "",
    os: "iOS",
    dimension: { height: 138.1, width: 67 },
    resolution: { height: 1334, width: 750 },
    pixelRatio: 2,
    ppi: 326
  },
  {
    name: "iPhone 5s",
    model: "",
    os: "iOS",
    dimension: { height: 123.8, width: 58.6 },
    resolution: { height: 1136, width: 640 },
    pixelRatio: 2,
    ppi: 326
  },
  {
    name: "iPhone 5c",
    model: "",
    os: "iOS",
    dimension: { height: 124.4, width: 59.2 },
    resolution: { height: 1136, width: 640 },
    pixelRatio: 2,
    ppi: 326
  },
  {
    name: "iPhone 5",
    model: "",
    os: "iOS",
    dimension: { height: 123.8, width: 58.6 },
    resolution: { height: 1136, width: 640 },
    pixelRatio: 2,
    ppi: 326
  },
  {
    name: "iPad Pro 2017 12.9",
    model: "",
    os: "iOS",
    dimension: { height: 305.7, width: 220.6 },
    resolution: { height: 2732, width: 2048 },
    pixelRatio: 2,
    ppi: 265
  },
  {
    name: "iPad Pro 2017 10.5",
    model: "",
    os: "iOS",
    dimension: { height: 250.6, width: 174.1 },
    resolution: { height: 2224, width: 1668 },
    pixelRatio: 1,
    ppi: 265
  },
  {
    name: "The iPad 2017",
    model: "",
    os: "iOS",
    dimension: { height: 240, width: 169.5 },
    resolution: { height: 2048, width: 1536 },
    pixelRatio: 2,
    ppi: 264
  },
  {
    name: "iPad Pro 2018 12.9",
    model: "",
    os: "iOS",
    dimension: { height: 280.6, width: 214.9 },
    resolution: { height: 2732, width: 2048 },
    pixelRatio: 2,
    ppi: 264
  },
  {
    name: "iPad Pro 2018 11",
    model: "",
    os: "iOS",
    dimension: { height: 247.6, width: 178.5 },
    resolution: { height: 2388, width: 1668 },
    pixelRatio: 2,
    ppi: 264
  },
  {
    name: "Nexus 5",
    model: "",
    os: "Android",
    dimension: { height: 137.9, width: 69.2 },
    resolution: { height: 1920, width: 1080 },
    pixelRatio: 3,
    ppi: 445
  },
  {
    name: "Nexus 6",
    model: "",
    os: "Android",
    dimension: { height: 159.3, width: 83 },
    resolution: { height: 2560, width: 1440 },
    pixelRatio: 4,
    ppi: 493
  },
  {
    name: "Nexus 5X",
    model: "",
    os: "Android",
    dimension: { height: 147, width: 72.6 },
    resolution: { height: 1920, width: 1080 },
    pixelRatio: 3,
    ppi: 423
  },
  {
    name: "Nexus 6P",
    model: "",
    os: "Android",
    dimension: { height: 159.3, width: 77.8 },
    resolution: { height: 2560, width: 1440 },
    pixelRatio: 4,
    ppi: 518
  },
  {
    name: "Google Pixel",
    model: "Pixel",
    os: "Android",
    dimension: { height: 143.8, width: 69.5 },
    resolution: { height: 1920, width: 1080 },
    pixelRatio: 3,
    ppi: 441
  },
  {
    name: "Google Pixel XL",
    model: "Pixel XL",
    os: "Android",
    dimension: { height: 154.7, width: 75.8 },
    resolution: { height: 2560, width: 1440 },
    pixelRatio: 4,
    ppi: 534
  },
  {
    name: "Google Pixel 2",
    model: "Pixel 2",
    os: "Android",
    dimension: { height: 145.7, width: 69.7 },
    resolution: { height: 1920, width: 1080 },
    pixelRatio: 3,
    ppi: 441
  },
  {
    name: "Google Pixel 2 XL",
    model: "Pixel 2 XL",
    os: "Android",
    dimension: { height: 157.9, width: 76.7 },
    resolution: { height: 2880, width: 1440 },
    pixelRatio: 4,
    ppi: 538
  },
  {
    name: "Google Pixel 3",
    model: "Pixel 3",
    os: "Android",
    dimension: { height: 145.6, width: 68.2 },
    resolution: { height: 2160, width: 1080 },
    pixelRatio: 3,
    ppi: 443
  },
  {
    name: "Google Pixel 3 XL",
    model: "Pixel 3 XL",
    os: "Android",
    dimension: { height: 158, width: 76.7 },
    resolution: { height: 2960, width: 1440 },
    pixelRatio: 4,
    ppi: 523
  },
  {
    name: "OnePlus 5",
    model: "ONEPLUS A5000",
    os: "Android",
    dimension: { height: 154.2, width: 74.1 },
    resolution: { height: 1920, width: 1080 },
    pixelRatio: 3,
    ppi: 401
  },
  {
    name: "OnePlus 5T",
    model: "ONEPLUS A5010",
    os: "Android",
    dimension: { height: 156.1, width: 75 },
    resolution: { height: 2160, width: 1080 },
    pixelRatio: 3,
    ppi: 401
  },
  {
    name: "OnePlus 6",
    model: "ONEPLUS A6000",
    os: "Android",
    dimension: { height: 155.7, width: 75.4 },
    resolution: { height: 2280, width: 1080 },
    pixelRatio: 3,
    ppi: 402
  },
  {
    name: "OnePlus 6T",
    model: "ONEPLUS A6010",
    os: "Android",
    dimension: { height: 157.5, width: 74.8 },
    resolution: { height: 2340, width: 1080 },
    pixelRatio: 3,
    ppi: 402
  },
  {
    name: "OnePlus 7",
    model: "OnePlus 7",
    os: "Android",
    dimension: { height: 157.5, width: 74.8 },
    resolution: { height: 2400, width: 1080 },
    pixelRatio: 3,
    ppi: 402
  },
  {
    name: "OnePlus 7 Pro",
    model: "OnePlus 7 Pro",
    os: "Android",
    dimension: { height: 157.5, width: 74.8 },
    resolution: { height: 3120, width: 1440 },
    pixelRatio: 4,
    ppi: 402
  },
  {
    name: "OnePlus 7T Pro",
    model: "OnePlus 7T Pro",
    os: "Android",
    dimension: { height: 157.5, width: 74.8 },
    resolution: { height: 3120, width: 1440 },
    pixelRatio: 4,
    ppi: 402
  },

  {
    name: "Apple TV",
    model: "Apple TV",
    os: "Web",
    dimension: { height: 1080, width: 1920 },
    resolution: { height: 1080, width: 1920 },
    pixelRatio: 1,
    ppi: 402
  },

  {
    name: "Chromebook Pixel",
    model: "Chromebook Pixel",
    os: "Web",
    dimension: { height: 1700, width: 2560 },
    resolution: { height: 1700, width: 2560 },
    pixelRatio: 1,
    ppi: 239
  },

  {
    name: "Web 1920",
    model: "Web 1920",
    os: "Web",
    dimension: { height: 1080, width: 1920 },
    resolution: { height: 1080, width: 1920 },
    pixelRatio: 1,
    ppi: 239
  },

  {
    name: "Web 1366",
    model: "Web 1366",
    os: "Web",
    dimension: { height: 768, width: 1366 },
    resolution: { height: 768, width: 1366 },
    pixelRatio: 1,
    ppi: 239
  },

  {
    name: "Web 1280",
    model: "Web 1280",
    os: "Web",
    dimension: { height: 800, width: 1280 },
    resolution: { height: 800, width: 1280 },
    pixelRatio: 1,
    ppi: 239
  },

  {
    name: "Laptop HiDPI",
    model: "Laptop HiDPI",
    os: "Web",
    dimension: { height: 900, width: 1400 },
    resolution: { height: 900, width: 1400 },
    pixelRatio: 1,
    ppi: 239
  },

  {
    name: "Laptop MDPI",
    model: "Laptop MDPI",
    os: "Web",
    dimension: { height: 900, width: 1280 },
    resolution: { height: 900, width: 1280 },
    pixelRatio: 1,
    ppi: 239
  },

  {
    name: "Laptop Touch",
    model: "Laptop Touch",
    os: "Web",
    dimension: { height: 950, width: 1280 },
    resolution: { height: 950, width: 1280 },
    pixelRatio: 1,
    ppi: 239
  }
];
