// @builderx-core/builderx/src/rendered-html/demo/App.tsx
import * as React from "react";
import { getRenderedHTML } from "../lib";

var classString = `class ThirdPartyFlatList extends React.Component {
  constructor() {
    super();
    const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 });
    this.state = { dataSource: ds.cloneWithRows(["row 1", "row 2"]) };
  }
  render() {
    const props = {
      ...this.props,
      style:{ ...this.props.style, top: 20 },
        dataSource:this.state.dataSource,
        renderRow:rowData => <Text>{rowData}</Text>
    }
    return (
      <ListView
        {...props}
      />
    );
  }
}`;
// var classString = `
// class ThirdPartyFlatList extends React.Component {
//   constructor() {    super();    }
//   render () {
//     return <View {...this.props} />
//   }
// }
// `;
class App extends React.Component<any, any> {
  constructor(props: any) {
    super(props);

    this.state = { html: "" };

    this.getRenderedHTML();
    // var data = `<View
    //        style={{
    //          backgroundColor: "red",
    //          position: "absolute",
    //          top: 20,
    //          left: 20,
    //          width: 100,
    //          height: 100
    //        }}
    //      />`;

    // transform(
    //   data,
    //   {
    //     plugins: [reactJsxPlugin]
    //   },
    //   (err: any, result: any) => {
    //     console.log(result, err, "hello 123123123");
    //     // ctx.postMessage({ type: "getRenderedHTML", html: result });
    //   }
    // );
  }

  async getRenderedHTML() {
    // var code = `<View
    //         style={{
    //           backgroundColor: "red",
    //           position: "absolute",
    //           top: 20,
    //           left: 20,
    //           width: 100,
    //           height: 100
    //         }}
    //       ><Text>Hello</Text></View>`;
    var code = classString;
    var html = await getRenderedHTML(code, {
      style: { backgroundColor: "red" }
    });

    this.setState({ html: html });

    // console.log(html, "hello html");
  }
  public render() {
    // var html = getRenderedHTMLFromComponent(
    //   <View>
    //     {3 < 2 ? (
    //       <View
    //         style={{
    //           backgroundColor: "red",
    //           position: "absolute",
    //           top: 20,
    //           left: 20,
    //           width: 100,
    //           height: 100
    //         }}
    //       />
    //     ) : (
    //       <View
    //         style={{
    //           backgroundColor: "blue",
    //           position: "absolute",
    //           top: 20,
    //           left: 20,
    //           width: 100,
    //           height: 100
    //         }}
    //       />
    //     )}
    //   </View>
    // );

    // console.log(html, "html");
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">Welcome to React</h1>
        </header>
        <p className="App-intro">
          To get started, edit <code>src/App.tsx</code> and save to reload.
        </p>
        <div dangerouslySetInnerHTML={{ __html: this.state.html }} />
      </div>
    );
  }
}

export default App;
