// @builderx-core/builderx/src/session-plugins/lib/plugins/PluginPlayButton/index.tsx
import * as React from "react";
import { ISessionPlugin } from "common-interfaces";
import { Session } from "session";
import { Button, Icon } from "components-x";
import urljoin from "url-join";

export default class PlayButton implements ISessionPlugin {
  name = "PlayButton";
  session: Session;
  id: string;
  init(session: Session) {
    this.session = session;
  }

  activate() {
    if (this.session.isPluginLoaded("ExternalComponentsProvider")) {
      this.id = this.session.executeFunction("ExternalComponentsProvider.add", {
        placement: "header-panel-right",
        component: () => (
          <Button
            className={"py-2 px-3 highlight-filter"}
            style={{ flexDirection: "column" }}
            transparent
            icon={<Icon size={14} name={"play-record"} />}
            data-cypress-id="play-button"
            onClick={() => {
              window.open(
                urljoin(
                  this.session.envConstants.WEBSITE_URL,
                  "player",
                  "projects",
                  this.session.project.projectId
                ),
                "_blank"
              );
            }}
          />
        )
      });
    }
  }
  deactivate() {
    //
  }
}
