// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Card/cardWithRightButtons.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { StyleSheet, View, TouchableOpacity, Image } from "react-native";
export default class Card8 extends Component {
  static containerStyle = {
    height: 266,
    width: 359
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Image
          source={require("../assets/cardImage.png")}
          style={styles.cardItemImagePlace}
        />
        <View style={styles.buttonGroup}>
          <TouchableOpacity style={styles.leftBtn}>
            <Icon
              name="heart"
              style={styles.icon1}
              type="MaterialCommunityIcons"
            />
          </TouchableOpacity>
          <TouchableOpacity style={styles.centerBtn}>
            <Icon
              name="book"
              style={styles.icon2}
              type="MaterialCommunityIcons"
            />
          </TouchableOpacity>
          <TouchableOpacity style={styles.rightBtn}>
            <Icon
              name="share"
              style={styles.icon3}
              type="MaterialCommunityIcons"
            />
          </TouchableOpacity>
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    borderWidth: 1,
    borderRadius: 2,
    borderColor: "#CCC",
    flexWrap: "nowrap",
    backgroundColor: "#FFF",
    shadowColor: "#000",
    shadowOffset: { width: -2, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 1.5,
    elevation: 3,
    overflow: "hidden"
  },
  cardItemImagePlace: {
    backgroundColor: "#ccc",
    flex: 1,
    minHeight: 210
  },
  buttonGroup: {
    padding: 8,
    flexDirection: "row",
    alignSelf: "flex-end"
  },
  leftBtn: {
    padding: 8,
    marginLeft: 8,
    marginRight: 8
  },
  centerBtn: {
    padding: 8,
    marginLeft: 8,
    marginRight: 8
  },
  rightBtn: {
    padding: 8,
    marginLeft: 8,
    marginRight: 8
  },
  icon1: {
    fontSize: 24,
    color: "#000",
    opacity: 0.5
  },
  icon2: {
    fontSize: 24,
    color: "#000",
    opacity: 0.5
  },
  icon3: {
    fontSize: 24,
    color: "#000",
    opacity: 0.5
  }
});
