// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Card/card6.js
import Icon from "@builderx/icons";
import React, { Component } from "react";
import { Image, StyleSheet, Text, TouchableOpacity, View } from "react-native";
export default class Card6 extends Component {
  static containerStyle = {
    height: 528,
    width: 359,
    defaultHeigh: "auto"
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Image
          source={require("../assets/cardImage.png")}
          style={styles.cardItemImagePlace}
        />
        <View style={styles.bodyContent}>
          <Text style={styles.titleStyle}>Title goes here</Text>
          <Text style={styles.subtitleStyle}>Subtitle here</Text>
        </View>
        <View style={styles.actionBody}>
          <TouchableOpacity style={styles.actionButton1}>
            <Text style={styles.actionText1}>ACTION 1</Text>
          </TouchableOpacity>
          <TouchableOpacity style={styles.actionButton2}>
            <Text style={styles.actionText2}>ACTION 2</Text>
          </TouchableOpacity>
          <TouchableOpacity style={styles.actionButton3}>
            <Icon
              name="chevron-up"
              style={styles.iconStyle}
              type="MaterialCommunityIcons"
            />
          </TouchableOpacity>
        </View>
        <View style={styles.body}>
          <Text style={styles.bodyText}>
            BuilderX is a screen design tool which codes React Native for you
            which design without boundaries, the code is generated
            simultaneously. Save your designed components as symbol and then
            simply add it to your design next time. Live preview works on real
            device. Shout out to the Expo team to make it happen.
          </Text>
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    borderWidth: 1,
    borderRadius: 2,
    borderColor: "#CCC",
    flexWrap: "nowrap",
    backgroundColor: "#FFF",
    shadowColor: "#000",
    shadowOffset: { width: -2, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 1.5,
    elevation: 3,
    overflow: "hidden"
  },
  bodyContent: {
    padding: 16,
    paddingTop: 24,
    justifyContent: "center"
  },
  titleStyle: {
    fontFamily: "Roboto",
    fontSize: 24,
    color: "#000",
    paddingBottom: 12
  },
  subtitleStyle: {
    fontFamily: "Roboto",
    fontSize: 14,
    color: "#000",
    lineHeight: 16,
    opacity: 0.5
  },
  bodyText: {
    lineHeight: 20,
    fontSize: 14,
    color: "#424242"
  },
  body: {
    padding: 16,
    paddingTop: 8
  },
  cardItemImagePlace: {
    backgroundColor: "#ccc",
    minHeight: 210,
    flex: 1
  },
  actionBody: {
    padding: 8,
    flexDirection: "row"
  },
  actionButton1: {
    padding: 8,
    height: 36
  },
  actionButton2: {
    padding: 8,
    height: 36
  },
  actionButton3: {
    padding: 8,
    height: 36,
    position: "absolute",
    right: 8,
    bottom: 12
  },
  iconStyle: {
    fontSize: 24,
    color: "#000",
    opacity: 0.7
  },
  actionText1: {
    fontSize: 14,
    color: "#000",
    opacity: 0.9
  },
  actionText2: {
    fontSize: 14,
    color: "#000",
    opacity: 0.9
  }
});
