// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Card/cardWithVerticalButtons.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { StyleSheet, View, TouchableOpacity, Image } from "react-native";
export default class Card13 extends Component {
  static containerStyle = {
    height: 272,
    width: 359
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Image
          source={require("../assets/cardImage.png")}
          style={styles.cardItemImagePlace}
        />
        <View style={styles.body}>
          <TouchableOpacity style={styles.topBtn}>
            <Icon
              name="heart"
              style={styles.icon1}
              type="MaterialCommunityIcons"
            />
          </TouchableOpacity>
          <TouchableOpacity style={styles.centerBtn}>
            <Icon
              name="book"
              style={styles.icon2}
              type="MaterialCommunityIcons"
            />
          </TouchableOpacity>
          <TouchableOpacity style={styles.bottomBtn}>
            <Icon
              name="share"
              style={styles.icon3}
              type="MaterialCommunityIcons"
            />
          </TouchableOpacity>
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    borderWidth: 1,
    borderRadius: 2,
    borderColor: "#CCC",
    flexWrap: "nowrap",
    backgroundColor: "#FFF",
    shadowColor: "#000",
    shadowOffset: { width: -2, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 1.5,
    elevation: 3,
    flexDirection: "row",
    padding: 16,
    justifyContent: "space-between",
    overflow: "hidden"
  },
  cardItemImagePlace: {
    backgroundColor: "#ccc",
    minHeight: 240,
    minWidth: 240,
    flex: 1
  },
  body: {
    padding: 8,
    paddingLeft: 16
  },
  topBtn: {
    padding: 8,
    marginTop: 8,
    marginBottom: 8
  },
  centerBtn: {
    padding: 8,
    marginTop: 8,
    marginBottom: 8
  },
  bottomBtn: {
    padding: 8,
    marginTop: 8,
    marginBottom: 8
  },
  icon1: {
    fontSize: 24,
    color: "#000",
    opacity: 0.5
  },
  icon2: {
    fontSize: 24,
    color: "#000",
    opacity: 0.5
  },
  icon3: {
    fontSize: 24,
    color: "#000",
    opacity: 0.5
  }
});
