// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Card/cardWithContent.js
import React, { Component } from "react";
import { StyleSheet, View, Text, Image, TouchableOpacity } from "react-native";
export default class Card2 extends Component {
  static containerStyle = {
    height: 352,
    width: 359
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <View style={styles.cardItem1Style}>
          <View style={styles.headerStyle}>
            <Image
              source={require("../assets/cardImage.png")}
              style={styles.leftImage}
            />
            <View style={styles.headerContent}>
              <Text style={styles.textStyle}>Title</Text>
              <Text style={styles.noteTextStyle}>Subhead</Text>
            </View>
          </View>
        </View>
        <Image
          source={require("../assets/cardImage.png")}
          style={styles.cardItemImagePlace}
        />
        <View style={styles.body}>
          <Text style={styles.bodyText}>
            BuilderX is a screen design tool which codes React Native for you.
          </Text>
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    borderWidth: 1,
    borderRadius: 2,
    borderColor: "#CCC",
    flexWrap: "nowrap",
    backgroundColor: "#FFF",
    shadowColor: "#000",
    shadowOffset: { width: -2, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 1.5,
    elevation: 3,
    overflow: "hidden"
  },
  headerContent: {
    paddingLeft: 16,
    justifyContent: "center"
  },
  textStyle: {
    fontFamily: "Roboto",
    fontSize: 16,
    color: "#000",
    lineHeight: 20
  },
  noteTextStyle: {
    fontFamily: "Roboto",
    fontSize: 14,
    color: "#000",
    lineHeight: 16,
    opacity: 0.5
  },
  leftImage: {
    height: 40,
    width: 40,
    borderRadius: 20
  },
  headerStyle: {
    flex: 1,
    flexDirection: "row",
    alignItems: "center"
  },
  cardItem1Style: {
    flexDirection: "row",
    alignItems: "center",
    padding: 16,
    height: 72
  },
  cardItemImagePlace: {
    backgroundColor: "#ccc",
    flex: 1,
    minHeight: 210
  },
  bodyText: {
    lineHeight: 20,
    fontSize: 14,
    color: "#424242"
  },
  body: {
    padding: 16
  }
});
