// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Tabs/basicTab.js
import React, { Component } from "react";
import { TabView, TabContent } from "@builderx/tab-view";
import { View, StyleSheet } from "react-native";
export default class BasicTab extends Component {
  // Only for displaying symbol in BuilderX.
  static containerStyle = {
    width: 375,
    height: 500
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <TabView
          style={styles.tabView1}
          tabIndicatorStyle={styles.tabIndicatorStyle}
          tabBarStyle={styles.tabBarStyle}
          activeTabIndex={0}
        >
          <TabContent title="Tab" style={styles.tabContent1}>
            <View style={styles.rect1} />
          </TabContent>
          <TabContent title="Tab1" style={styles.tabContent2}>
            <View style={styles.rect2} />
          </TabContent>
        </TabView>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "transparent"
  },
  tabView1: {
    width: 375,
    height: 500
  },
  rect1: {
    flex: 1,
    backgroundColor: "#eeeeee"
  },
  rect2: {
    flex: 1,
    backgroundColor: "#eeeeee"
  },
  tabContent1: {
    flex: 1
  },
  tabContent2: {
    flex: 1
  },
  tabBarStyle: {
    backgroundColor: "#3F51B5"
  },
  tabIndicatorStyle: {
    backgroundColor: "#ffffff",
    elevation: 3,
    shadowColor: "#000",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.2,
    shadowRadius: 1.2
  }
});
