// @builderx-core/builderx/cypress/integration/login.spec.js
import deleteShapeSelecter from "./actions/deleteShape";

export default function loginTests(environment = "production") {
  context("Login", () => {
    //var env = process.env.NODE_ENV;
    var env = environment;
    var skipLoginProcedure = false;
    before(function() {
      if (env === "production") {
        cy.visit("https://pre-cloud.builderx.io/login");
        cy.url().then(url => {
          if (url.includes("dashboard")) {
            skipLoginProcedure = true;
          }
        });
      } else {
        cy.request("POST", "http://builderx-api.geekydev.com/authentication", {
          strategy: "local",
          email: "mukulp@geekyants.com",
          password: "goldtree9"
        }).then(response => {
          expect(response.body).to.have.property("accessToken");
          cy.setCookie("access-token", response.body.accessToken);
          if (
            Cypress.env("projUrl") === undefined ||
            Cypress.env("projUrl") === "undefined"
          ) {
            cy.visit("http://localhost:3000/");
          } else {
            cy.visit(Cypress.env("projUrl"));
          }
        });
      }
    });

    beforeEach(function() {
      Cypress.Cookies.preserveOnce("access-token");
    });
    if (env === "production") {
      it("can visit the login page", function() {
        if (skipLoginProcedure) {
          this.skip();
        }
        cy.url().should("include", "https://pre-cloud.builderx.io/login");
      });

      it("displays login with email", function() {
        if (skipLoginProcedure) {
          this.skip();
        }
        cy.get("a")
          .contains("Login with email")
          .should("exist");
      });

      it("opens login form", function() {
        if (skipLoginProcedure) {
          this.skip();
        }
        cy.get("a")
          .contains("Login with email")
          .click("center");
        cy.get(".card form").should("exist");
      });

      it("can add credentials and login", function() {
        if (skipLoginProcedure) {
          this.skip();
        }
        cy.get(".card form input[name='email']")
          .click("center")
          .type("testuser@geekyants.com");
        cy.get(".card form input[name='password']")
          .click("center")
          .type("goldtree9");
        cy.get(".card form button[type='submit']").click("center");
        cy.url().should("include", "https://pre-cloud.builderx.io/dashboard");
      });
      it("can create a new project and open the core app", function() {
        cy.contains(/^New Project$/i).click({ force: true });
        cy.url().should("include", "https://pre-cloud.builderx.io/app/");
        console.log("proj url in login", cy.url());
        // cy.get("div.textX")
        //   .contains("Untitled")
        //   .should("exist");
      });
    } else {
      it("can open the core app", function() {
        cy.exec("echo hello");
        cy.url().should("include", "http://localhost:3000/");
        cy.get(".bx-artboard-Untitled").should("exist");
      });
    }

    it("shows the scratch demo board", function() {
      cy.get(
        ".bx-artboard-Untitled > .view-render > .view-render > .text-here"
      ).should("exist");
      deleteShapeSelecter(
        ".bx-artboard-Untitled > .view-render > .view-render > .text-here"
      );
      cy.get(
        ".bx-artboard-Untitled > .view-render > .view-render > .text-here"
      ).should("not.exist");
    });
  });
}
