// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Chips/chipWithImageAndCloseButton.js
import React, { Component } from "react";
import Icon from "VectorIcons";
import { View, StyleSheet, Text, Image } from "react-native";
export default class Chip4 extends Component {
  static containerStyle = {
    width: 150,
    height: 32
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Image
          source={require("../assets/cardImage.png")}
          style={styles.leftImage}
        />
        <Text style={styles.chipText}>Example Chip</Text>
        <Icon
          style={styles.iconStyle}
          name="close-circle"
          type="MaterialCommunityIcons"
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    alignItems: "center",
    justifyContent: "space-between",
    backgroundColor: "rgb(230,230,230)",
    borderRadius: 50,
    flexDirection: "row"
  },
  text1: {
    fontSize: 13,
    color: "rgba(0,0,0,0.87)",
    paddingLeft: 4,
    paddingRight: 4
  },
  leftImage: {
    height: 32,
    width: 32,
    backgroundColor: "#CCC",
    borderRadius: 16
  },
  iconStyle: {
    color: "#9E9E9E",
    fontSize: 24,
    marginRight: 4
  }
});
