// @builderx-core/builderx/src/shared-components/SharedComponentsJSON/Cupertino/index.ts
import SharedComponentLibrary from "../../SharedComponentLibrary";
import ButtonInfo from "./Symbols/ButtonInfo";
import ButtonWhiteTextColor from "./Symbols/ButtonWhiteTextColor";
import ButtonDelete from "./Symbols/ButtonDelete";
// import ButtonQuestion from "./Symbols/ButtonQuestion";
import ButtonGrey from "./Symbols/ButtonGrey";
import ButtonPurple from "./Symbols/ButtonPurple";
import ButtonSuccess from "./Symbols/ButtonSuccess";
import ButtonWarning from "./Symbols/ButtonWarning";
import ButtonLight from "./Symbols/ButtonLight";
import ButtonDanger from "./Symbols/ButtonDanger";
import ButtonBlueTextColor from "./Symbols/ButtonBlueTextColor";
import ButtonBlackTextColor from "./Symbols/ButtonBlackTextColor";
import Checkbox from "./Symbols/Checkbox";
import Footer1 from "./Symbols/Footer1";
import Footer2 from "./Symbols/Footer2";
import HeaderWithAddButton from "./Symbols/HeaderWithAddButton";
import HeaderWithActionButton from "./Symbols/HeaderWithActionButton";
import HeaderWithSubTitle from "./Symbols/HeaderWithSubTitle";
import HeaderWithBackground from "./Symbols/HeaderWithBackground";
import HeaderWithLargeTitle from "./Symbols/HeaderWithLargeTitle";
// import List1 from "./Symbols/List1";
// import List2 from "./Symbols/List2";
// import List3 from "./Symbols/List3";
// import List4 from "./Symbols/List4";
// import List5 from "./Symbols/List5";
// import List6 from "./Symbols/List6";
// import List7 from "./Symbols/List7";
import Radio from "./Symbols/Radio";
import SearchBar from "./Symbols/SearchBar";
import SearchBarWithCancelButton from "./Symbols/SearchBarWithCancelButton";
import SearchBarBasic from "./Symbols/SearchBarBasic";
import SearchBarWithMic from "./Symbols/SearchBarWithMic";
import SegmentWithTwoTabs from "./Symbols/SegmentWithTwoTabs";
import SegmentWithThreeTabs from "./Symbols/SegmentWithThreeTabs";
import SegmentWithFourTabs from "./Symbols/SegmentWithFourTabs";
import SegmentWithIcons from "./Symbols/SegmentWithIcons";
import Slider from "./Symbols/Slider";
import Stepper from "./Symbols/Stepper";
import Switch from "./Symbols/Switch";
import Toolbar from "./Symbols/Toolbar";

// Cupertino Component Definitions Imports
import ButtonInfoComponentDefinition from "./ComponentDefinitions/Button/ButtonInfo";
import ButtonWhiteTextColorComponentDefinition from "./ComponentDefinitions/Button/ButtonWhiteTextColor";
import ButtonDeleteComponentDefinition from "./ComponentDefinitions/Button/ButtonDelete";
// import ButtonQuestionComponentDefinition from "./ComponentDefinitions/Button/ButtonQuestion";
import ButtonGreyComponentDefinition from "./ComponentDefinitions/Button/ButtonGrey";
import ButtonPurpleComponentDefinition from "./ComponentDefinitions/Button/ButtonPurple";
import ButtonSuccessComponentDefinition from "./ComponentDefinitions/Button/ButtonSuccess";
import ButtonWarningComponentDefinition from "./ComponentDefinitions/Button/ButtonWarning";
import ButtonLightComponentDefinition from "./ComponentDefinitions/Button/ButtonLight";
import ButtonDangerComponentDefinition from "./ComponentDefinitions/Button/ButtonDanger";
import ButtonBlueTextColorComponentDefinition from "./ComponentDefinitions/Button/ButtonBlueTextColor";
import ButtonBlackTextColorComponentDefinition from "./ComponentDefinitions/Button/ButtonBlackTextColor";
// import CheckboxComponentDefinition from "./ComponentDefinitions/Checkbox/Checkbox";
import Footer1ComponentDefinition from "./ComponentDefinitions/Footer/Footer1";
import Footer2ComponentDefinition from "./ComponentDefinitions/Footer/Footer2";
import HeaderWithAddButtonComponentDefinition from "./ComponentDefinitions/Header/HeaderWithAddButton";
import HeaderWithActionButtonComponentDefinition from "./ComponentDefinitions/Header/HeaderWithActionButton";
import HeaderWithSubTitleComponentDefinition from "./ComponentDefinitions/Header/HeaderWithSubTitle";
import HeaderWithBackgroundComponentDefinition from "./ComponentDefinitions/Header/HeaderWithBackground";
import HeaderWithLargeTitleComponentDefinition from "./ComponentDefinitions/Header/HeaderWithLargeTitle";
// import List1ComponentDefinition from "./ComponentDefinitions/List/List1";
// import List2ComponentDefinition from "./ComponentDefinitions/List/List2";
// import List3ComponentDefinition from "./ComponentDefinitions/List/List3";
// import List4ComponentDefinition from "./ComponentDefinitions/List/List4";
// import List5ComponentDefinition from "./ComponentDefinitions/List/List5";
// import List6ComponentDefinition from "./ComponentDefinitions/List/List6";
// import List7ComponentDefinition from "./ComponentDefinitions/List/List7";
import RadioComponentDefinition from "./ComponentDefinitions/Radio/Radio";
import SearchBarComponentDefinition from "./ComponentDefinitions/SearchBar/SearchBar";
import SearchBarWithCancelButtonComponentDefinition from "./ComponentDefinitions/SearchBar/SearchBarWithCancelButton";
import SearchBarBasicComponentDefinition from "./ComponentDefinitions/SearchBar/SearchBarBasic";
import SearchBarWithMicComponentDefinition from "./ComponentDefinitions/SearchBar/SearchBarWithMic";
import SegmentWithTwoTabsComponentDefinition from "./ComponentDefinitions/Segments/SegmentWithTwoTabs";
import SegmentWithThreeTabsComponentDefinition from "./ComponentDefinitions/Segments/SegmentWithThreeTabs";
import SegmentWithFourTabsComponentDefinition from "./ComponentDefinitions/Segments/SegmentWithFourTabs";
import SegmentWithIconsComponentDefinition from "./ComponentDefinitions/Segments/SegmentWithIcons";
import SliderComponentDefinition from "./ComponentDefinitions/Slider/Slider";
import StepperComponentDefinition from "./ComponentDefinitions/Stepper/Stepper";
import SwitchComponentDefinition from "./ComponentDefinitions/Switch/Switch";
import ToolbarComponentDefinition from "./ComponentDefinitions/Toolbar/Toolbar";

import { entries } from "lodash";

export class CupertinoComponentLibrary extends SharedComponentLibrary {
  constructor() {
    super();
    this.name = "Cupertino";
    this.symbols = new Map(
      entries({
        CupertinoButtonInfo: ButtonInfo,
        CupertinoButtonWhiteTextColor: ButtonWhiteTextColor,
        CupertinoButtonDelete: ButtonDelete,
        // CupertinoButtonQuestion: ButtonQuestion,
        CupertinoButtonGrey: ButtonGrey,
        CupertinoButtonPurple: ButtonPurple,
        CupertinoButtonSuccess: ButtonSuccess,
        CupertinoButtonWarning: ButtonWarning,
        CupertinoButtonLight: ButtonLight,
        CupertinoButtonDanger: ButtonDanger,
        CupertinoButtonBlueTextColor: ButtonBlueTextColor,
        CupertinoButtonBlackTextColor: ButtonBlackTextColor,
        CupertinoCheckbox: Checkbox,
        CupertinoFooter1: Footer1,
        CupertinoFooter2: Footer2,
        CupertinoHeaderWithAddButton: HeaderWithAddButton,
        CupertinoHeaderWithActionButton: HeaderWithActionButton,
        CupertinoHeaderWithSubTitle: HeaderWithSubTitle,
        CupertinoHeaderWithBackground: HeaderWithBackground,
        CupertinoHeaderWithLargeTitle: HeaderWithLargeTitle,
        // CupertinoList1: List1,
        // CupertinoList2: List2,
        // CupertinoList3: List3,
        // CupertinoList4: List4,
        // CupertinoList5: List5,
        // CupertinoList6: List6,
        // CupertinoList7: List7,
        CupertinoRadio: Radio,
        CupertinoSearchBar: SearchBar,
        CupertinoSearchBarWithCancelButton: SearchBarWithCancelButton,
        CupertinoSearchBarBasic: SearchBarBasic,
        CupertinoSearchBarWithMic: SearchBarWithMic,
        CupertinoSegmentWithTwoTabs: SegmentWithTwoTabs,
        CupertinoSegmentWithThreeTabs: SegmentWithThreeTabs,
        CupertinoSegmentWithFourTabs: SegmentWithFourTabs,
        CupertinoSegmentWithIcons: SegmentWithIcons,
        CupertinoSlider: Slider,
        CupertinoStepper: Stepper,
        CupertinoSwitch: Switch,
        CupertinoToolbar: Toolbar
      })
    );

    this.templates = {
      Button: [
        ButtonInfoComponentDefinition,
        ButtonWhiteTextColorComponentDefinition,
        ButtonDeleteComponentDefinition,
        // ButtonQuestionComponentDefinition,
        ButtonGreyComponentDefinition,
        ButtonPurpleComponentDefinition,
        ButtonSuccessComponentDefinition,
        ButtonWarningComponentDefinition,
        ButtonLightComponentDefinition,
        ButtonDangerComponentDefinition,
        ButtonBlueTextColorComponentDefinition,
        ButtonBlackTextColorComponentDefinition
      ],
      // Checkbox: [CheckboxComponentDefinition],
      Footer: [Footer1ComponentDefinition, Footer2ComponentDefinition],
      Header: [
        HeaderWithAddButtonComponentDefinition,
        HeaderWithActionButtonComponentDefinition,
        HeaderWithSubTitleComponentDefinition,
        HeaderWithBackgroundComponentDefinition,
        HeaderWithLargeTitleComponentDefinition
      ],
      // List: [
      //   List1ComponentDefinition,
      //   List2ComponentDefinition,
      //   List3ComponentDefinition,
      //   List4ComponentDefinition,
      //   List5ComponentDefinition,
      //   List6ComponentDefinition,
      //   List7ComponentDefinition
      // ],
      Radio: [RadioComponentDefinition],
      SearchBar: [
        SearchBarComponentDefinition,
        SearchBarWithCancelButtonComponentDefinition,
        SearchBarBasicComponentDefinition,
        SearchBarWithMicComponentDefinition
      ],
      Segments: [
        SegmentWithTwoTabsComponentDefinition,
        SegmentWithThreeTabsComponentDefinition,
        SegmentWithFourTabsComponentDefinition,
        SegmentWithIconsComponentDefinition
      ],
      Slider: [SliderComponentDefinition],
      Stepper: [StepperComponentDefinition],
      Switch: [SwitchComponentDefinition],
      Toolbar: [ToolbarComponentDefinition]
    };
  }
  getSize() {
    return this.symbols.size;
  }
}

export const cupertinoComponentLibrary = new CupertinoComponentLibrary();
