// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Toast/toast3.js
import React, { Component } from "react";
import { View, StyleSheet, Text, TouchableOpacity } from "react-native";
export default class Toast3 extends Component {
  static containerStyle = {
    width: 322,
    height: 80
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Text style={styles.text1}>
          Multiline Text added to the toast of BuilderX
        </Text>
        <TouchableOpacity style={styles.button}>
          <Text style={styles.buttonText}>UNDO</Text>
        </TouchableOpacity>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    minWidth: 288,
    justifyContent: "space-between",
    backgroundColor: "#323232",
    opacity: 1,
    padding: 24,
    borderRadius: 2,
    flexDirection: "row",
    alignItems: "center"
  },
  text1: {
    fontSize: 14,
    fontFamily: "Roboto",
    color: "rgba(255,255,255,1)",
    lineHeight: 20
  },
  buttonText: {
    color: "#4CAF50",
    fontSize: 14,
    fontFamily: "Roboto"
  },
  button: {
    marginLeft: 24
  }
});
