// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Toast/toast4.js
import React, { Component } from "react";
import { View, StyleSheet, Text, TouchableOpacity } from "react-native";
export default class Toast4 extends Component {
  static containerStyle = {
    width: 322,
    height: 112
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Text style={styles.text1}>
          Multiline Text added to the toast of BuilderX
        </Text>
        <TouchableOpacity style={styles.button}>
          <Text style={styles.buttonText}>UNDO</Text>
        </TouchableOpacity>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    minWidth: 288,
    backgroundColor: "#323232",
    opacity: 1,
    paddingLeft: 24,
    paddingRight: 24,
    borderRadius: 2,
    flexDirection: "row"
  },
  text1: {
    fontSize: 14,
    lineHeight: 20,
    fontFamily: "Roboto",
    color: "rgba(255,255,255,1)",
    paddingRight: 24,
    paddingTop: 24,
    paddingBottom: 24,
    lineHeight: 20
  },
  buttonText: {
    color: "#4CAF50",
    fontSize: 14,
    fontFamily: "Roboto"
  },
  button: {
    alignSelf: "flex-end",
    marginBottom: 14
  }
});
