// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/Segments/segmentWithThreeTabs.js
import React, { Component } from "react";
import { View, TouchableOpacity, Text, StyleSheet } from "react-native";
export default class CupertinoSegment2 extends Component {
  static containerStyle = {
    height: 56,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <View style={styles.textWrapper}>
          <TouchableOpacity style={styles.segmentTextWrapper1}>
            <Text style={styles.text1}>Puppies</Text>
          </TouchableOpacity>
          <TouchableOpacity style={styles.segmentTextWrapper2}>
            <Text style={styles.text2}>Kittens</Text>
          </TouchableOpacity>
          <TouchableOpacity style={styles.segmentTextWrapper3}>
            <Text style={styles.text3}>Cubs</Text>
          </TouchableOpacity>
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: "#FFF"
  },
  textWrapper: {
    height: 29,
    flex: 1,
    paddingLeft: 30,
    paddingRight: 30,
    flexDirection: "row"
  },
  segmentTextWrapper1: {
    flex: 1,
    alignItems: "center",
    backgroundColor: "#007AFF",
    padding: 6,
    borderWidth: 1,
    borderColor: "#007AFF",
    borderBottomLeftRadius: 5,
    borderTopLeftRadius: 5
  },
  segmentTextWrapper2: {
    flex: 1,
    alignItems: "center",
    backgroundColor: "#FFFFFF",
    padding: 6,
    borderWidth: 1,
    borderColor: "#007AFF",
    borderRightWidth: 0,
    borderLeftWidth: 0
  },
  text1: {
    fontSize: 13,
    fontFamily: "Roboto",
    color: "#FFFFFF"
  },
  segmentTextWrapper3: {
    flex: 1,
    alignItems: "center",
    backgroundColor: "#FFFFFF",
    padding: 6,
    borderWidth: 1,
    borderColor: "#007AFF",
    borderBottomRightRadius: 5,
    borderTopRightRadius: 5
  },
  text2: {
    fontSize: 13,
    fontFamily: "Roboto",
    color: "#007AFF"
  },
  text3: {
    fontSize: 13,
    fontFamily: "Roboto",
    color: "#007AFF"
  }
});
