// @builderx-core/builderx/src/form-builder/lib/FormBuilder/Editors/PickerEditor/index.tsx
import * as React from "react";
import { FormGroup } from "components-x";
import { DomainLayer } from "domain-x";
import TextPropEditor from "./../TextPropEditor";
import InputField from "./../../Components/InputField";
import { IEvent, Font } from "common-interfaces";
import BooleanInput from "../../Components/BooleanInput";
import DropdownInput from "../../Components/DropdownInput";
import { Session } from "session";

type propType = {
  layer: DomainLayer;
  onEvent: (event: IEvent) => void;
  hideAutoToggle?: boolean;
  hideAlignmentEditor?: boolean;
  getFontsList: () => Array<Font>;
  session: Session;
  resolveNodeValue: (node: any) => any;
  robot: any;
};

export default class PickerEditor extends React.Component<propType, any> {
  onEvent = (event: IEvent) => {
    this.props.onEvent(event);
  };
  render() {
    const platform = this.props.layer.platform();
    const { layer, resolveNodeValue } = this.props;
    return platform === "Android" ? (
      <FormGroup noPadding>
        <BooleanInput
          key={"picker-editor-enabled"}
          layer={layer}
          session={this.props.session}
          onEvent={this.onEvent}
          property={{
            propertyName: "enabled",
            type: "boolean",
            title: "Enabled",
            value: {
              defaultValue: true
            },
            category: "props"
          }}
        />
        <InputField
          key={"picker-editor-prompt"}
          session={this.props.session}
          resolveNodeValue={resolveNodeValue}
          property={{
            propertyName: "prompt",
            type: "string",
            title: "Prompt",
            value: {
              defaultValue: ""
            },
            category: "props"
          }}
          layer={layer}
          onEvent={this.onEvent}
        />
        <DropdownInput
          key={"picker-editor-mode"}
          layer={layer}
          session={this.props.session}
          onEvent={this.onEvent}
          property={{
            propertyName: "mode",
            title: "Mode",
            type: "enum",
            value: {
              options: ["dialog", "dropdown"],
              defaultValue: "dialog"
            },
            category: "props"
          }}
          propertyName={"mode"}
          propertyTitle={"Mode"}
          category={"props"}
          options={["dialog", "dropdown"]}
          defaultValue={"dialog"}
        />
      </FormGroup>
    ) : (
      <FormGroup noPadding>
        <TextPropEditor
          {...this.props}
          hideAutoToggle
          onEvent={this.onEvent}
          stylePropName={"itemStyle"}
          robot={this.props.robot}
        />
      </FormGroup>
    );
  }
}
