// @builderx-core/builderx/src/dialog-x/lib/ProgressIndicatorModal.tsx
import * as React from "react";
import {
  Window,
  WindowBody,
  Pane,
  FormGroup,
  FormRow,
  FormCol,
  Title,
  SubTitle,
  Spinner
} from "components-x";
type propType = {
  heading: string;
  text: string;
  value: number;
  resolvePromise: (val: boolean, data?: any) => void;
  loaderOnly?: boolean;
  closeDialog: () => void;
};
export default class ProgressIndicatorModal extends React.Component<
  propType,
  any
> {
  render() {
    const { heading, text } = this.props;
    return (
      <Window>
        <WindowBody>
          <Pane uiBackground={800}>
            <FormGroup style={{ paddingBottom: 0 }}>
              <FormRow center className="p-0">
                {!this.props.loaderOnly && (
                  <FormCol uiSize={4} style={{ padding: 8 }}>
                    <FormRow className="p-3">
                      <Title uiSize="m" uiColor={300}>
                        {heading}
                      </Title>
                    </FormRow>
                    <FormRow className="pl-3 p-2">
                      <SubTitle
                        uiSize="m"
                        className="mb-0"
                        uiColor={400}
                        style={{ lineHeight: "1.4em" }}
                      >
                        {text}
                      </SubTitle>
                    </FormRow>
                  </FormCol>
                )}
                <FormCol style={{ display: "flex" }}>
                  <Spinner uiSize="l" color="white" />
                </FormCol>
              </FormRow>
            </FormGroup>
          </Pane>
        </WindowBody>
      </Window>
    );
  }
}
