// @builderx-core/builderx/src/sketch-import/demo/App.tsx
import * as React from "react";
import SketchImport from "../lib/src/SketchImport";
import IPC from "../demo/IPC";

class App extends React.Component<any, any> {
  sketchImport: any;
  constructor(props: any) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.state = {
      file: File
    };
  }

  init() {
    const ipc = new IPC();
    this.sketchImport = new SketchImport(ipc as any);
  }

  handleChange(selectorFile: any) {
    this.setState({
      file: selectorFile[0]
    });
  }

  handleSubmit = async () => {
    await this.sketchImport.extractSketchFile(this.state.file);
    // TODO: test symbols
    let outputJson = await this.sketchImport.getOutputJSON(
      0.5,
      false,
      "",
      {
        x: 0,
        y: 0
      },
      []
    );

    console.log(outputJson);
  };

  public render() {
    this.init();
    return (
      <div>
        <form
          id="form"
          onSubmit={e => {
            e.preventDefault();
            this.handleSubmit();
          }}
        >
          <input
            type="file"
            onChange={e => this.handleChange(e.target.files)}
          />
          <input type="submit" name="btn-file" value="Upload File" />
        </form>
      </div>
    );
  }
}

export default App;
