// @builderx-core/builderx/src/bx-constants/index.ts
import { StylePropNameType } from "common-interfaces";
import * as ExpoFiles from "./ExpoFiles";
import * as CRAFiles from "./CRAFiles";
export { CRAFiles };

export const ACCEPTED_IMAGE_FORMATS = [
  "image/gif",
  "image/png",
  "image/jpeg",
  "image/jpg"
];
export const MARGIN_PROPERTIES = [
  "margin",
  "marginBottom",
  "marginHorizontal",
  "marginLeft",
  "marginRight",
  "marginTop",
  "marginVertical"
];
export const FLEX_PROPERTIES = [
  "flexDirection",
  "justifyContent",
  "alignItems",
  "alignSelf",
  "flex"
];

export const LAYOUT_PROPERTIES = [
  "bottom",
  "direction",
  "fontSize",
  "fontFamily",
  "height",
  "left",
  "maxHeight",
  "maxWidth",
  "minHeight",
  "minWidth",
  "padding",
  "paddingBottom",
  "paddingHorizontal",
  "paddingLeft",
  "paddingRight",
  "paddingTop",
  "paddingVertical",
  "position",
  "right",
  "rotation",
  "width",
  "top",
  "transform",
  "lineHeight",
  "letterSpacing",
  ...FLEX_PROPERTIES,
  ...MARGIN_PROPERTIES
];

export const ROUNDABLE_LAYOUT_PROPERTIES = [
  "bottom",
  "fontSize",
  "height",
  "left",
  "maxHeight",
  "maxWidth",
  "minHeight",
  "minWidth",
  "padding",
  "paddingBottom",
  "paddingHorizontal",
  "paddingLeft",
  "paddingRight",
  "paddingTop",
  "paddingVertical",
  "right",
  "width",
  "top",
  ...MARGIN_PROPERTIES
];

export const ROUNDABLE_TO_TWO_DECIMALS_LAYOUT_PROPERTIES = ["flex"];

export const LAYOUT_PROPERTIES_RENDER_FILTER = [
  "bottom",
  "height",
  "left",
  "position",
  "right",
  "rotation",
  "width",
  "top",
  "transform",
  "flexDirection",
  "justifyContent",
  "alignItems",
  "alignSelf",
  "flex",
  ...MARGIN_PROPERTIES
];
export const TAG_NAME_STYLE_NAME_MAPPING = {
  View: "rect",
  Text: "text",
  TouchableOpacity: "button",
  Image: "image",
  Icon: "icon",
  TextInput: "textInput",
  ScrollView: "scrollArea",
  Slider: "slider",
  Switch: "switch",
  Picker: "itemPicker",
  ActivityIndicator: "activityIndicator",
  TabView: "tab",
  Swiper: "carousel",
  FlatList: "list",
  Ellipse: "ellipse",
  Group: "group"
};
export const DEBUG_POINTER_EVENTS = false;
export const RENDER_WIREFRAME = false;
export const START_STORYBOOK =
  process.env.NODE_ENV === "production" ? false : false; // Set this true to use testpad
export const ENABLE_EXTENSION_LOG = true;
export { default as DeviceStore } from "./DeviceStore";
export { Device } from "./DeviceStore/Device";
export { devices } from "./DeviceStore/deviceDatabase";
export { IconOptions, overideIconOptions } from "./IconsStore";
export { MapStyles } from "./MapStyle";
export { ExpoFiles };

export const STYLE_PROP_NAMES: StylePropNameType[] = [
  "style",
  "contentContainerStyle",
  "itemStyle",
  "labelStyle",
  "indicatorStyle",
  "tabStyle",
  "tabBarStyle"
];

export const LAYOUT_PROPERTIES_AFFECTING_CHILDREN = [
  "alignContent",
  "alignItems",
  "direction",
  "flexDirection",
  "flexWrap",
  "justifyContent",
  "overflow",
  "padding",
  "paddingBottom",
  "paddingEnd",
  "paddingHorizontal",
  "paddingLeft",
  "paddingRight",
  "paddingStart",
  "paddingTop",
  "paddingVertical"
];

export const FLATLIST_PROPS = [
  "renderItem",
  "ListFooterComponent",
  "ListHeaderComponent",
  "ItemSeparatorComponent"
];

export const CAN_MOVE_LAYER_INSIDE = [
  "view",
  "scrollView",
  "touchableOpacity",
  "root"
];
export const BX_PREVIEW_FILE_PATH = "src/preview/Index.js";
export const fontWeightMap = {
  "100": "Thin",
  "200": "Extra Light",
  "300": "Light",
  "400": "Normal",
  "500": "Medium",
  "600": "Semi Bold",
  "700": "Bold",
  "800": "Extra Bold",
  "900": "Ultra Bold",
  "100italic": "Thin Italic",
  "200italic": "Extra Light Italic",
  "300italic": "Light Italic",
  "400italic": "Normal Italic",
  "500italic": "Medium Italic",
  "600italic": "Semi Bold Italic",
  "700italic": "Bold Italic",
  "800italic": "Extra Bold Italic",
  "900italic": "Ultra Bold Italic",
  regular: "Regular",
  italic: "Italic",
  regularitalic: "Regular Italic",
  bold: "Bold"
};
export const ASSETS_URL =
  typeof window !== "undefined" && window["bxEnv"]
    ? window["bxEnv"].BUILDERX_ASSETS_URL
    : process.env.REACT_APP_ASSETS_URL
    ? String(process.env.REACT_APP_ASSETS_URL)
    : "https://pre-cloud.builderx.io/assets/plugins/component";

export const BUILDERX_JSON_VERSION = "4.0";

export const LAYER_TYPES = [
  "view",
  "root",
  "file",
  "text",
  "image",
  "ellipse",
  "scrollView",
  "symbol",
  "textInput",
  "touchableOpacity",
  "path",
  "icon",
  "activityIndicator",
  "mapView",
  "slider",
  "switch"
];
