// @builderx-core/builderx/src/sketch-import/lib/src/__tests__/layerJSONs.ts
export const viewLayerJSON = {
  type: "view",
  locked: false,
  styleName: "rectangle",
  import: {
    layerName: "View",
    packageName: "native",
    isDefaultImport: false
  },
  template: {
    centerVertical: false,
    centerHorizontal: false
  },
  props: {
    style: {
      position: "absolute",
      top: "11.09%",
      left: "14.67%",
      height: "67.47%",
      width: "67.73%",
      borderWidth: 1,
      borderColor: "rgba(151,151,151,1)",
      backgroundColor: "rgba(216,216,216,1)"
    }
  },
  children: []
};
export const textLayerJSON = {
  type: "text",
  innerText: "Type something can  Lorem Ipsum",
  objectId: "E61299D5-E9F0-4F8F-96BD-1674F1A9453E",
  locked: false,
  styleName: "typeSomethingCanL",
  import: {
    layerName: "Text",
    packageName: "native",
    isDefaultImport: false
  },
  template: {
    centerVertical: false,
    centerHorizontal: false
  },
  props: {
    style: {
      position: "absolute",
      top: "29.24%",
      left: "5.07%",
      height: "10.19%",
      width: "90.13%",
      opacity: 0.6040969122023809,
      backgroundColor: "transparent",
      lineHeight: 20,
      textAlign: "justify",
      color: "rgba(219,37,37,1)",
      fontSize: 25,
      fontFamily: "AmericanTypewriter-Bold",
      letterSpacing: 10
    }
  },
  children: []
};

export const ovalLayerJSON = {
  type: "path",
  locked: false,
  styleName: "oval",
  import: {
    layerName: "Path",
    packageName: "react-native-svg",
    isDefaultImport: false
  },
  template: {
    centerVertical: false,
    centerHorizontal: false
  },
  props: {
    strokeWidth: 1,
    fill: "rgba(241,50,50,1)",
    stroke: "rgba(151,151,151,1)",
    viewBox: "0 0 214 493",
    isClosed: true,
    points: [
      {
        cornerRadius: 0,
        curveMode: 2,
        hasCurveTo: true,
        hasCurveFrom: true,
        curveFrom: {
          x: 165.3183,
          y: 492
        },
        curveTo: {
          x: 47.6817,
          y: 492
        },
        point: {
          x: 106.5,
          y: 492
        }
      },
      {
        cornerRadius: 0,
        curveMode: 2,
        hasCurveTo: true,
        hasCurveFrom: true,
        curveFrom: {
          x: 213,
          y: 110.138
        },
        curveTo: {
          x: 213,
          y: 381.862
        },
        point: {
          x: 213,
          y: 246
        }
      },
      {
        cornerRadius: 0,
        curveMode: 2,
        hasCurveTo: true,
        hasCurveFrom: true,
        curveFrom: {
          x: 47.6817,
          y: 0
        },
        curveTo: {
          x: 165.3183,
          y: 0
        },
        point: {
          x: 106.5,
          y: 0
        }
      },
      {
        cornerRadius: 0,
        curveMode: 2,
        hasCurveTo: true,
        hasCurveFrom: true,
        curveFrom: {
          x: 0,
          y: 381.862
        },
        curveTo: {
          x: 0,
          y: 110.138
        },
        point: {
          x: 0,
          y: 246
        }
      }
    ],
    style: {
      position: "absolute",
      height: "74.06%",
      width: "57.33%",
      top: "8.62%",
      left: "21.47%",
      backgroundColor: "transparent",
      borderColor: "transparent"
    }
  },
  children: []
};
