// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Carousel/carousel.js
import React, { Component } from "react";
import Swiper from "@builderx/react-native-swiper";
import { View, StyleSheet, Text } from "react-native";
export default class Carousel extends Component {
  static containerStyle = {
    height: 400,
    width: 330
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <Swiper style={styles.carousel1} index={0}>
          <View style={styles.rect1}>
            <Text style={styles.text1}>Page 1</Text>
          </View>
          <View style={styles.rect2}>
            <Text style={styles.text2}>Page 2</Text>
          </View>
          <View style={styles.rect3}>
            <Text style={styles.text3}>Page 3</Text>
          </View>
        </Swiper>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1
  },
  carousel1: {
    height: 400,
    width: 330
  },
  rect1: {
    flex: 1,
    backgroundColor: "#5C6BC0",
    justifyContent: "center",
    alignItems: "center"
  },
  text1: {
    color: "#FFF",
    fontSize: 18,
    fontWeight: "700"
  },
  text2: {
    color: "#FFF",
    fontSize: 18,
    fontWeight: "700"
  },
  text3: {
    color: "#FFF",
    fontSize: 18,
    fontWeight: "700"
  },
  rect2: {
    flex: 1,
    backgroundColor: "#2196F3",
    justifyContent: "center",
    alignItems: "center"
  },
  rect3: {
    flex: 1,
    backgroundColor: "#009688",
    justifyContent: "center",
    alignItems: "center"
  }
});
