// @builderx-core/builderx/src/domain-x/lib/services/ThirdPartyComponents/Picker/index.ts
import { ComponentDefinition } from "common-interfaces";
const pickerSchema: ComponentDefinition = {
  version: "0.0.1",
  template: {
    type: "thirdParty",
    import: {
      layerName: "Picker",
      packageName: "react-native",
      isDefaultImport: false,
      asName: ""
    },
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    },
    props: {
      style: {
        position: "absolute",
        top: 50,
        left: 50,
        width: 50,
        height: 50
      },
      itemStyle: {},
      enabled: true
    },
    children: [
      {
        type: "thirdParty",
        import: {
          layerName: "Picker.Item"
        } as any,
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        },
        props: {
          label: "Lorem Ipsum Dolor Sit",
          value: "one"
        },
        children: []
      },
      {
        type: "thirdParty",
        import: {
          layerName: "Picker.Item"
        } as any,
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        },
        props: {
          label: "Lorem Ipsum Dolor Sit",
          value: "two"
        },
        children: []
      },
      {
        type: "thirdParty",
        import: {
          layerName: "Picker.Item"
        } as any,
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        },
        props: {
          label: "Lorem Ipsum Dolor Sit",
          value: "three"
        },
        children: []
      }
    ]
  },
  thumbnail: "",
  name: "Picker",
  controlsProperty: [
    {
      type: "AlignmentEditor",
      title: "Alignment Editor",
      value: {},
      showForMultiple: true
    },
    {
      type: "OrientationEditor",
      title: "Orientation Editor",
      value: {},
      showForMultiple: true
    },
    {
      type: "ResizingTypeEditor",
      title: "ResizingType Editor",
      value: {}
    },
    {
      type: "group",
      title: "Styling",
      value: {},
      properties: [
        {
          propertyName: "backgroundColor",
          type: "color",
          title: "Fill",
          value: {
            defaultValue: "#fff"
          },
          category: "style"
        }
      ]
    },
    {
      type: "group",
      title: "Picker",
      value: {},
      properties: [
        {
          type: "PickerEditor",
          title: "Picker Editor",
          value: {}
        }
      ]
    }
  ],
  componentRenderDefinition: `class ThirdPartyFlatList extends React.Component {
      constructor() {      super();      const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 });
      this.state = { dataSource: ds.cloneWithRows([\"row 1\", \"row 2\"]) };    }    render() {      const props = {
        this.props,        style:{ ...this.props.style},          dataSource:this.state.dataSource,
        renderRow:rowData => <Text>{rowData}</Text>      };      return (        <ListView          {...props}
                  />      );    }  }`
};

export default pickerSchema;
