// @builderx-core/builderx/src/domain-x/lib/services/ThirdPartyComponents/Picker/PickerItem.ts
import { ComponentDefinition } from "common-interfaces";
const pickerItemSchema: ComponentDefinition = {
  version: "0.0.1",
  template: {
    type: "thirdParty",
    import: {
      layerName: "Picker.Item"
    } as any,
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    },
    props: {
      label: "Lorem Ipsum Dolor Sit",
      value: "one"
    },
    children: []
  },
  thumbnail: "",
  name: "Picker.Item",
  controlsProperty: [],
  componentRenderDefinition: `class ThirdPartyFlatList extends React.Component {    constructor() {      super();
      const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 });      this.state = {
        dataSource: ds.cloneWithRows([\"row 1\", \"row 2\"]) };    }    render() {      const props = {
          ...this.props,        style:{ ...this.props.style},          dataSource:this.state.dataSource,
          renderRow:rowData => <Text>{rowData}</Text>      };      return (        <ListView          {...props}
               />      );    }  }`
};

export default pickerItemSchema;
