// @builderx-core/builderx/src/session-plugins/lib/plugins/ExternalComponentsProvider/index.tsx
import { ISessionPlugin } from "common-interfaces";
import { Session } from "session";
import { action } from "model";
export default class ExternalComponentsProvider implements ISessionPlugin {
  name = "ExternalComponentsProvider";
  session: Session;
  externalComponents: Array<any>;
  init(session: Session) {
    this.session = session;
    this.externalComponents = [];
  }
  activate() {
    this.session.addFunction(
      `${this.name}.add`,
      (session: Session, componentData: any, visible: boolean = true) => {
        this.externalComponents.push({
          ...componentData,
          visible
        });
        action(() => {
          this.session.emit("externalComponents");
        });
        return this.externalComponents.length - 1;
      }
    );
    this.session.addFunction(`${this.name}.get`, () => this.externalComponents);
    this.session.addFunction(
      `${this.name}.remove`,
      (session: Session, id: number) => {
        this.externalComponents.splice(id, 1);
        action(() => {
          this.session.emit("externalComponents");
        });
      }
    );
    this.session.addFunction(
      `${this.name}.show`,
      (session: Session, id: any) => {
        if (id !== -1) {
          this.externalComponents[id]["visible"] = true;
          action(() => {
            this.session.emit("externalComponents");
          });
        }
      }
    );
    this.session.addFunction(
      `${this.name}.hide`,
      (session: Session, id: any) => {
        if (id !== -1) {
          this.externalComponents[id]["visible"] = false;
          action(() => {
            this.session.emit("externalComponents");
          });
        }
      }
    );
  }
  deactivate() {
    //
  }
}

// DOC: List of placement ids
// [
// "header-panel-right",
// "header-panel-left",
// "header-panel-center:before",
// "header-panel-center:after"
// "user-account-panel",
// "left-panel-expandable",
// "layer-panel",
// "header-panel-report-bug:before",
// "header-panel-user-acount:before",
// "header-panel-user-acount:after",
// "header-panel-right-group:before",
// "header-panel-right-group:after",
// "header-panel-right-zoom:before",
// "header-panel-right-zoom:after",
// "header-panel-right-share:before",
// "header-panel-right-share:after",
// "insert-toolbar:before",
// "insert-toolbar:after",
// "insert-toolbar-menu:before",
// "insert-toolbar-menu:after",
// "insert-toolbar-default:before",
// "insert-toolbar-default:after",
// "insert-toolbar-rectangle:before",
// "insert-toolbar-rectangle:after",
// "insert-toolbar-ellipse:before",
// "insert-toolbar-ellipse:after",
// "insert-toolbar-image:before",
// "insert-toolbar-image:after",
// "insert-toolbar-text:before",
// "insert-toolbar-text:after",
// "insert-toolbar-icon:before",
// "insert-toolbar-icon:after",
// "left-panel-expandable-layer:before",
// "left-panel-expandable-layer:after",
// "left-panel-expandable-shared-components:before",
// "left-panel-expandable-shared-components:after",
// "code-panel-header-center",
// "code-panel-header-cog-options",
// "code-panel-header-cog:before",
// "code-panel-header-cog:after",
// "code-panel-header-cog-options",
// "code-panel-header-code-engines:before",
// "code-panel-header-code-engines:after",
// "code-panel-header-donwload:before",
// "code-panel-header-donwload:after"
// "overlay",
// "window-menu:before",
// "window-menu:after"
// ]
