// @builderx-core/builderx/src/code-engine/react/layers/SwiperCodeGen.ts
import { Context, ExtraParams } from "../types";
import { generateChildren, addImportToContext } from "../utils";
import LayerCodeGen from "../LayerCodeGen";

export default class SwiperCodeGen extends LayerCodeGen {
  context: Context;
  constructor(context: Context) {
    super();
    this.context = context;
  }
  generate(json: any, extraParams: ExtraParams): string {
    this.context.styles[json.styleName] = {
      style: json.props.style,
      layerName: "ThirdParty"
    };
    addImportToContext(this.context, json);
    json.props = {
      ...json.props,
      ...{
        totalSlides: json.children.length,
        naturalSlideWidth: json.props.style.width,
        naturalSlideHeight: json.props.style.width
      }
    };
    return `
          <CarouselProvider ${this.generateProps(
            json,
            extraParams
          )} style={${this.generateJSLiteralCode(
      json.props.style,
      extraParams
    )}}>
         <CarouselSlider>
          ${generateChildren(json, this.context.baseCodeGen.generate, {
            parent: json
          })}
          </CarouselSlider>
          </CarouselProvider>
      `;
  }
}
