// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/Footer/basicFooter.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, StyleSheet, Text, TouchableOpacity } from "react-native";
export default class Basic extends Component {
  static containerStyle = {
    height: 56,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <TouchableOpacity style={styles.buttonWrapper1}>
          <Icon
            name="television"
            style={styles.icon1}
            type="MaterialCommunityIcons"
          />
          <Text style={styles.btn1Text}>Movies</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.activeButtonWrapper}>
          <Icon
            name="music-note"
            style={styles.activeIcon}
            type="MaterialCommunityIcons"
          />
          <Text style={styles.activeText}>Music</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.buttonWrapper2}>
          <Icon
            name="book"
            style={styles.icon2}
            type="MaterialCommunityIcons"
          />
          <Text style={styles.btn2Text}>Books</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.buttonWrapper3}>
          <Icon
            name="calendar-text"
            style={styles.icon3}
            type="MaterialCommunityIcons"
          />
          <Text style={styles.btn3Text}>Calendar</Text>
        </TouchableOpacity>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    backgroundColor: "#3f51b5",
    flexDirection: "row",
    alignItems: "center",
    shadowColor: "#111",
    shadowOffset: { width: 0, height: -2 },
    shadowOpacity: 0.2,
    shadowRadius: 1.2,
    elevation: 3
  },
  buttonWrapper1: {
    flex: 1,
    paddingTop: 8,
    paddingBottom: 6,
    paddingHorizontal: 12,
    minWidth: 80,
    maxWidth: 168,
    alignItems: "center",
    flexDirection: "column"
  },
  buttonWrapper2: {
    flex: 1,
    paddingTop: 8,
    paddingBottom: 6,
    paddingHorizontal: 12,
    minWidth: 80,
    maxWidth: 168,
    alignItems: "center",
    flexDirection: "column"
  },
  buttonWrapper3: {
    flex: 1,
    paddingTop: 8,
    paddingBottom: 6,
    paddingHorizontal: 12,
    minWidth: 80,
    maxWidth: 168,
    alignItems: "center",
    flexDirection: "column"
  },
  activeButtonWrapper: {
    flex: 1,
    paddingTop: 6,
    paddingBottom: 10,
    paddingHorizontal: 12,
    minWidth: 80,
    maxWidth: 168,
    alignItems: "center",
    flexDirection: "column"
  },
  icon1: {
    backgroundColor: "transparent",
    color: "#FFFFFF",
    fontSize: 24,
    opacity: 0.8
  },
  icon2: {
    backgroundColor: "transparent",
    color: "#FFFFFF",
    fontSize: 24,
    opacity: 0.8
  },
  icon3: {
    backgroundColor: "transparent",
    color: "#FFFFFF",
    fontSize: 24,
    opacity: 0.8
  },
  activeIcon: {
    backgroundColor: "transparent",
    color: "#FFFFFF",
    fontSize: 24,
    opacity: 1
  },
  btn1Text: {
    color: "#FFFFFF",
    opacity: 0.8,
    fontFamily: "Roboto"
  },
  btn2Text: {
    color: "#FFFFFF",
    opacity: 0.8,
    fontFamily: "Roboto"
  },
  btn3Text: {
    color: "#FFFFFF",
    opacity: 0.8,
    fontFamily: "Roboto"
  },
  activeText: {
    color: "#FFFFFF",
    fontSize: 14,
    opacity: 1,
    fontFamily: "Roboto",
    paddingTop: 4
  }
});
