// @builderx-core/builderx/src/form-builder/lib/FormBuilder/Editors/OrientationEditor/index.tsx
import * as React from "react";
import PositionEditor from "./PositionEditor";
import SizeEditor from "./SizeEditor";
import TransformEditor from "./TransformEditor";
import { WrapperDiv } from "components-x";
import { Layer } from "domain-x";
import { IEvent } from "common-interfaces";
import { Session } from "session";

export default class OrientationEditor extends React.PureComponent<
  {
    layers: Array<Layer>;
    rootLayer: Layer;
    session: Session;
    onEvent: (event: IEvent) => void;
    getBoundingBox: () => {
      x: number;
      y: number;
      width: number;
      height: number;
      rotation: number;
    };
    captureErrorForSentry: (err: Error) => void;
    getResizerHandlesForLayers: (layers: Array<Layer>) => Array<string>;
    additionalOptions?: any;
  },
  any
> {
  render() {
    if (
      !this.props.layers ||
      (this.props.layers[0] &&
        this.props.layers[0].import.layerName === "ScrollView" &&
        this.props.layers[0] === this.props.rootLayer)
    ) {
      return null;
    }
    return (
      <WrapperDiv id="app-tutorial-size-editor" className="mt-1">
        <PositionEditor {...this.props} />
        {!(
          this.props.additionalOptions && this.props.additionalOptions.hideSize
        ) && <SizeEditor {...this.props} />}
        <TransformEditor
          {...this.props}
          additionalOptions={this.props.additionalOptions}
        />
      </WrapperDiv>
    );
  }
}
