// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/List/list13.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, StyleSheet, FlatList, Text, Image } from "react-native";
export default class List13 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Image
                  source={require("../assets/cardImage.png")}
                  style={styles.avatarImageStyle}
                />
                <View style={styles.contentColor}>
                  <Text style={styles.rowTitle}>Two-line Item</Text>
                  <Text style={styles.rowSubTitle}>
                    BuilderX is a screen design tool which codes React Native
                    for you.
                  </Text>
                </View>
                <Icon
                  name="message"
                  style={styles.iconStyle}
                  type="MaterialCommunityIcons"
                />
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF",
    paddingTop: 8
  },
  list: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  rowBgColor: {
    flexDirection: "row",
    paddingLeft: 16,
    alignItems: "center",
    height: 88
  },
  iconStyle: {
    fontSize: 24,
    color: "#CCCCCC",
    position: "absolute",
    right: 16
  },
  rowTitle: {
    color: "#212121",
    fontSize: 16,
    paddingBottom: 8
  },
  rowSubTitle: {
    color: "#9E9E9E",
    fontSize: 13,
    lineHeight: 16
  },
  avatarImageStyle: {
    width: 40,
    height: 40,
    borderRadius: 20,
    backgroundColor: "#CCCCCC"
  },
  contentColor: {
    position: "absolute",
    left: 72,
    right: 0,
    paddingRight: 40,
    height: 88,
    justifyContent: "center",
    borderBottomWidth: 0.5,
    borderColor: "#BDBDBD"
  }
});
