// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/List/list3.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, StyleSheet, FlatList, Text } from "react-native";
export default class List3 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Icon
                  name="checkbox-blank"
                  style={styles.iconStyle}
                  type="MaterialCommunityIcons"
                />
                <View style={styles.contentColor}>
                  <Text style={styles.rowText}>Single-line Item</Text>
                </View>
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF",
    paddingTop: 8
  },
  list: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  rowBgColor: {
    flexDirection: "row",
    paddingLeft: 16,
    alignItems: "center",
    height: 48
  },
  rowText: {
    color: "#212121",
    fontSize: 16
  },
  iconStyle: {
    fontSize: 32,
    color: "#616161"
  },
  contentColor: {
    position: "absolute",
    left: 72,
    right: 0,
    flex: 1,
    paddingRight: 16,
    height: 48,
    justifyContent: "center",
    borderBottomWidth: 0.5,
    borderColor: "#BDBDBD"
  }
});
