// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/List/list2.js
import React, { Component } from "react";
import { View, StyleSheet, FlatList, Text } from "react-native";
export default class List2 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Text style={styles.rowText}>Single-line Item</Text>
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF",
    paddingTop: 4
  },
  list: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  rowBgColor: {
    flexDirection: "row",
    padding: 16,
    paddingLeft: 0,
    height: 40,
    borderBottomWidth: 0.5,
    borderBottomColor: "#BDBDBD",
    marginLeft: 16
  },
  rowText: {
    color: "#212121",
    fontSize: 13
  }
});
