// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/List/list14.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, StyleSheet, FlatList, Text } from "react-native";
export default class List14 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Text style={styles.text}>Two-line Item</Text>
                <Text style={styles.subText}>
                  BuilderX is a screen design tool which codes React Native for
                  you.
                </Text>
              </View>
            );
          }}
          ListHeaderComponent={({ highlighted }) => {
            return (
              <View style={styles.listHeaderColor}>
                <Text style={styles.headerText}>List Header</Text>
              </View>
            );
          }}
          ListFooterComponent={({ highlighted }) => {
            return (
              <View style={styles.listFooterColor}>
                <Text style={styles.footerText}>List Footer</Text>
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF",
    paddingTop: 8
  },
  list: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  rowBgColor: {
    padding: 16,
    paddingLeft: 0,
    height: 88,
    marginLeft: 16
  },
  text: {
    color: "#212121",
    fontSize: 16,
    paddingBottom: 8
  },
  subText: {
    color: "#9E9E9E",
    fontSize: 13,
    lineHeight: 16
  },
  listHeaderColor: {
    padding: 16,
    backgroundColor: "#fff"
  },
  headerText: {
    color: "#616161"
  },
  listFooterColor: {
    padding: 16,
    backgroundColor: "#fff"
  },
  footerText: {
    color: "#616161"
  }
});
