// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/List/list15.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, StyleSheet, FlatList, Text } from "react-native";
export default class List15 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Icon
                  name="school"
                  style={styles.iconStyle}
                  type="MaterialCommunityIcons"
                />
                <View style={styles.contentColor}>
                  <Text style={styles.rowText}>Single-line Item</Text>
                </View>
                <Icon
                  name="checkbox-marked"
                  style={styles.icon2Style}
                  type="MaterialCommunityIcons"
                />
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF",
    paddingTop: 8
  },
  list: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  rowBgColor: {
    flexDirection: "row",
    paddingLeft: 16,
    alignItems: "center",
    height: 56
  },
  iconStyle: {
    fontSize: 24,
    color: "#616161"
  },
  icon2Style: {
    fontSize: 24,
    color: "#616161",
    position: "absolute",
    right: 16
  },
  rowText: {
    color: "#212121",
    fontSize: 16
  },
  contentColor: {
    position: "absolute",
    left: 72,
    right: 0,
    paddingRight: 16,
    height: 56,
    justifyContent: "center"
  }
});
