// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Material/List/list10.js
import React, { Component } from "react";
import { View, StyleSheet, FlatList, Text } from "react-native";
export default class List10 extends Component {
  static containerStyle = {
    height: 500,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <FlatList
          style={styles.list}
          renderItem={({ item, separators }) => {
            return (
              <View style={styles.rowBgColor}>
                <Text style={styles.rowTitle}>Three-line Item</Text>
                <Text style={styles.rowSubTitle}>
                  BuilderX is a screen design tool which codes React Native for
                  you.
                </Text>
              </View>
            );
          }}
        />
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    flex: 1,
    backgroundColor: "#FFF",
    paddingTop: 8
  },
  list: {
    flex: 1,
    backgroundColor: "#FFF"
  },
  rowBgColor: {
    padding: 16,
    paddingLeft: 0,
    height: 88,
    marginLeft: 16
  },
  rowTitle: {
    color: "#212121",
    fontSize: 16,
    paddingBottom: 8
  },
  rowSubTitle: {
    color: "#9E9E9E",
    fontSize: 13,
    lineHeight: 16
  }
});
