// @builderx-core/builderx/src/domain-x/lib/services/ThirdPartyComponents/Carousel.ts
import { ComponentDefinition } from "common-interfaces";

const carouselSchema: ComponentDefinition = {
  version: "0.0.1",
  template: {
    type: "thirdParty",
    import: {
      layerName: "Swiper",
      packageName: "@builderx/react-native-swiper",
      isDefaultImport: true,
      asName: ""
    },
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    },
    props: {
      style: {
        position: "absolute",
        top: 50,
        left: 50
      },
      index: 0
    },
    children: [
      {
        type: "view",
        children: [],
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false,
          asName: ""
        },
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "#5C6BC0"
          }
        }
      },
      {
        type: "view",
        children: [],
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false,
          asName: ""
        },
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "#2196F3"
          }
        }
      },
      {
        type: "view",
        children: [],
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false,
          asName: ""
        },
        locked: false,
        template: {
          centerVertical: false,
          centerHorizontal: false
        },
        props: {
          style: {
            flex: 1,
            backgroundColor: "#009688"
          }
        }
      }
    ]
  },
  thumbnail: "",
  name: "Swiper",
  controlsProperty: [
    {
      type: "AlignmentEditor",
      title: "Alignment Editor",
      value: {},
      showForMultiple: true
    },
    {
      type: "OrientationEditor",
      title: "Orientation Editor",
      value: {},
      showForMultiple: true
    },
    {
      type: "ResizingTypeEditor",
      title: "ResizingType Editor",
      value: {}
    },
    {
      type: "group",
      title: "Carousel",
      value: {},
      properties: [
        {
          type: "CarouselEditor",
          title: "Carousel Editor",
          value: {}
        },
        {
          propertyName: "loop",
          type: "boolean",
          title: "Loop",
          value: {
            defaultValue: true
          },
          category: "props"
        },
        {
          propertyName: "autoplay",
          type: "boolean",
          title: "Auto Play",
          value: {
            defaultValue: false
          },
          category: "props"
        },
        {
          propertyName: "autoplayTimeout",
          type: "number",
          title: "Autoplay Timeout",
          value: {
            defaultValue: 2.5
          },
          category: "props"
        },
        {
          propertyName: "autoplayDirection",
          title: "Autoplay Direction",
          type: "enum",
          value: {
            options: ["Clockwise", "Anti clockwise"],
            defaultValue: "Clockwise"
          },
          category: "props"
        },
        {
          propertyName: "showsButtons",
          type: "boolean",
          title: "Show Buttons",
          value: {
            defaultValue: false
          },
          category: "props"
        },
        {
          propertyName: "showsPagination",
          type: "boolean",
          title: "Show Pagination",
          value: {
            defaultValue: true
          },
          category: "props"
        },
        {
          propertyName: "horizontal",
          type: "boolean",
          title: "Horizontal",
          value: {
            defaultValue: true
          },
          category: "props"
        },
        {
          propertyName: "bounces",
          type: "boolean",
          title: "Bounces",
          value: {
            defaultValue: false
          },
          category: "props"
        }
      ]
    },
    {
      type: "group",
      title: "Styling",
      value: {},
      properties: [
        {
          propertyName: "backgroundColor",
          type: "color",
          title: "Fill",
          value: {
            defaultValue: "#fff"
          },
          category: "style"
        },
        {
          type: "BorderWidthEditor",
          title: "BorderWidth Editor",
          value: {}
        },
        {
          type: "BorderRadiusEditor",
          title: "BorderRadius Editor",
          value: {}
        }
      ]
    },
    {
      type: "group",
      title: "Layout",
      value: {},
      properties: [
        {
          type: "MarginPaddingEditor",
          title: "MarginPadding Editor",
          value: {}
        }
      ]
    }
  ],
  componentRenderDefinition: `class ThirdPartyFlatList extends React.Component {
    constructor() {      super();      const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 });
    this.state = { dataSource: ds.cloneWithRows([\"row 1\", \"row 2\"]) };    }    render() {      const props = {
      this.props,        style:{ ...this.props.style},          dataSource:this.state.dataSource,
                renderRow:rowData => <Text>{rowData}</Text>      };      return (
                  <ListView          {...props}        />      );    }  }`
};

export default carouselSchema;
