// @builderx-core/builderx/src/domain-x/lib/services/ThirdPartyComponents/FlatList.ts
import { ComponentDefinition } from "common-interfaces";

const faltListSchema: ComponentDefinition = {
  version: "0.0.1",
  template: {
    type: "thirdParty",
    import: {
      layerName: "FlatList",
      packageName: "react-native",
      isDefaultImport: false,
      asName: ""
    },
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    },
    props: {
      style: {
        top: 50,
        left: 50,
        width: 50,
        height: 50,
        position: "absolute"
      },
      renderItem: {
        type: "view",
        import: {
          layerName: "View",
          packageName: "native",
          isDefaultImport: false,
          asName: ""
        },
        template: {
          centerVertical: false,
          centerHorizontal: false
        },
        props: {
          style: {
            padding: 16,
            height: 48,
            width: "100%",
            backgroundColor: "#fff"
          }
        },
        children: [
          {
            type: "text",
            import: {
              layerName: "Text",
              packageName: "native",
              isDefaultImport: false,
              asName: ""
            },
            props: {
              style: {
                fontSize: 14,
                color: "#121212"
              }
            },
            children: [
              {
                type: "literal",
                value: "List Item"
              }
            ],
            locked: false,
            template: {
              centerVertical: false,
              centerHorizontal: false
            }
          }
        ],
        locked: false
      }
    },
    children: []
  },
  thumbnail: "",
  name: "FlatList",

  controlsProperty: [
    {
      type: "AlignmentEditor",
      title: "Alignment Editor",
      value: {},
      showForMultiple: true
    },
    {
      type: "OrientationEditor",
      title: "Orientation Editor",
      value: {},
      showForMultiple: true
    },
    {
      type: "ResizingTypeEditor",
      title: "ResizingType Editor",
      value: {}
    },
    // {
    //   type: "renderProp",
    //   title: "List Item",
    //   propertyName: "renderItem",
    //   category: "props",
    //   value: {
    //     defaultValue: {
    //       type: "view",
    //       import: {
    //         layerName: "View",
    //         packageName: "native",
    //         isDefaultImport: false,
    //         asName: ""
    //       },
    //       template: {
    //         centerVertical: false,
    //         centerHorizontal: false
    //       },
    //       props: {
    //         style: {
    //           padding: 16,
    //           height: 48,
    //           width: "100%",
    //           backgroundColor: "#fff"
    //         }
    //       },
    //       children: [
    //         {
    //           type: "text",
    //           import: {
    //             layerName: "Text",
    //             packageName: "native",
    //             isDefaultImport: false,
    //             asName: ""
    //           },
    //           props: {
    //             style: {
    //               fontSize: 14,
    //               color: "#121212"
    //             }
    //           },
    //           children: [
    //             {
    //               type: "literal",
    //               value: "List Item"
    //             }
    //           ],
    //           locked: false,
    //           template: {
    //             centerVertical: false,
    //             centerHorizontal: false
    //           }
    //         }
    //       ],
    //       locked: false
    //     }
    //   }
    // },
    // {
    //   type: "component",
    //   title: "List Header",
    //   propertyName: "ListHeaderComponent",
    //   category: "props",
    //   value: {
    //     defaultValue: {
    //       type: "view",
    //       import: {
    //         layerName: "View",
    //         packageName: "native",
    //         isDefaultImport: false,
    //         asName: ""
    //       },
    //       props: {
    //         style: {
    //           height: 48,
    //           width: "100%",
    //           backgroundColor: "#fff"
    //         }
    //       },
    //       children: [
    //         {
    //           type: "text",
    //           import: {
    //             layerName: "Text",
    //             packageName: "native",
    //             isDefaultImport: false,
    //             asName: ""
    //           },
    //           props: {
    //             style: {
    //               padding: 16,
    //               color: "#616161"
    //             }
    //           },
    //           children: [
    //             {
    //               type: "literal",
    //               value: "List Header"
    //             }
    //           ],
    //           locked: false,
    //           template: {
    //             centerHorizontal: false,
    //             centerVertical: false
    //           }
    //         }
    //       ],
    //       locked: false,
    //       template: {
    //         centerHorizontal: false,
    //         centerVertical: false
    //       }
    //     }
    //   }
    // },
    // {
    //   type: "component",
    //   title: "List Footer",
    //   propertyName: "ListFooterComponent",
    //   category: "props",
    //   value: {
    //     defaultValue: {
    //       type: "view",
    //       import: {
    //         layerName: "View",
    //         packageName: "native",
    //         isDefaultImport: false,
    //         asName: ""
    //       },
    //       props: {
    //         style: {
    //           height: 48,
    //           width: "100%",
    //           backgroundColor: "#fff"
    //         }
    //       },
    //       children: [
    //         {
    //           type: "text",
    //           import: {
    //             layerName: "Text",
    //             packageName: "native",
    //             isDefaultImport: false,
    //             asName: ""
    //           },
    //           props: {
    //             style: {
    //               padding: 16,
    //               color: "#616161"
    //             }
    //           },
    //           children: [
    //             {
    //               type: "literal",
    //               value: "List Footer"
    //             }
    //           ],
    //           locked: false,
    //           template: {
    //             centerHorizontal: false,
    //             centerVertical: false
    //           }
    //         }
    //       ],
    //       locked: false,
    //       template: {
    //         centerHorizontal: false,
    //         centerVertical: false
    //       }
    //     }
    //   }
    // },
    // {
    //   type: "component",
    //   title: "Item Separator",
    //   propertyName: "ItemSeparatorComponent",
    //   category: "props",
    //   value: {
    //     defaultValue: {
    //       type: "view",
    //       import: {
    //         layerName: "View",
    //         packageName: "react-native",
    //         isDefaultImport: false,
    //         asName: ""
    //       },
    //       locked: false,
    //       template: {
    //         centerVertical: false,
    //         centerHorizontal: false
    //       },
    //       props: {
    //         style: {
    //           width: "100%",
    //           height: 1,
    //           backgroundColor: "black"
    //         }
    //       },
    //       children: []
    //     }
    //   }
    // },
    {
      type: "boolean",
      title: "Horizontal",
      propertyName: "horizontal",
      category: "props",
      value: {
        defaultValue: false
      }
    },
    {
      type: "group",
      title: "Styling",
      value: {},
      properties: [
        {
          propertyName: "backgroundColor",
          type: "color",
          title: "Fill",
          value: {
            defaultValue: "#fff"
          },
          category: "style"
        },
        {
          type: "BorderWidthEditor",
          title: "BorderWidth Editor",
          value: {}
        },
        {
          type: "BorderRadiusEditor",
          title: "BorderRadius Editor",
          value: {}
        },
        {
          type: "ShadowEditor",
          title: "Shadow Editor",
          value: {}
        }
      ]
    }
  ],
  componentRenderDefinition: `class ThirdPartyFlatList extends React.Component {    constructor() {      super();
      const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 });
      this.state = { dataSource: ds.cloneWithRows([\"row 1\", \"row 2\"]) };    }    render() {
        const props = {        ...this.props,        style:{ ...this.props.style},
        dataSource:this.state.dataSource,          renderRow:rowData => <Text>{rowData}</Text>      };
        return (        <ListView          {...props}        />      );    }  }`
};

export default faltListSchema;
