// @builderx-core/builderx/src/domain-x/lib/services/ThirdPartyComponents/Slider.ts
import { ComponentDefinition } from "common-interfaces";

const sliderSchema: ComponentDefinition = {
  version: "0.0.1",
  template: {
    type: "thirdParty",
    import: {
      layerName: "Slider",
      packageName: "react-native",
      isDefaultImport: false,
      asName: ""
    },
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    },
    props: {
      style: {
        position: "absolute",
        top: 50,
        left: 50,
        height: 50,
        width: 50
      },
      value: 50
    },
    children: []
  },
  thumbnail: "",
  name: "Slider",
  controlsProperty: [
    {
      type: "AlignmentEditor",
      title: "Alignment Editor",
      value: {},
      showForMultiple: true
    },
    {
      type: "OrientationEditor",
      title: "Orientation Editor",
      value: {},
      showForMultiple: true
    },
    {
      type: "ResizingTypeEditor",
      title: "ResizingType Editor",
      value: {}
    },
    {
      type: "group",
      title: "Styling",
      value: {},
      properties: [
        {
          propertyName: "opacity",
          type: "range",
          title: "Opacity",
          value: {
            defaultValue: 1,
            min: 0,
            max: 1
          },
          category: "style"
        }
      ]
    },
    {
      type: "group",
      title: "Slider",
      value: {},
      properties: [
        {
          propertyName: "disabled",
          type: "boolean",
          title: "Disabled",
          value: {
            defaultValue: false
          },
          category: "props"
        },
        {
          propertyName: "value",
          type: "range",
          title: "Value",
          value: {
            defaultValue: 0,
            min: 0,
            max: 100
          },
          category: "props"
        },
        {
          propertyName: "minimumValue",
          type: "number",
          title: "Minimum Value",
          value: {
            defaultValue: 0
          },
          category: "props"
        },
        {
          propertyName: "maximumValue",
          type: "number",
          title: "Maximum Value",
          value: {
            defaultValue: 100
          },
          category: "props"
        },
        {
          propertyName: "step",
          type: "number",
          title: "Step",
          value: {
            defaultValue: 0
          },
          category: "props"
        },
        {
          propertyName: "minimumTrackTintColor",
          type: "bareColor",
          title: "Minimum Track",
          value: {
            defaultValue: "#000"
          },
          category: "props"
        },
        {
          propertyName: "maximumTrackTintColor",
          type: "bareColor",
          title: "Maximum Track",
          value: {
            defaultValue: "#000"
          },
          category: "props"
        },
        {
          propertyName: "thumbTintColor",
          type: "bareColor",
          title: "Thumb",
          value: {
            defaultValue: "#000"
          },
          category: "props"
        }
      ]
    }
  ],
  componentRenderDefinition: `class ThirdPartyFlatList extends React.Component {    constructor() {      super();
      const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 });
      this.state = { dataSource: ds.cloneWithRows([\"row 1\", \"row 2\"]) };    }    render() {
        const props = {        ...this.props,        style:{ ...this.props.style},
         dataSource:this.state.dataSource,          renderRow:rowData => <Text>{rowData}</Text>
         };      return (        <ListView          {...props}        />      );    }  }`
};

export default sliderSchema;
