// @builderx-core/builderx/src/domain-x/lib/services/ThirdPartyComponents/Switch.ts
import { ComponentDefinition } from "common-interfaces";

const switchSchema: ComponentDefinition = {
  version: "0.0.1",
  template: {
    type: "thirdParty",
    import: {
      layerName: "Switch",
      packageName: "react-native",
      isDefaultImport: false,
      asName: ""
    },
    locked: false,
    template: {
      centerVertical: false,
      centerHorizontal: false
    },
    props: {
      style: {
        position: "absolute",
        top: 50,
        left: 50
      },
      value: true
    },
    children: []
  },
  thumbnail: "",
  name: "Switch",

  controlsProperty: [
    {
      type: "AlignmentEditor",
      title: "Alignment Editor",
      value: {},
      showForMultiple: true
    },
    {
      type: "OrientationEditor",
      title: "Orientation Editor",
      value: {},
      showForMultiple: true,
      additionalOptions: {
        hideSize: true
      }
    },
    {
      type: "ResizingTypeEditor",
      title: "ResizingType Editor",
      value: {},
      additionalOptions: {
        hideSize: true
      }
    },
    {
      type: "group",
      title: "Styling",
      value: {},
      properties: [
        {
          propertyName: "opacity",
          type: "range",
          title: "Opacity",
          value: {
            defaultValue: 1,
            min: 0,
            max: 1
          },
          category: "style"
        }
      ]
    },
    {
      type: "group",
      title: "Switch",
      value: {},
      properties: [
        {
          propertyName: "value",
          category: "props",
          type: "boolean",
          title: "Value",
          value: {
            defaultValue: false
          }
        },
        {
          propertyName: "disabled",
          category: "props",
          type: "boolean",
          title: "Disabled",
          value: {
            defaultValue: false
          }
        },
        {
          propertyName: "thumbTintColor",
          type: "bareColor",
          title: "Grip",
          value: {
            defaultValue: "#fff"
          },
          category: "props"
        },
        {
          propertyName: "onTintColor",
          type: "bareColor",
          title: "Turned On",
          value: {
            defaultValue: "#fff"
          },
          category: "props"
        },
        {
          propertyName: "tintColor",
          type: "bareColor",
          title: "Turned Off",
          value: {
            defaultValue: "#fff"
          },
          category: "props"
        }
      ]
    }
  ],
  componentRenderDefinition: `class ThirdPartyFlatList extends React.Component {    constructor() {      super();
      const ds = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 });
      this.state = { dataSource: ds.cloneWithRows([\"row 1\", \"row 2\"]) };    }    render() {
        const props = {        ...this.props,        style:{ ...this.props.style},
        dataSource:this.state.dataSource,          renderRow:rowData => <Text>{rowData}</Text>      };
      return (        <ListView          {...props}        />      );    }  }`
};

export default switchSchema;
