// @builderx-core/builderx/src/session-plugins/lib/plugins/ExternalDesignerPluginsProvider/index.tsx
import { ISessionPlugin } from "common-interfaces";
import { Session } from "session";
import { action } from "model";

export default class ExternalDesignerPluginsProvider implements ISessionPlugin {
  name = "ExternalDesignerPluginsProvider";
  session: Session;
  externalDesignerPlugins: Array<any>;
  init(session: Session) {
    this.session = session;
    this.externalDesignerPlugins = [];
  }

  activate() {
    this.session.addFunction(
      `${this.name}.add`,
      (session: Session, pluginData: any) => {
        this.externalDesignerPlugins.push(pluginData);
        action(() => {
          this.session.emit("externalDesignerPlugins");
        });
        return this.externalDesignerPlugins.length - 1;
      }
    );
    this.session.addFunction(
      `${this.name}.get`,
      () => this.externalDesignerPlugins
    );
    this.session.addFunction(
      `${this.name}.remove`,
      (session: Session, id: any) => {
        this.externalDesignerPlugins.splice(id, 1);
        this.session.emit("externalDesignerPlugins");
      }
    );
  }
  deactivate() {
    //
  }
}
