// @builderx-core/builderx/src/session-plugins/lib/plugins/TestPlugin/index.tsx
import * as React from "react";
import { ISessionPlugin } from "common-interfaces";
import { Session } from "session";
import { LayerCollection } from "domain-x";
export default class TestPlugin implements ISessionPlugin {
  name = "TestPlugin";
  session: Session;
  init(session: Session) {
    this.session = session;
  }

  activate() {
    if (this.session.isPluginLoaded("ExternalComponentsProvider")) {
      // NOTE: uncomment for WindowAPIProvider plugin example
      // let windowId = this.session.executeFunction(
      //   "WindowAPIProvider.add",
      //   (props: any) => <h1>Hello</h1>,
      //   () => <h4>Footer</h4>,
      //   { backdropBackground: 200, modalSize: "m", background: 800 }
      // );
      this.session.executeFunction("ExternalComponentsProvider.add", {
        id: `placement-example-header-panel-left`,
        placement: "header-panel-left",
        component: () => (
          <>
            <button
              onClick={() => {
                //
              }}
              style={{ height: 20, width: 50 }}
            >
              Click1
            </button>
          </>
        )
      });
      // NOTE: MenuAPIProvider example
      // this.session.executeFunction(
      //   "MenuProvider.add",
      //   "insert-toolbar:before"
      //   "test",
      //   (e: any) => console.log("*** 🔥herer"),
      // );
      // this.session.executeFunction("MenuProvider.add", () => {
      //   return {
      //     caption: "test",
      //     // callback: (e: any) => console.log("*** 🔥herer"),
      //     placement: "insert-toolbar-menu:after",
      //     shortcut: "mod+b",
      //     // rightIcon: "right-arrow",
      //     children: [
      //       () => {
      //         return {
      //           caption: "test nest",
      //           callback: (e: any) => console.log("*** 🔥herer in 1"),
      //           shortcut: "mod+e"
      //         };
      //       }
      //       // () => {
      //       //   return {
      //       //     caption: "test nest 2",
      //       //     callback: (e: any) => console.log("*** 🔥herer in 2"),
      //       //     rightIcon: "right-arrow",
      //       //     children: [
      //       //       () => {
      //       //         return {
      //       //           caption: "test nest 2 1",
      //       //           callback: (e: any) => console.log("*** 🔥herer in 2 1"),
      //       //           shortcut: "T3"
      //       //         };
      //       //       }
      //       //     ]
      //       //   };
      //       // }
      //     ]
      //   };
      // });
      // this.session.executeFunction("MenuProvider.add", () => {
      //   return {
      //     placement: "window-menu:before",
      //     caption: "test nest",
      //     callback: (e: any) => console.log("*** 🔥herer in 1"),
      //     shortcut: "T1"
      //   };
      // });
      // this.session.executeFunction("MenuProvider.add", () => {
      //   return {
      //     placement: "window-menu:after",
      //     caption: "test nest",
      //     callback: (e: any) => console.log("*** 🔥herer in 1"),
      //     shortcut: "mod+b"
      //   };
      // });
      // this.session.executeFunction("MenuProvider.add", () => {
      //   return {
      //     placement: "user-account-menu:before",
      //     caption: "test nest",
      //     callback: (e: any) => console.log("*** 🔥herer in 1"),
      //     shortcut: "mod+u"
      //   };
      // });
      // this.session.executeFunction("MenuProvider.add", () => {
      //   return {
      //     placement: "user-account-menu:after",
      //     caption: "test nest",
      //     callback: (e: any) => console.log("*** 🔥herer in 1"),
      //     shortcut: "T1"
      //   };
      // });
    }
  }
  deactivate() {
    //
  }
}
