// @builderx-core/builderx/src/realtime-io/demo/App.tsx
import * as React from "react";
import { apiSdk as ApiSdk } from "@builderx-core/api-sdk";

class App extends React.Component {
  public render() {
    const io = ApiSdk.poll.connect(
      "http://localhost:8080/poll",
      {
        project_id: "1ii7a92j4l8kcgc0ockgc8cwskww48",
        session_id: "skldjflskjdf"
      },
      {
        headers: {
          Accept: "application/json",
          // tslint:disable-next-line:max-line-length
          Authorization: `Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImp0aSI6IjRjOTA5M2QyNjEyZDZhMzBiZTI1YWU2YTQ3Y2VhMmRhMmRjOGM1MTE1ZGEyZTQzN2ZjZDUwNGE1YmY5ODM2MDQ3NzZmNzFmZmY0NzdmN2VlIn0.eyJhdWQiOiIxIiwianRpIjoiNGM5MDkzZDI2MTJkNmEzMGJlMjVhZTZhNDdjZWEyZGEyZGM4YzUxMTVkYTJlNDM3ZmNkNTA0YTViZjk4MzYwNDc3NmY3MWZmZjQ3N2Y3ZWUiLCJpYXQiOjE1NjYzODMxNjksIm5iZiI6MTU2NjM4MzE2OSwiZXhwIjoxNTk4MDA1NTY5LCJzdWIiOiIxMyIsInNjb3BlcyI6W119.JyEgk8mAUp-UZ3Z5-hDUJLtxRzVHiPZBa3Jv95S0Rm1hm3YMDe5bWApAYBXDNKJWdI9Adn3tNt2peIUZ9m6l4ADJF-EEbdsVRNNwJ36oJVw9Ta8hosV1pZ-SOl4HxQ3VtnVXqbvCyPkiJEErGWKmwsXPXy1bLvE_qFw5aH_15KEuAi2MgXWkDHPthn_RJRf4zumSl6h_khcqDxabvcuSlG6xXsseCgEjAkpcppCBCz_5hNjtfT__TKHbGP_IkwV30jmqUOIiIq58jClRalUAk4o0h9ew7DUZ-8NTJa_GFnWqXy1VyNTilr8eGevGYVJRZy3B8cuPiwxL2MZR-prefPtapyFMwtyVTgD2etfbbajn0SbuYmlGMRNQABmEFcvjHHwN8pj49-9LuI45QAj5c_CHIWRtkGV71a-UdlO7bUzw81T52Z1J0FI2q7hSwOqdlAMWlfXCPu58UMFpK8qngnh6yUwovriowplklhS01HV3TuUpQNw2ru3hQtHV1oM8iVk8RtF7nOaWtSVvb0ivmWy_XZAba0grZ_MCPLOugZAhbBcFuxe5fAl140cFI0_lgV7rOIsKGZz6QvyO61iQO3FLOBpkpO3FpJv6yyzFmgrDMPRpGxug-98SiwxdAJzTiL1sgskN_996k8geoyVTzneOjeepPgMUp3sTLYifWiQ`
        }
      }
    );
    io.on("connect", () => {
      console.log("connected");
    });

    io.on("connection-failed", (data: any) => {
      console.log("connection failed", data);
    });

    io.on("poll", (data: any) => {
      console.log("poll", data);
    });

    return (
      <div>
        <p>Use this sample project from test. {Text}</p>
      </div>
    );
  }
}

export default App;
