// @builderx-core/builderx/src/shared-components/SharedComponentsJS/Cupertino/Header/headerWithLargeTitle.js
import React, { Component } from "react";
import Icon from "@builderx/icons";
import { View, TouchableOpacity, Text, StyleSheet } from "react-native";
export default class CupertinoHeader5 extends Component {
  static containerStyle = {
    height: 96,
    width: 375
  };
  render() {
    return (
      <View style={[styles.root, this.props.style]}>
        <View style={styles.header}>
          <View style={styles.leftWrapper}>
            <TouchableOpacity style={styles.leftIconButton}>
              <Icon
                style={styles.leftIcon}
                name="ios-arrow-back"
                type="Ionicons"
              />
              <Text style={styles.leftText}>Back</Text>
            </TouchableOpacity>
          </View>
          <View style={styles.rightWrapper}>
            <TouchableOpacity>
              <Text style={styles.rightText}>Action</Text>
            </TouchableOpacity>
          </View>
        </View>
        <View style={styles.textWrapper}>
          <Text style={styles.title} numberOfLines={1}>
            Large Title
          </Text>
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  root: {
    backgroundColor: "#EFEFF4",
    paddingRight: 8,
    paddingLeft: 8
  },
  header: {
    width: "100%",
    height: 44,
    flexDirection: "row"
  },
  leftWrapper: {
    flex: 1,
    alignItems: "flex-start",
    justifyContent: "center"
  },
  leftIconButton: {
    flexDirection: "row"
  },
  leftIcon: {
    color: "#007AFF",
    fontSize: 32
  },
  leftText: {
    fontSize: 17,
    color: "#007AFF",
    paddingLeft: 5,
    alignSelf: "center"
  },
  textWrapper: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center"
  },
  title: {
    fontSize: 17,
    fontWeight: "600",
    fontFamily: "Roboto",
    lineHeight: 17,
    color: "#000"
  },
  rightWrapper: {
    flex: 1,
    alignItems: "flex-end",
    justifyContent: "center"
  },
  rightText: {
    color: "#007AFF",
    fontSize: 17,
    alignSelf: "center"
  },
  textWrapper: {
    height: 52,
    paddingLeft: 5,
    justifyContent: "center"
  },
  title: {
    fontSize: 34,
    fontWeight: "600",
    fontFamily: "Roboto",
    lineHeight: 40,
    color: "#000"
  }
});
